---
title: "R Notebook"
output: html_notebook
editor_options: 
  chunk_output_type: console
---

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute 
code within the notebook, the results appear beneath the code. 

Try executing this chunk by clicking the *Run* button within the chunk or by 
placing your cursor inside it and pressing *Ctrl+Shift+Enter*. 

Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by 
pressing *Ctrl+Alt+I*.

When you save the notebook, an HTML file containing the code and output will be 
saved alongside it (click the *Preview* button or press *Ctrl+Shift+K* to 
preview the HTML file).

The preview shows you a rendered HTML copy of the contents of the editor. 
Consequently, unlike *Knit*, *Preview* does not run any R code chunks. Instead, 
the output of the chunk when it was last run in the editor is displayed.



#loading required libraries and setting up workspace
```{r}
#cleaning the work environment
rm(list = ls(all.names = T))

#load libraries
{
  library(Seurat)
  library(ggplot2)
  library(ggExtra)
  library(sctransform)
  library(future)
  library(DESeq2)
  library(plotly)
  library(reshape2)
  library(scDblFinder)
  library(BiocParallel)
  library(slingshot)
  library(patchwork)
  library(igraph)
  library(SeuratWrappers)
  library(scales)
  library(monocle)
  library(dplyr)
  library(tidyverse)
  library(ggrepel)
  library(ggpubr)
  library(rstatix)
  library(stringr)
  library(ggpattern)
  library(gage)
  library(pathview)
  library(cluster)
  library(ggrastr)
  library(loomR)
}

#setting up workspace and location variables
{
  old_wrkdir <- getwd()
  setwd("/home/tareens/Workspace/AAV_IL2/")
  loc_workspace <- getwd()
  
  loc_genomes <- paste0(loc_workspace, "/00_genomes/")
  loc_raw_data <- paste0(loc_workspace, "/00_raw_data/")
  loc_metadata <- paste0(loc_workspace, "/00_metadata/")
  loc_scripts <- paste0(loc_workspace, "/00_scripts/")
  
  loc_04mv24m_results <- paste0(loc_workspace, "/04_04mv24m")
  loc_ext_data <- paste0(loc_04mv24m_results, "/04_other_datasets/")
  loc_cellranger_results <- paste0(loc_04mv24m_results, 
                                   "/01_cellranger_results/")
  loc_seurat_results <- paste0(loc_04mv24m_results, 
                               "/02_seurat_results/")
  loc_superclusters <- paste0(loc_04mv24m_results, 
                               "/03_superclusters/")
  loc_other_datasets <- paste0(loc_04mv24m_results, 
                               "/04_other_datasets/")
  loc_monocle <- paste0(loc_04mv24m_results, 
                        "/05_monocle/")
  loc_diff_exprs_results <- paste0(loc_04mv24m_results, 
                                   "/06_diff_exprs_results/")
  loc_pathview_results <- paste0(loc_04mv24m_results, 
                                   "/07_pathview_results/")
  
}

#defining functions
{
  # get plots of cell proportions by age in different clusters
  get_cell_counts_props <- function(seurat_object, 
                                    cluster_column = "seurat_clusters", 
                                    category_column = "label", 
                                    proportions = T, 
                                    display_stats = T
                                    ){
    
    if(!cluster_column %in% colnames(seurat_object@meta.data)){
      stop(paste0("Error: specified cluster column not found in the seurate ", 
                  "object. Did you forget to run the respective clustering?"))
      }
    if(!category_column %in% colnames(seurat_object@meta.data)){
      stop(paste0("Error: specified category column not found in the ", 
                  "seurate object. Double-check the column names of the meta ", 
                  "data."))
      }
    if(length(category_column) > 1){
      stop(paste0("Error: please provide only a single category column."))
      }
    if(length(cluster_column) > 1){
      stop(paste0("Error: please provide only a single cluster column."))
      }
    
    plot_data <- table(seurat_object@meta.data[,category_column], 
                       seurat_object@meta.data[,cluster_column])
    
    if(proportions){
      plot_data <- (plot_data/rowSums(plot_data))*100
    }
    
    plot_data <- melt(plot_data)
    
    #formatting age_treatment grouping for x-axis
    plot_data$age <- gsub(pattern = "_.*$", 
                             replacement = "", 
                             x = plot_data$Var1)
    plot_data$sample <- gsub(pattern = "^.*_", 
                                replacement = "", 
                                x = plot_data$Var1)
    plot_data$treatment <- "PHP.GFAP-GFP"
    plot_data$treatment[
      which((plot_data$age %in% c("12m") &
               plot_data$sample %in% c("S01", "S03")) |
              (!(plot_data$age %in% c("12m")) &
                 plot_data$sample %in% c("S02", "S04", "S06")) )] <-
          "PHP.GFAP-IL2"
    plot_data$age_treatment <- paste0(plot_data$age, 
                                         #"_", 
                                      ".", 
                                      plot_data$treatment)
    
    colplot_list <- list()
    
    for(i in unique(plot_data$Var2)){
      
      colplot_data <- plot_data[plot_data$Var2==i,]
        
      if(!("Var1" %in% colnames(colplot_data))){
        colplot_data$Var1 <- rownames(colplot_data)
        rownames(colplot_data) <- seq(1, nrow(colplot_data))
      }
      
      col_plot <- 
        ggplot(data = colplot_data, 
               aes(y = value, 
                   x = age)) + 
        geom_dotplot(aes(fill = treatment), 
                     binaxis = "y", 
                     position = "dodge", 
                     stackdir = "center", 
                     dotsize = 1.5) + 
        stat_summary(aes(fill = treatment), 
                     fun.data=mean_sdl, fun.args = list(mult=1),
                     geom="errorbar", color="black", width=0.2, 
                     position = position_dodge(0.9)) +
        stat_summary(aes(fill = treatment), 
                     fun.y=mean, geom="point", color="black", size = 0.8, 
                     position = position_dodge(0.9)) + 
        labs(fill = "Treatment", 
             y = "Percentage of cells", 
             x = "Age group", 
             subtitle = as.character(i)
             ) + 
        expand_limits(y = 0) + 
        theme_light(base_size = 15) + 
        theme(panel.grid.minor = element_blank(), 
              legend.position = "top")
      
      if(display_stats){
        stat.test.within.age <- colplot_data %>%
          group_by(age) %>%
          t_test(formula = value ~ treatment) %>%
          adjust_pvalue(method = "bonferroni") %>%
          add_significance("p.adj")
        
        stat.test.within.age <- stat.test.within.age %>% 
          add_xy_position(x = "age", dodge = 0.8)
        
        stat.test.within.age$y.position <- stat.test.within.age$y.position + 
          (stat.test.within.age$y.position * 0.05)
        
        col_plot <- col_plot + 
          stat_pvalue_manual(data = stat.test.within.age, 
                             label = "p.adj", tip.length = 0, size = 4.5) 
        
        stat.test.bw.age <- colplot_data %>%
          t_test(formula = value ~ age) %>%
          adjust_pvalue(method = "bonferroni")
        
        stat.test.bw.age <- stat.test.bw.age %>% 
          add_xy_position(x = "age")
        
        stat.test.bw.age$y.position <- stat.test.bw.age$y.position + 
          (stat.test.bw.age$y.position * 0.2)
        
        col_plot <- col_plot + 
          stat_pvalue_manual(stat.test.bw.age, label = "p.adj", 
                             tip.length = 0.02, step.increase = 0.5, size = 4.5) + 
          scale_y_continuous(expand = expansion(mult = c(0.05,0.1)))
      }
      
      colplot_list[[length(colplot_list)+1]] <- col_plot
      
    }
    
    return(colplot_list)
    
  }
  
  # function to calculate gene scores for given vector of genes
  cell_gene_scoring <- function(seurat_object, 
                                genes_of_interest, 
                                normalised = T){
    
    exprs_data <- 
      as.matrix(seurat_object[[seurat_object@active.assay]]@counts[
        genes_of_interest,]) #"SCT" in our case
    
    rowMaxes <- rowMax(exprs_data)
    exprs_data <- exprs_data/rowMaxes
    
    if(normalised){
      gene_scores <- colSums(exprs_data)/length(genes_of_interest)
    }
    else{
      gene_scores <- colSums(exprs_data)
    }
    
    return(gene_scores)
  }
  
  # get differential expression results in a list
  get_diff_exprs_list <- function(seurat_object, 
                                  cluster_column = "seurat_clusters"){
    
      clusters <- unique(as.character(seurat_object@meta.data[,cluster_column]))
      clusters <- clusters[order((clusters))]
      markers_list <- list()
      
      for(clust in clusters){
        print(paste0("Now processing cluster: ", as.character(clust),
                     " - number ", as.character(which(clusters %in% clust)),
                     " out of ", length(clusters)))
        
        markers_list[[length(markers_list)+1]] <- 
          FindMarkers(object = seurat_object, 
                      ident.1 = clust, 
                      group.by = cluster_column, 
                      logfc.threshold = 0, min.pct = 0,
                      test.use = "negbinom"
                      )
        names(markers_list)[length(markers_list)] <- clust
        
        markers_list[[length(markers_list)]] <- 
          markers_list[[length(markers_list)]][
            order(markers_list[[length(markers_list)]]$avg_log2FC, decreasing = T),]
      }
      
      markers_list <- markers_list[order(as.numeric(names(markers_list)))]
      
      return(markers_list)
  }
  
  # 2D differential expression plot (to combine GFP and IL2 volcano plots)
  get_2d_diff_exprs_plot <- function(FindMarkers_result_x, 
                                     FindMarkers_result_y, 
                                     label_x = "", 
                                     label_y = "", 
                                     main_title = "", 
                                     label_quad1 = "", 
                                     label_quad3 = "", 
                                     fold_change_thresh = 1, 
                                     adj_p_val_thresh = 0.001, 
                                     axisdh = 0.5, 
                                     show_corr = FALSE){
    require(ggplot2)
    require(ggExtra)
    
    #create the initial data structure for plotting
    plot_data <- merge(x = FindMarkers_result_x, y = FindMarkers_result_y, 
                       by = "row.names", all = T)
    rownames(plot_data) <- plot_data$Row.names
    
    #finding which genes to colour
    plot_data$coloured <- "not_coloured"
    plot_data$coloured[which(
      (plot_data$p_val_adj.x <= adj_p_val_thresh | 
        plot_data$p_val_adj.y <= adj_p_val_thresh) & 
        (abs(plot_data$avg_log2FC.x) >= fold_change_thresh | 
        abs(plot_data$avg_log2FC.y) >= fold_change_thresh))] <- "coloured"
    plot_data <- plot_data[order(plot_data$coloured, decreasing = T),]
    
    #get legend labels and point colours
    if(length(unique(plot_data$coloured)) < 2){
      labels <- "Other genes"
      values <- "darkgrey"
    }
    else{
      labels <- c(paste0("Log |FC|\u2265 ", 
                         fold_change_thresh, 
                         " and \nadj.p.value \u2264 ", 
                         adj_p_val_thresh, 
                         "\n(along either axis)"), 
                  "Other genes")
      values <- c("#fd7239", "darkgrey")
    }
    
    #calculating x-axis so that inf values can be replaced
    x.axis <- c(plot_data$avg_log2FC.x, plot_data$avg_log2FC.y)
    x.axis.max <- x.axis
    x.axis.max[x.axis.max == Inf | x.axis.max == -Inf] <- NA
    x.axis.max <- max(abs(x.axis.max), na.rm = T)
    x.axis.max.nodetect <- x.axis.max + (0.2*x.axis.max)
    
    xbreaks <- 
      sort(c(-fold_change_thresh, fold_change_thresh,
             seq(0, floor(-x.axis.max/0.2)*0.2, -axisdh),
             seq(0, ceiling(x.axis.max/0.2)*0.2, axisdh)))
             
    if(x.axis.max.nodetect < max(max(xbreaks), abs(min(xbreaks)))){
      x.axis.max.nodetect <- max(max(xbreaks), abs(min(xbreaks))) + 
        (0.5*max(max(xbreaks), abs(min(xbreaks))))
    }
    xbreaks <- sort(unique(xbreaks))
    
    x.axis <- plot_data$avg_log2FC.x
    x.axis[x.axis == Inf] <- x.axis.max.nodetect
    x.axis[x.axis == -Inf] <- -x.axis.max.nodetect
    x.axis[is.na(x.axis)] <- 0
    
    xlabels <- as.character(xbreaks)
    xlabels[xlabels == x.axis.max.nodetect | xlabels == -x.axis.max.nodetect] <- 
      sprintf("Detected\nonly in\none group")
    
    #mirroring x.axis as y.axis too
    y.axis <- plot_data$avg_log2FC.y
    y.axis[y.axis == Inf] <- x.axis.max.nodetect
    y.axis[y.axis == -Inf] <- -x.axis.max.nodetect
    y.axis[is.na(y.axis)] <- 0
    ybreaks <- xbreaks
    ylabels <- xlabels
    
    #updating plot_data
    plot_data$avg_log2FC.x <- x.axis
    plot_data$avg_log2FC.y <- y.axis
    
    #generating the ggplot object
    scatter_plot <- ggplot(data = plot_data,
                           aes(x = plot_data$avg_log2FC.x, 
                               y = plot_data$avg_log2FC.y, 
                               colour = plot_data$coloured, 
                               text = rownames(plot_data))) + 
      geom_segment(aes(x = -fold_change_thresh, y = fold_change_thresh, 
                       xend = fold_change_thresh, yend = fold_change_thresh), 
                       colour = "grey", size = 0.25, linetype = 2) + 
      geom_segment(aes(x = fold_change_thresh, y = fold_change_thresh, 
                       xend = fold_change_thresh, yend = -fold_change_thresh), 
                       colour = "grey", size = 0.25, linetype = 2) + 
      geom_segment(aes(x = fold_change_thresh, y = -fold_change_thresh, 
                       xend = -fold_change_thresh, yend = -fold_change_thresh), 
                       colour = "grey", size = 0.25, linetype = 2) + 
      geom_segment(aes(x = -fold_change_thresh, y = -fold_change_thresh, 
                       xend = -fold_change_thresh, yend = fold_change_thresh), 
                       colour = "grey", size = 0.25, linetype = 2) + 
      geom_hline(yintercept = 0, colour = "#4d4d4d", size = 0.25) + 
      geom_vline(xintercept = 0, colour = "#4d4d4d", size = 0.25) + 
      geom_abline(slope = 1, intercept = 0, alpha = 0.25)
      
    #adding a regression line and adding pearson correlation values
    if(show_corr){
      #geom_smooth not producing regression line, so calculating coef and adding
      # line manually
      lm_res <- lm(formula = y ~ x, 
                   data = data.frame(x = plot_data$avg_log2FC.x, 
                                     y = plot_data$avg_log2FC.y))
      cor_res <- cor(x = plot_data$avg_log2FC.x, 
                     y = plot_data$avg_log2FC.y, 
                     method = "pearson", use = "complete.obs")
      cor_res <- round(x = cor_res, digits = 3)
      
      scatter_plot <- scatter_plot + 
        geom_abline(slope = coef(lm_res)[['x']], 
                    intercept = coef(lm_res)[['(Intercept)']], 
                    linetype = "dashed", alpha = 0.65) + 
        annotate("text", 
                 x = 0+(max(xbreaks)*0.05), 
                 #y = max(ybreaks)+0.25,
                 y = max(ybreaks)+(0.1*max(ybreaks)), 
                 alpha = 0.5, label = paste0("r = ", cor_res),
                 hjust = 0, vjust = 1)#, 
                 #parse = parse_labels)
      
    }
    scatter_plot <- scatter_plot + 
      geom_point(shape = 19, size = 1.5) + 
      scale_color_manual(
        labels = labels,
        values = values) +
      labs(colour = "Genes", 
           y = paste0("Average log2 fold change", " ", label_y), 
           x = paste0("Average log2 fold change", " ", label_x), 
           title = main_title
           ) + 
      theme_light(base_size = 15) + 
      theme(plot.caption = element_text(hjust = 0, face= "italic"), 
            legend.key.width = unit(0.15,"line")
            )
    
    scatter_plot <- scatter_plot +
      scale_x_continuous(breaks = xbreaks, 
                         limits = c(min(xbreaks)+(0.1*min(xbreaks)), 
                                    max(xbreaks)+(0.1*max(xbreaks))), 
                         labels = xlabels) +
      scale_y_continuous(breaks = ybreaks, 
                         limits = c(min(ybreaks)+(0.1*min(ybreaks)), 
                                    max(ybreaks)+(0.1*max(ybreaks))), 
                         labels = ylabels) + 
      coord_fixed() + 
      theme(panel.grid.minor.x = element_blank(), 
            panel.grid.minor.y = element_blank(), 
            panel.grid.major = element_line(colour = "#4d4d4d"),
            axis.ticks = element_line(colour = "#4d4d4d")
            )
    
    #adding quadrant labels
    if(label_quad1 != "" & label_quad3 != ""){
      
      scatter_plot <- scatter_plot + 
        annotate("text", 
                 x = max(xbreaks)+(0.1*max(xbreaks)), 
                 y = max(ybreaks)+(0.1*max(ybreaks)), 
                 alpha = 0.35, label = paste0("Higher in\n", label_quad1), 
                 hjust = 1, vjust = 1) + 
        annotate("text", 
                 x = min(xbreaks)+(0.1*min(xbreaks)), 
                 y = min(ybreaks)+(0.1*min(ybreaks)), 
                 alpha = 0.35, label = paste0("Higher in\n", label_quad3), 
                 hjust = 0, vjust = 0)
    }
    
    scatter_plot <- scatter_plot + ggExtra::removeGrid()
    
    return(scatter_plot)
    
    
  }
  
}

```


#loading 04m and 24m data and performing QC with filtering and doublet removal
```{r}
#loading individual cellranger count results for manual normalisation
{
  #loading filtered_features results, which have been filtered by cellranger to 
  # remove background barcodes, into a list
  samples_list <- dir(path = paste0(loc_workspace, 
                                    "/01_cellranger_results/", 
                                    "all_individual_custom/"))
  samples_list <- samples_list[!grepl(pattern = "(1_5y|1y)", 
                                      x = samples_list, 
                                      ignore.case = T)]
  
  
  individual_count_data <- list()
  for(sample_name in samples_list){
    print(sample_name)
    individual_count_data[[length(individual_count_data)+1]] <- 
      Read10X(data.dir = paste0(loc_workspace, 
                                "/01_cellranger_results/", 
                                "all_individual_custom/", 
                                sample_name, 
                                "/filtered_feature_bc_matrix/"))
    colnames(individual_count_data[[length(individual_count_data)]]) <- 
      paste0(gsub(pattern = "^1_", 
                  replacement = "1.", 
                  x = sample_name, 
                  ignore.case = T), 
             "__",  
             gsub(pattern = "\\-[0-9]+$", 
                  replacement = "", 
                  x = colnames(
                    individual_count_data[[length(individual_count_data)]]),
                  ignore.case = T)
             )
    names(individual_count_data)[length(individual_count_data)] <- sample_name
  }
  
  #converting data to Seurat objects
  batch2 <- grep(pattern = "^(1_5y|2y)", x = samples_list, value = T)
  for(i in seq(1, length(individual_count_data))){
    
    individual_count_data[[i]] <- 
      CreateSeuratObject(counts = individual_count_data[[i]], 
                         project = names(individual_count_data)[i])
    
    individual_count_data[[i]]@meta.data$label <- 
      names(individual_count_data)[i]
    
    individual_count_data[[i]]@meta.data$age <- 
      gsub(pattern = "_[S0-9]+$", 
           replacement = "", 
           x = names(individual_count_data)[i], 
           ignore.case = T)
    
    individual_count_data[[i]]@meta.data$sample <- 
      gsub(pattern = "^.*_", 
           replacement = "", 
           x = names(individual_count_data)[i], 
           ignore.case = T)
    
    individual_count_data[[i]]@meta.data$batch <- 
      ifelse(names(individual_count_data)[i] %in% batch2, "2", "1")
  }
  
}

#performing QC/filtering on SequratObjects -- individual data
{
  #calculating mitochondrial gene percentage - High mt-genes %age indicates 
  # dead/dying cells so we can filter these cells out
  mt_gene_list <- list()
  for(i in seq(1, length(individual_count_data))){
    
    mt_gene_list[[i]] <- grep(pattern = "mt-", 
                     x = rownames(individual_count_data[[i]]), 
                     value = T, ignore.case = T)
    names(mt_gene_list)[i] <- names(individual_count_data)[i]
    individual_count_data[[i]] <- 
      PercentageFeatureSet(object = individual_count_data[[i]], 
                           pattern = "^mt-", 
                           col.name = "percent_mt_genes")
  }
  
  #filtering out cells with >5% mitochondrial gene exression based on 
  # https://academic.oup.com/bioinformatics/advance-article-abstract/doi/10.1093/bioinformatics/btaa751/5896986?redirectedFrom=fulltext
  # >5% was filtering out too many cells, so went with >10%
  for(i in seq(1, length(individual_count_data))){
    individual_count_data[[i]] <- subset(x = individual_count_data[[i]], 
                                         subset = percent_mt_genes<=10)
  }
  
  #adding metadata for later visualisation
  for(i in seq(1, length(individual_count_data))){
    temp <- individual_count_data[[i]]@meta.data
    temp$treatment <- "eGFP"
    if(temp$age != "1y" & temp$sample%in% c("S02", "S04", "S06")){
      temp$treatment <- "IL2"
    }
    else if(temp$age == "1y" & temp$sample %in% c("S01", "S03")){
      temp$treatment <- "IL2"
    }
    else{
      temp$treatment <- "eGFP"
    }
    
    individual_count_data[[i]]@meta.data$treatment <- temp$treatment
    
  }
  
}

#performing SCT with batch_var = "batch" for batch correction
{
  
  #merge Seurat objects using Seurat::merge
  merged_data <- merge(x = individual_count_data[[1]], 
                       y = individual_count_data[-1])
  
  #updating metadata for later visualisation
  {
    metadata <- merged_data@meta.data
    
    #cell count by cell type
    metadata$age <- gsub(pattern = "^1_5y",
                         replacement = "18m",
                         x = metadata$age,
                         ignore.case = T)
    metadata$label <- gsub(pattern = "^1_5y",
                         replacement = "18m",
                         x = metadata$label,
                         ignore.case = T)
    metadata$age <- gsub(pattern = "^1y",
                         replacement = "12m",
                         x = metadata$age,
                         ignore.case = T)
    metadata$label <- gsub(pattern = "^1y",
                         replacement = "12m",
                         x = metadata$label,
                         ignore.case = T)
    metadata$age <- gsub(pattern = "^2y",
                         replacement = "24m",
                         x = metadata$age,
                         ignore.case = T)
    metadata$label <- gsub(pattern = "^2y",
                         replacement = "24m",
                         x = metadata$label,
                         ignore.case = T)
    metadata$age <- gsub(pattern = "^4m",
                         replacement = "04m",
                         x = metadata$age,
                         ignore.case = T)
    metadata$label <- gsub(pattern = "^4m",
                         replacement = "04m",
                         x = metadata$label,
                         ignore.case = T)
    
    metadata$treatment[metadata$treatment == "eGFP"] <- "PHP.GFAP-GFP"
    metadata$treatment[metadata$treatment == "IL2"] <- "PHP.GFAP-IL2"
  
    metadata$age_treatment <- paste(metadata$age, metadata$treatment, sep = "_")
    
    merged_data@meta.data$age <- metadata$age
    merged_data@meta.data$treatment <- metadata$treatment
    merged_data@meta.data$age_treatment <- metadata$age_treatment
    merged_data@meta.data$label <- metadata$label
    rm(metadata)
    
  }
  
  #SCTransform normalisation with batch correction
  SCT_norm_data <- SCTransform(object = merged_data, 
                                        batch_var = "batch", 
                                        vars.to.regress = c("percent_mt_genes", 
                                                            "nCount_RNA"), 
                                        return.only.var.genes = F, 
                                        verbose = T)
  
}

#performing doublet analysis and removal
{
  
  data <- SCT_norm_data
  rm(SCT_norm_data)
  
  #running PCA
  data <- RunPCA(data, npcs = 100)
  
  #elbow plot
  ElbowPlot(object = data, ndims = 100)
  
  #running doublet analysis and removal
  metadata <- data@meta.data
  data_sce <- as.SingleCellExperiment(x = data)
  data_sce <- scDblFinder(sce = data_sce, 
                          clusters = NULL, 
                          samples = metadata$label, 
                          trajectoryMode = F, 
                          use.cxds = T, 
                          nfeatures = 3000, 
                          dims = 75, 
                          includePCs = 1:75, 
                          returnType = "full", 
                          nrounds = NULL, 
                          BPPARAM = MulticoreParam(6))
  
  metadata$doublet_status <- as.character(data_sce$scDblFinder.class)
  data@meta.data <- metadata
  
  #running umap to visualise doublets
  {
    #running UMAP
    data <- RunUMAP(object = data, 
                    dims = 1:75, 
                    reduction = "pca"
                    )
    
    #finding cell neighbours using t-SNE
    data <- FindNeighbors(object = data, 
                          reduction = "umap", 
                          dims = 1:2, 
                          nn.eps = 0)
    
    #finding clusters
    data <- FindClusters(object = data, 
                         resolution = 0.013, 
                         n.start = 50)
    
    DimPlot(object = data, 
            reduction = "umap", group.by = "doublet_status")
  }
  
  data <- subset(data, subset = doublet_status == "singlet")
  
}

#doublechecking QC 
# following (https://hbctraining.github.io/In-depth-NGS-Data-Analysis-Course/sessionIV/lessons/SC_quality_control_analysis.html)
{
  # Visualize the number UMIs/transcripts per cell
  ggplot(data = data@meta.data, aes(color=sample, x=nCount_RNA, fill= sample)) + 
        geom_density() + 
        scale_x_log10() + 
        ylab("log10 cell density") +
        geom_vline(xintercept = 500) + theme_classic()
  ggplot(data = data@meta.data, aes(color=sample, x=nCount_SCT, fill= sample)) + 
        geom_density() + 
        scale_x_log10() + 
        ylab("log10 cell density") +
        geom_vline(xintercept = 500) + theme_classic()
  
  # Visualize the distribution of genes detected per cell via histogram
  ggplot(data = data@meta.data, aes(color=sample, x=nFeature_RNA, fill= sample)) + 
        geom_density() + 
        scale_x_log10() + 
        ylab("log10 cell density") +
        geom_vline(xintercept = 500) + theme_classic()
  ggplot(data = data@meta.data, aes(color=sample, x=nFeature_SCT, fill= sample)) + 
        geom_density() + 
        scale_x_log10() + 
        ylab("log10 cell density") +
        geom_vline(xintercept = 500) + theme_classic()
  
  # Visualize the correlation between genes detected and number of UMIs and determine whether strong presence of cells with low numbers of genes/UMIs
  ggplot(data = data@meta.data, 
         aes(x=nCount_SCT, y=nFeature_SCT, color=label)) + 
    geom_point() + 
    stat_smooth(method=lm) +
    scale_x_log10() + 
    scale_y_log10() + 
    geom_vline(xintercept = 800) + theme_classic() 
  
  # Visualize the distribution of mitochondrial gene expression detected per cell
  ggplot(data = data@meta.data, aes(color=label, x=percent_mt_genes, fill=label)) + 
        geom_density() + 
        scale_x_log10() + 
        geom_vline(xintercept = 0.1) + theme_classic()
}

#saving the preprocessed and normalised seurat object
saveRDS(object = data,
        file = paste0(loc_seurat_results,
                      "04.04mv24m_SCT_noDbl.rds"), 
        compress = T)


```


#Batch effect normalised via SCT batch_var = "batch"
```{r}
#for the SCTransform batch_var = "batch" normalised data
{
  data <- readRDS(file = paste0(loc_seurat_results,
                                "04.04mv24m_SCT_noDbl.rds"))
  
  #running PCA
  data <- RunPCA(data, npcs = 100)
  
  #elbow plot
  ElbowPlot(object = data, ndims = 100) + geom_hline(yintercept = 2)
  
  #running UMAP
  data <- RunUMAP(object = data, 
                  dims = 1:75,  
                  reduction = "pca"
                  )
  
  #finding cell neighbours using t-SNE
  data <- FindNeighbors(object = data,  
                        reduction = "umap", 
                        dims = 1:2, 
                        nn.eps = 0)
  
  #finding clusters
  data <- FindClusters(object = data, 
                       resolution = 0.015, 
                       n.start = 50)
  
  DimPlot(object = data, 
          label = F, 
          # label = T, 
          reduction = "umap", 
          raster = T) + 
    NoLegend() + xlab("Dimension 1") + ylab("Dimension 2") + 
    coord_fixed() + theme(text = element_text(size = 20), 
                          axis.text = element_text(size = 16))
  
  saveRDS(object = data,
          file = paste0(loc_seurat_results,
                          "04.04mv24m_SCT_noDbl.rds"),
            compress = T)
  # data <- readRDS(file = paste0(loc_seurat_results, 
  #                               "04.04mv24m_SCT_noDbl.rds"))
  
  FeaturePlot(object = data,
              features = c("Tmem119", "P2ry12", "Cx3cr1", "Hexb",   #micro
                           "Mog", "Cldn11", "Mobp", "Mbp",          #oligo
                           "Aqp4", "Fgfr3", "Slc4a4", "Sox9",       #astro
                           "Pdgfra", "C1ql1", "Olig2", "Neu4"),     #OPC
              ncol = 4, reduction = "umap", slot = "data", 
              pt.size = 0.01) & coord_fixed()
  
  #differential expression of each cluster vs everything else
  all_markers <- FindAllMarkers(object = data, 
                                test.use = "negbinom")
  
  saveRDS(object = all_markers,
          file = paste0(loc_seurat_results,
                          "all_markers.rds"),
          compress = T)
  # all_markers <- readRDS(file = paste0(loc_seurat_results, "all_markers.rds"))
  
  FeaturePlot(object = data, 
              features = c("S100a4", "Camp", "S100a9", "Cd3d", "Cd3e", "Cd3g",  
                           "Mrc1", "Cd74", "Cd163", "Cd38", 
                           "Ctsz", "Cst7", "Ccl12", "Ccl3", 
                           "Ccl4", "Cxcl16", "Rbfox3", "Ptprc"), 
              ncol = 4, 
              reduction = "umap", 
              slot = "data")
  
  #minor markers
  #rearranged
  FeaturePlot(object = data, 
              features = c("Hbb-bs", "S100a8", "Cd74", "Ttr", "Stmn3", "Chgb", 
                           "C1qtnf4", "Ppp1r1b", "Fabp7", "Trh", "Sox11", 
                           "Bex2", "Ccnd2", "Tmem212", "Sspo", "Cldn5", "Dcn", 
                           "Vtn", "Acta2"), 
              ncol = 4, 
              reduction = "umap", 
              slot = "data", pt.size = 0.01) & coord_fixed()
  
  #creating the PCA plot using sample proportions across the clusters
  {
    #checking if non-major clusters create any changes to the PCA -- they don't
    #data <- subset(x = data, subset = seurat_clusters %in% c(0,1,13,19,2,3,4,6,9,11))
    
    data_plots <- get_cell_counts_props(seurat_object = data, 
                                        proportions = T, display_stats = T)
    patchwork::wrap_plots(data_plots, guides = "collect") & 
      theme(legend.position = "bottom")
    
    plot_data <- table(data@meta.data[,"label"], 
                       data@meta.data[,"seurat_clusters"])
    
    #reusing this obsolete part to send by sample proportions to Pierre
    for(i in seq(1, nrow(plot_data))){

      if(length(dim(plot_data))>2){
        for(j in seq(1, dim(plot_data)[3])){

          plot_data[i,,j] <- plot_data[i,,j]/(rowSums(plot_data[,,j])[i])
        }
      }
      else{
        plot_data[i,] <- plot_data[i,]/(rowSums(plot_data))[i]
      }
    }
    plot_data <- plot_data*100
    
    data_pca <- as.data.frame(prcomp(x = plot_data)$x)
    pca_vars <- summary(prcomp(x = plot_data))$importance[2,]
    
    data_pca$Age <- 
      unlist(strsplit(x = rownames(data_pca), split = "_"))[seq(1,24,2)]
    data_pca$Age <- as.factor(data_pca$Age)
    
    data_pca$Treatment <- "PHP.GFAP-GFP"
    data_pca$Treatment[seq(2,12,2)] <- "PHP.GFAP-IL2"
    data_pca$Treatment <- as.factor(data_pca$Treatment)
    
    data_pca$Sample <- 
      unlist(strsplit(x = rownames(data_pca), split = "_"))[seq(2,24,2)]
    
    ggplot(data = data_pca, 
           aes(x = PC1, y = PC2, colour = Treatment, shape = Age)) + 
      geom_point(size = 5) + 
      geom_text_repel(aes(label = Sample)) + 
      xlab(label = paste0("PC1 (", percent(pca_vars["PC1"]), 
                          " of total variance)")) + 
      ylab(label = paste0("PC2 (", percent(pca_vars["PC2"]), 
                          " of total variance)")) + 
      xlim(c(-8,8)) + ylim(c(-8,8)) + 
      coord_fixed(ratio = 1) + 
      theme_classic(base_size = 14) + 
      coord_fixed() + theme(text = element_text(size = 16), 
                            axis.text = element_text(size = 14)) 
    
    ggplot(data = data_pca, 
           aes(x = PC2, y = PC3, colour = Treatment, shape = Age)) + 
      geom_point(size = 5) + 
      geom_text_repel(aes(label = Sample)) + 
      xlab(label = paste0("PC2 (", percent(pca_vars["PC2"]), 
                          " of total variance)")) + 
      ylab(label = paste0("PC3 (", percent(pca_vars["PC3"]), 
                          " of total variance)")) + 
      xlim(c(-8,8)) + ylim(c(-8,8)) + 
      theme_classic(base_size = 14)
    
    ggplot(data = data_pca, 
           aes(x = PC1, y = PC3, colour = Treatment, shape = Age)) + 
      geom_point(size = 5) + 
      geom_text_repel(aes(label = Sample)) + 
      xlab(label = paste0("PC1 (", percent(pca_vars["PC1"]), 
                          " of total variance)")) + 
      ylab(label = paste0("PC3 (", percent(pca_vars["PC3"]), 
                          " of total variance)")) + 
      xlim(c(-8,8)) + ylim(c(-8,8)) + 
      theme_classic(base_size = 14)
    
  }
  
  
}


#subsetting micro data and looking at cell cluster membership
{
  micro_data <- subset(x = data, subset = seurat_clusters %in% c(0,22))
  
  #running SCT on a copy of micro_data to get VariableFeatures specific to this
  # cluster
  {
    VariableFeatures(micro_data) <- 
      VariableFeatures(SCTransform(object = micro_data, 
                                   batch_var = "batch", 
                                   vars.to.regress = c("percent_mt_genes", 
                                                       "nCount_RNA"), 
                                   return.only.var.genes = F, 
                                   verbose = T))
  }
  
  micro_data <- RunPCA(object = micro_data, 
                       features = VariableFeatures(object = micro_data), 
                       npcs = 100)
  
  #elbow plot
  ElbowPlot(object = micro_data, ndims = 100) + 
    geom_hline(yintercept = 2) + 
    geom_hline(yintercept = 3)
  
  #running UMAP
  micro_data <- RunUMAP(object = micro_data, 
                        dims = 1:40
                        )
  
  #finding cell neighbours using t-SNE
  micro_data <- FindNeighbors(object = micro_data, 
                                  reduction = "umap", 
                                  dims = 1:2, 
                                  nn.eps = 0)
  
  #finding clusters
  micro_data <- FindClusters(object = micro_data, 
                             resolution = 0.035, 
                             n.start = 50)
  
  DimPlot(object = micro_data, 
          label = F, 
          reduction = "umap", 
          raster = T) + 
    NoLegend() + xlab("Dimension 1") + ylab("Dimension 2") + 
    coord_fixed() + theme(text = element_text(size = 20), 
                          axis.text = element_text(size = 16))
  
  DimPlot(object = micro_data, 
          group.by = "treatment", 
          #label = TRUE, 
          reduction = "umap", 
          raster = T) + 
    xlab("Dimension 1") + ylab("Dimension 2") + 
    labs(colour = "Treatment", title = "") + 
    coord_fixed() + theme(text = element_text(size = 20), 
                          axis.text = element_text(size = 16))
  
  DimPlot(object = micro_data, 
          group.by = "age", 
          split.by = "treatment", 
          #label = TRUE, 
          reduction = "umap", 
          cols = c("#00BA38", "#F564E3"), 
          raster = T) + 
    xlab("Dimension 1") + ylab("Dimension 2") + 
    labs(color = "Age", title = "") + 
    coord_fixed() + theme(text = element_text(size = 20), 
                          axis.text = element_text(size = 16))
  
  #assigning new names to clusters for visualisations
  {
    micro_data$seurat_clusters_renamed <- 
      as.character(micro_data$seurat_clusters)
    micro_data$seurat_clusters_renamed[
      micro_data$seurat_clusters_renamed == "3"] <- "MICRO1"
    micro_data$seurat_clusters_renamed[
      micro_data$seurat_clusters_renamed == "0"] <- "MICRO2"
    micro_data$seurat_clusters_renamed[
      micro_data$seurat_clusters_renamed == "1"] <- "MICRO3"
    micro_data$seurat_clusters_renamed[
      micro_data$seurat_clusters_renamed == "4"] <- "MICRO4"
    micro_data$seurat_clusters_renamed[
      micro_data$seurat_clusters_renamed == "2"] <- "MICRO5"
    micro_data$seurat_clusters_renamed[
      micro_data$seurat_clusters_renamed == "5"] <- "MICRO6"
    micro_data$seurat_clusters_renamed <- 
      as.factor(micro_data$seurat_clusters_renamed)
    Idents(micro_data) <- micro_data$seurat_clusters_renamed
  }
  
  saveRDS(object = micro_data,
          file = paste0(loc_superclusters,
                      "micro/micro_data.rds"),
          compress = T)
  # micro_data <- readRDS(file = paste0(loc_superclusters,
  #                                     "micro/micro_data.rds"))
  
  
  micro_plots <- get_cell_counts_props(seurat_object = micro_data, 
                                       cluster_column = "seurat_clusters_renamed", 
                                       proportions = T, display_stats = T)
  
  wrap_plots(micro_plots, guides = "collect") & xlab(NULL) & 
    theme(legend.position = "bottom", 
          text = element_text(size = 20), 
          axis.text = element_text(size = 16)) 
  
  FeaturePlot(object = micro_data,
              features = c("Tyrobp", "Apoe", "B2m", "Trem2", "Lpl", "Cst7", 
                           "Axl", "Itgax", "Spp1", "Cd9", "Ccl6", "Csf1"),
              ncol = 3,
              slot = "data")
  FeaturePlot(object = micro_data,
              features = c("Cx3cr1", "P2ry12", "Tmem119", "Hexb", "Cst3", 
                           "Apoe", "Apoe", "Lpl", "Cst7", "Axl", "Itgax", "Spp1", 
                           "Ccl6", "Csf1"),
              ncol = 3,
              slot = "data") & coord_fixed()
  
}
#finding markers for reclustered microglia
{
  markers_list <- 
    get_diff_exprs_list(seurat_object = micro_data, 
                        cluster_column = "seurat_clusters_renamed")
  
  saveRDS(object = markers_list,
          file = paste0(loc_superclusters,
                        "micro/micro_markers_list.rds"),
          compress = T)
  markers_list <- readRDS(file = paste0(loc_superclusters,
                                        "micro/micro_markers_list.rds"))
  
  
  #exploring feature plots of clusters, one at a time
  {
    head(markers_list[[1]])
    FeaturePlot(object = micro_data,
                features = rownames(markers_list[[1]])[1:6], 
                ncol = 3, reduction = "umap", slot = "data", pt.size = 0.01)
  }
  
}
#DAM-scoring micro_data
{
  micro_data <- readRDS(file = paste0(loc_superclusters,
                                      "micro/micro_data.rds"))
  
  gene_scores <- cell_gene_scoring(seurat_object = micro_data, 
                                   genes_of_interest = c("Apoe", "Lpl", 
                                                         "Cst7", "Axl", 
                                                         "Itgax", "Spp1", 
                                                         "Ccl6", "Csf1"), 
                                   normalised = T)
  
  threshold <- 0.07  #selected
  #threshold <- 0.01   #testing
  micro_data$DAM_score <- gene_scores
  micro_data$DAM_cell <- gene_scores>threshold
  
  #histogram using base R
  hist(x = gene_scores, 
       breaks = 50, 
       xlab = "DAM Score", 
       main = "")
  abline(v = threshold, lty = 3)
  text(x = threshold+0.12, y = 5000, paste0("Threshold for DAM: ", threshold))
  
  #histogram using ggplot2
  ggplot(data = micro_data@meta.data, 
         aes(x = DAM_score)) + 
    geom_histogram(colour = "#595959", size = 1.5) + 
    geom_vline(xintercept = 0.07, linetype="dotted") + 
    scale_y_continuous(trans = scales::log_trans(10), 
                       labels = scales::trans_format("log10", scales::math_format(10^.x))) + 
    theme_light() + xlab("DAM score") + ylab("Counts") + 
    theme(text = element_text(size = 20), 
          axis.text = element_text(size = 16))
  
  colours <- scales::hue_pal()(2)
  
  DimPlot(object = micro_data,
          #label = TRUE,
          reduction = "umap", 
          group.by = "DAM_cell") + 
    scale_colour_manual(labels = c("Ordinary", "DAM"), 
                        values = colours)
  
  FeaturePlot(object = micro_data, 
              features = "DAM_score", 
              reduction = "umap", pt.size = 0.5)
  
  #proportion of age in the DAM cells -- pie chart
  {
    metadata <- micro_data@meta.data
    slices <- (table(subset(metadata, metadata$DAM_cell == TRUE)$age)/
      sum(table(subset(metadata, metadata$DAM_cell == TRUE)$age))) #* 100
    slices <- as.data.frame(slices)
    
    ggplot(data = slices) + 
      geom_bar(aes(x = "", y = Freq, fill = Var1), 
               width = 1, stat = "identity", 
               position = position_fill(reverse = T)
               ) + 
      coord_polar("y", start = 0) + 
      theme_void() + 
      theme(
        axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        panel.border = element_blank(),
        panel.grid=element_blank(),
        axis.ticks = element_blank(),
        plot.title=element_text(size=14, face="bold")
        ) + 
      geom_text(aes(x = 1, y = cumsum(Freq) - Freq/2,
                    label = percent(Freq)), size = 5) + 
      labs(title = paste0("Threshold Score: ", 
                          format(x = threshold, digits = 3),
                          "; DAM Pop: ", 
                          100 - as.numeric(gsub(pattern = "%", 
                                                replacement = "", 
                                                x = names(threshold), 
                                                ignore.case = T)), 
                          "%"
                          ), 
           fill = "Ages"
           )
  }
  
  #proportion of DAM cells in each age_treatment group -- column dot plot
  {
    metadata <- micro_data@meta.data
    plot_data <- table(metadata$sample, 
                       metadata$DAM_cell, 
                       metadata$age_treatment)
    
    for(i in seq(1, dim(plot_data)[3])){
      plot_data[,,i] <- (plot_data[,,i]/rowSums(plot_data[,,i]))*100
    }
    
    plot_data <- melt(plot_data)
    plot_data <- plot_data[complete.cases(plot_data),]
    plot_data <- plot_data[plot_data$Var2 == TRUE,]
    plot_data$sample <- plot_data$Var1
    plot_data$treatment <- gsub(pattern = "^.*m\\.?", 
                                replacement = "", 
                                x = plot_data$Var3, 
                                ignore.case = T)
    plot_data$age <- gsub(pattern = "\\..*$", replacement = "", 
                          x = plot_data$Var3, ignore.case = T)
    plot_data$Var1 <- paste0(plot_data$age, "_", plot_data$Var1)
    plot_data$age_treatment <- paste0(plot_data$age, "_", plot_data$treatment)
    
    col_plot <- 
      ggplot(data = plot_data, 
             aes(y = value, 
                 x = age)) + 
      geom_dotplot(aes(fill = treatment), 
                   binaxis = "y", 
                   position = "dodge", 
                   stackdir = "center", 
                   dotsize = 1.5) + 
      stat_summary(aes(fill = treatment), 
                   fun.data=mean_sdl, fun.args = list(mult=1),
                   geom="errorbar", color="black", width=0.2, 
                   position = position_dodge(0.9)) +
      stat_summary(aes(fill = treatment), 
                   fun.y=mean, geom="point", color="black", size = 0.8, 
                   position = position_dodge(0.9)) + 
      labs(fill = "Treatment",
           y = "Proportion of DAM cells",
           x = "Age group"
           ) + 
      expand_limits(y = 0) + 
      theme_light(base_size = 15) + 
      theme(panel.grid.minor = element_blank(), 
            legend.position = "bottom")
    
    #if(display_stats){
    stat.test.within.age <- plot_data %>%
      group_by(age) %>%
      t_test(formula = value ~ treatment) %>%
      adjust_pvalue(method = "bonferroni") %>%
      add_significance("p.adj")
    
    stat.test.within.age <- stat.test.within.age %>% 
      add_xy_position(x = "age", dodge = 0.8)
    
    for(i in stat.test.within.age$age){
      stat.test.within.age[stat.test.within.age$age == i,"y.position"] <- 
        max(plot_data[plot_data$age == i, "value"]) + 
        max(max(plot_data[plot_data$age == i, "value"])*0.1, 0.5)
    }
    
    col_plot <- col_plot + 
      stat_pvalue_manual(data = stat.test.within.age, 
                         label = "p.adj", tip.length = 0, size = 4.5) + 
      scale_y_continuous(expand = expansion(mult = c(0,0.1)))
    
    stat.test.bw.age <- plot_data %>%
      t_test(formula = value ~ age) %>%
      adjust_pvalue(method = "bonferroni")
    
    stat.test.bw.age <- stat.test.bw.age %>% 
      add_xy_position(x = "age")
    
    stat.test.bw.age$y.position <- 
        max(stat.test.within.age$y.position) + 
        max(max(stat.test.within.age$y.position)*0.1, 0.5)
    
    col_plot <- col_plot + 
      stat_pvalue_manual(stat.test.bw.age, label = "p.adj", 
                         tip.length = 0.02, step.increase = 0.5, size = 4.5) + 
      scale_y_continuous(expand = expansion(mult = c(0.05,0.1)))
    #}
    
    wrap_plots(list(col_plot, col_plot, col_plot), guides = "collect") & 
      theme(legend.position = "bottom", 
            text = element_text(size = 20), 
            axis.text = element_text(size = 16)) 
    
  }
  
}
#pca on proportions using samples
{
  micro_data <- readRDS(file = paste0(loc_superclusters,
                                      "micro/micro_data.rds"))
  
  exprs_data_per_smpl <- AverageExpression(object = micro_data, 
                                           assays = "SCT", 
                                           group.by = "label", verbose = T)
  exprs_data_per_smpl <- as.data.frame(t(exprs_data_per_smpl[[1]]))
  
  data_pca <- as.data.frame(prcomp(x = exprs_data_per_smpl)$x)
  pca_vars <- summary(prcomp(x = exprs_data_per_smpl))$importance[2,]
  
  data_pca$Age <- gsub(pattern = "_.*$", replacement = "",
                   x = rownames(data_pca), ignore.case = T)
  data_pca$Age <- as.factor(data_pca$Age)
  
  data_pca$Label <- rownames(data_pca)
  data_pca$Treatment <- "PHP.GFAP-GFP"
  data_pca$Treatment[gsub(pattern = "^.*_", 
                          replacement = "", 
                          x = rownames(data_pca), 
                          ignore.case = T) %in% 
                       c("S02", "S04", "S06")] <- "PHP.GFAP-IL2"
  data_pca$Treatment <- as.factor(data_pca$Treatment)
  
  data_pca$Sample <- gsub(pattern = "^.*_", replacement = "", 
                          x = data_pca$Label, ignore.case = T)
  
  data_pca$age_treatment <- paste0(data_pca$Age, "_", data_pca$Treatment)
  
  centroids <- 
    aggregate(cbind(data_pca$PC1, data_pca$PC2)~data_pca$age_treatment, 
              data_pca, mean)
  colnames(centroids) <- c("age_treatment", "mean.PC1", "mean.PC2")
  
  data_pca <- merge(data_pca, centroids, by = "age_treatment")
  
  ggplot(data = data_pca) + 
    geom_hline(yintercept = 0, linetype = "dotted") + 
    geom_vline(xintercept = 0, linetype = "dotted") + 
    geom_segment(aes(x = mean.PC1, y = mean.PC2, 
                     xend = PC1, yend = PC2, colour = Treatment), 
                 linetype = "dashed") + 
    geom_point(size = 5, 
               aes(x = PC1, y = PC2, colour = Treatment, shape = Age)) + 
    geom_point(aes(x = mean.PC1, y = mean.PC2), shape = 8) + 
    geom_text_repel(aes(x = PC1, y = PC2, 
                        label = Sample, colour = Treatment)) + 
    xlab(label = paste0("PC1 (", percent(pca_vars["PC1"]), 
                        " of total variance)")) + 
    ylab(label = paste0("PC2 (", percent(pca_vars["PC2"]), 
                        " of total variance)")) + 
    xlim(c(-(max(abs(data_pca$PC1), abs(data_pca$PC2))),
           max(abs(data_pca$PC1), abs(data_pca$PC2)))) + 
    ylim(c(-(max(abs(data_pca$PC1), abs(data_pca$PC2))),
           max(abs(data_pca$PC1), abs(data_pca$PC2)))) + 
    coord_fixed(ratio = 1) + 
    theme_classic(base_size = 14)
  
  # # Determine the proportion of variance of each component
  # # Proportion of variance equals (PC stdev^2) / (sum all PCs stdev^2)
  # pca_res <- prcomp(x = exprs_data_per_smpl)
  # propvariances <- ((pca_res$sdev ^ 2) / (sum(pca_res$sdev ^ 2))) * 100
  # 
  # # check rotations to see which genes are contributing to which PC
  # pca_res$rotation
  
}


#subsetting oligo data and looking at cell cluster membership
{
  oligo_data <- subset(x = data, subset = seurat_clusters %in% c(1,13,19))
  
  #running SCT on a copy of oligo_data to get VariableFeatures specific to this
  # cluster
  {
    VariableFeatures(oligo_data) <- 
      VariableFeatures(SCTransform(object = oligo_data, 
                                   batch_var = "batch", 
                                   vars.to.regress = c("percent_mt_genes", 
                                                       "nCount_RNA"), 
                                   return.only.var.genes = F, 
                                   verbose = T))
  }
  
  oligo_data <- RunPCA(object = oligo_data, 
                       features = VariableFeatures(object = oligo_data), 
                       npcs = 100)
  
  #elbow plot
  ElbowPlot(object = oligo_data, ndims = 100) + geom_hline(yintercept = 2) + geom_hline(yintercept = 3)
  
  #running UMAP
  oligo_data <- RunUMAP(object = oligo_data, 
                            dims = 1:50
                            )
  
  #finding cell neighbours using t-SNE
  oligo_data <- FindNeighbors(object = oligo_data, 
                                  reduction = "umap", 
                                  dims = 1:2, 
                                  nn.eps = 0)
  
  #finding clusters
  oligo_data <- FindClusters(object = oligo_data, 
                             resolution = 0.01, 
                             n.start = 50)
  
  DimPlot(object = oligo_data, 
          label = F, 
          reduction = "umap", 
          raster = T) + 
    NoLegend() + xlab("Dimension 1") + ylab("Dimension 2") + 
    coord_fixed() + theme(text = element_text(size = 20), 
                          axis.text = element_text(size = 16))
  
  DimPlot(object = oligo_data, 
          group.by = "treatment", 
          #label = TRUE, 
          reduction = "umap", 
          raster = T) + 
    xlab("Dimension 1") + ylab("Dimension 2") + 
    labs(title = "", colour = "Treatment") + 
    coord_fixed() + theme(text = element_text(size = 20), 
                          axis.text = element_text(size = 16))
  
  DimPlot(object = oligo_data, 
          group.by = "age", 
          split.by = "treatment", 
          #label = TRUE, 
          reduction = "umap", 
          cols = c("#00BA38", "#F564E3"), 
          raster = T) + 
    xlab("Dimension 1") + ylab("Dimension 2") + 
    labs(color = "Age", title = "") + 
    coord_fixed() + theme(text = element_text(size = 20), 
                          axis.text = element_text(size = 16))
  
  #assigning new names to clusters for visualisations
  {
    oligo_data$seurat_clusters_renamed <- 
      as.character(oligo_data$seurat_clusters)
    oligo_data$seurat_clusters_renamed[
      oligo_data$seurat_clusters_renamed == "5"] <- "OLIGO1"
    oligo_data$seurat_clusters_renamed[
      oligo_data$seurat_clusters_renamed == "2"] <- "OLIGO2"
    oligo_data$seurat_clusters_renamed[
      oligo_data$seurat_clusters_renamed == "1"] <- "OLIGO3"
    oligo_data$seurat_clusters_renamed[
      oligo_data$seurat_clusters_renamed == "0"] <- "OLIGO4"
    oligo_data$seurat_clusters_renamed[
      oligo_data$seurat_clusters_renamed == "3"] <- "OLIGO5"
    oligo_data$seurat_clusters_renamed[
      oligo_data$seurat_clusters_renamed == "4"] <- "OPCs"
    oligo_data$seurat_clusters_renamed <- 
      as.factor(oligo_data$seurat_clusters_renamed)
    Idents(oligo_data) <- oligo_data$seurat_clusters_renamed
  }
  
  saveRDS(object = oligo_data,
          file = paste0(loc_superclusters,
                      "oligo/oligo_data.rds"),
          compress = T)
  # oligo_data <- readRDS(file = paste0(loc_superclusters,
  #                                     "oligo/oligo_data.rds"))
  
  oligo_plots <- get_cell_counts_props(seurat_object = oligo_data, 
                                       cluster_column = "seurat_clusters_renamed", 
                                       proportions = T, display_stats = T)
  
  wrap_plots(c(oligo_plots[length(oligo_plots)], 
             oligo_plots[1:length(oligo_plots)-1]), 
             guides = "collect", ncol = 6) & 
    xlab(NULL) & 
    theme(legend.position = "bottom", 
          text = element_text(size = 20), 
          axis.text = element_text(size = 16))
  
  FeaturePlot(object = oligo_data, 
              features = c("S100b"), 
              slot = "data", 
              raster = T) + 
    xlab("Dimension 1") + ylab("Dimension 2") + 
    coord_fixed() + theme(text = element_text(size = 20), 
                          axis.text = element_text(size = 16))
  
  FeaturePlot(object = oligo_data, 
              features = c("Gpr17", "Fyn",       #Oligo1
                           "Tmod2", "Cpm",       #Oligo2
                           "S100b", "Selenop",   #Oligo3
                           "Ptgds", "Opalin",    #Oligo4
                           "Ifi27l2a", "H2-D1",  #Oligo5
                           "Pdgfra", "C1ql1"     #OPCs
                           ), 
              ncol = , 
              reduction = "umap", 
              slot = "data", 
              pt.size = 0.01) & coord_fixed()
  
}
#finding markers for reclustered oligodendrocytes 
{
  markers_list <- 
    get_diff_exprs_list(seurat_object = oligo_data, 
                        cluster_column = "seurat_clusters_renamed")
  
  saveRDS(object = markers_list,
          file = paste0(loc_superclusters,
                        "oligo/oligo_markers_list.rds"),
          compress = T)
  # markers_list <- readRDS(file = paste0(loc_superclusters,
  #                                       "oligo/oligo_markers_list.rds"))
  
  
  #exploring feature plots of clusters, one at a time
  {
    head(markers_list[[1]])
    FeaturePlot(object = oligo_data,
                features = rownames(markers_list[[1]])[1:6], 
                ncol = 3, reduction = "umap", slot = "data", pt.size = 0.01)
  }
  
}
#pca on proportions using samples
{
  oligo_data <- readRDS(file = paste0(loc_superclusters,
                                      "oligo/oligo_data.rds"))
  
  exprs_data_per_smpl <- AverageExpression(object = oligo_data, 
                                           assays = "SCT", 
                                           group.by = "label", verbose = T)
  exprs_data_per_smpl <- as.data.frame(t(exprs_data_per_smpl[[1]]))
  
  data_pca <- as.data.frame(prcomp(x = exprs_data_per_smpl)$x)
  pca_vars <- summary(prcomp(x = exprs_data_per_smpl))$importance[2,]
  
  data_pca$Age <- gsub(pattern = "_.*$", replacement = "",
                   x = rownames(data_pca), ignore.case = T)
  data_pca$Age <- as.factor(data_pca$Age)
  
  data_pca$Label <- rownames(data_pca)
  data_pca$Treatment <- "PHP.GFAP-GFP"
  data_pca$Treatment[gsub(pattern = "^.*_", 
                          replacement = "", 
                          x = rownames(data_pca), 
                          ignore.case = T) %in% 
                       c("S02", "S04", "S06")] <- "PHP.GFAP-IL2"
  data_pca$Treatment <- as.factor(data_pca$Treatment)
  
  data_pca$Sample <- gsub(pattern = "^.*_", replacement = "", 
                          x = data_pca$Label, ignore.case = T)
  
  data_pca$age_treatment <- paste0(data_pca$Age, "_", data_pca$Treatment)
  
  centroids <- 
    aggregate(cbind(data_pca$PC1, data_pca$PC2)~data_pca$age_treatment, 
              data_pca, mean)
  colnames(centroids) <- c("age_treatment", "mean.PC1", "mean.PC2")
  
  data_pca <- merge(data_pca, centroids, by = "age_treatment")
  
  ggplot(data = data_pca) + 
    geom_hline(yintercept = 0, linetype = "dotted") + 
    geom_vline(xintercept = 0, linetype = "dotted") + 
    geom_segment(aes(x = mean.PC1, y = mean.PC2, 
                     xend = PC1, yend = PC2, colour = Treatment), 
                 linetype = "dashed") + 
    geom_point(size = 5, 
               aes(x = PC1, y = PC2, colour = Treatment, shape = Age)) + 
    geom_point(aes(x = mean.PC1, y = mean.PC2), shape = 8) + 
    geom_text_repel(aes(x = PC1, y = PC2, 
                        label = Sample, colour = Treatment)) + 
    xlab(label = paste0("PC1 (", percent(pca_vars["PC1"]), 
                        " of total variance)")) + 
    ylab(label = paste0("PC2 (", percent(pca_vars["PC2"]), 
                        " of total variance)")) + 
    xlim(c(-(max(abs(data_pca$PC1), abs(data_pca$PC2))),
           max(abs(data_pca$PC1), abs(data_pca$PC2)))) + 
    ylim(c(-(max(abs(data_pca$PC1), abs(data_pca$PC2))),
           max(abs(data_pca$PC1), abs(data_pca$PC2)))) + 
    coord_fixed(ratio = 1) + 
    theme_classic(base_size = 14)
  
  # # Determine the proportion of variance of each component
  # # Proportion of variance equals (PC stdev^2) / (sum all PCs stdev^2)
  # pca_res <- prcomp(x = exprs_data_per_smpl)
  # propvariances <- ((pca_res$sdev ^ 2) / (sum(pca_res$sdev ^ 2))) * 100
  # 
  # # check rotations to see which genes are contributing to which PC
  # pca_res$rotation
  
}


#subsetting astro data and looking at cell cluster membership
{
  astro_data <- subset(x = data, subset = seurat_clusters %in% c(2,3,4,6,9,11))
  
  #running SCT on a copy of astro_data to get VariableFeatures specific to this
  # cluster
  {
    VariableFeatures(astro_data) <- 
      VariableFeatures(SCTransform(object = astro_data, 
                                   batch_var = "batch", 
                                   vars.to.regress = c("percent_mt_genes", 
                                                       "nCount_RNA"), 
                                   return.only.var.genes = F, 
                                   verbose = T))
  }
  
  astro_data <- RunPCA(object = astro_data, 
                       features = VariableFeatures(object = astro_data), 
                       npcs = 100)
  
  #elbow plot
  ElbowPlot(object = astro_data, ndims = 100) + geom_hline(yintercept = 2) + geom_hline(yintercept = 3)
  
  #running UMAP
  astro_data <- RunUMAP(object = astro_data, 
                        dims = 1:50
                        )
  
  #finding cell neighbours using t-SNE
  astro_data <- FindNeighbors(object = astro_data, 
                                  reduction = "umap", 
                                  dims = 1:2, 
                                  nn.eps = 0)
  
  #finding clusters
  astro_data <- FindClusters(object = astro_data,  
                                 resolution = 0.04, 
                                 n.start = 50)
  
  DimPlot(object = astro_data, 
          label = F, 
          reduction = "umap", 
          raster = T) + 
    NoLegend() + xlab("Dimension 1") + ylab("Dimension 2") + 
    coord_fixed() + theme(text = element_text(size = 20), 
                          axis.text = element_text(size = 16))
  
  DimPlot(object = astro_data, 
          group.by = "treatment", 
          #label = TRUE, 
          reduction = "umap", 
          raster = T) + 
    xlab("Dimension 1") + ylab("Dimension 2") + 
    labs(colour = "Treatment", title = "") + 
    coord_fixed() + theme(text = element_text(size = 20), 
                          axis.text = element_text(size = 16))
  
  DimPlot(object = astro_data, 
          group.by = "age", 
          split.by = "treatment", 
          #label = TRUE, 
          reduction = "umap", 
          cols = c("#00BA38", "#F564E3"), 
          raster = T) + 
    xlab("Dimension 1") + ylab("Dimension 2") + 
    labs(color = "Age", title = "") + 
    coord_fixed() + theme(text = element_text(size = 20), 
                          axis.text = element_text(size = 16))
  
  #assigning new names to clusters for visualisations
  {
    astro_data$seurat_clusters_renamed <- 
      as.character(astro_data$seurat_clusters)
    astro_data$seurat_clusters_renamed[
      astro_data$seurat_clusters_renamed == "0"] <- "\nNon-telencephalon\nAstrocytes"
    astro_data$seurat_clusters_renamed[
      astro_data$seurat_clusters_renamed == "1"] <- "\nTelencephalon\nAstrocytes"
    astro_data$seurat_clusters_renamed[
      astro_data$seurat_clusters_renamed == "2"] <- "\nBergmann Glia"
    astro_data$seurat_clusters_renamed[
      astro_data$seurat_clusters_renamed == "3"] <- "\nBergmann Glia"
    astro_data$seurat_clusters_renamed[
      astro_data$seurat_clusters_renamed == "4"] <- "\nCerebellar Astrocytes"
    astro_data$seurat_clusters_renamed[
      astro_data$seurat_clusters_renamed == "5"] <- "\nOlfactory Astrocytes"
    astro_data$seurat_clusters_renamed[
      astro_data$seurat_clusters_renamed == "6"] <- "\nStriatal Astrocytes"
    astro_data$seurat_clusters_renamed <- 
      as.factor(astro_data$seurat_clusters_renamed)
    #Idents(astro_data) <- astro_data$seurat_clusters_renamed
  }
  
  saveRDS(object = astro_data,
          file = paste0(loc_superclusters,
                      "astro/astro_data.rds"),
          compress = T)
  # astro_data <- readRDS(file = paste0(loc_superclusters,
  #                                     "astro/astro_data.rds"))
  
  astro_plots <- get_cell_counts_props(seurat_object = astro_data, 
                                       cluster_column = "seurat_clusters_renamed", 
                                       proportions = T, display_stats = T)
  
  wrap_plots(astro_plots, ncol = 6, guides = "collect") & xlab(NULL) & 
    theme(legend.position = "bottom", 
          text = element_text(size = 20), 
          axis.text = element_text(size = 16)) 
  
  #for supplementary figure
  FeaturePlot(object = astro_data,
              features = c("Septin4", "Gdf10",    #Bergmann Glia
                           "Aqp4", "Cyp26b1",    #Cerebellar
                           "Agt", "Slc6a11",      #Non-telencephalon
                           "Islr", "Gria2",      #Olfactory
                           "Crym",                #Striatal
                           "Cspg5", "Slco1c1"    #Telencephalon
                           ),
              ncol = 4,
              slot = "data", pt.size = 0.01) & coord_fixed()
  
}
#finding markers for reclustered astrocytes 
{
  markers_list <- 
    get_diff_exprs_list(seurat_object = astro_data, 
                        cluster_column = "seurat_clusters_renamed")
  
  saveRDS(object = markers_list,
          file = paste0(loc_superclusters,
                        "astro/astro_markers_list.rds"),
          compress = T)
  # markers_list <- readRDS(file = paste0(loc_superclusters,
  #                                       "astro/astro_markers_list.rds"))
  
  # Find markers for all clusters, set do.print=TRUE to output progress
  markers.all <- FindAllMarkers(object = astro_data, 
                                test.use = "negbinom")
  
  #exploring feature plots of clusters
  {
    head(markers_list[[1]])
    FeaturePlot(object = astro_data,
                features = rownames(markers_list[[1]])[1:6], 
                ncol = 3, reduction = "umap", slot = "data", pt.size = 0.01)
  }
  
}
#DAA-scoring astro_data
{
  astro_data <- readRDS(file = paste0(loc_superclusters,
                                      "astro/astro_data.rds"))
  
  #genes taken from 10.1038/s41593-020-0624-8 Extended Data Fig 4c and 4j
  gene_scores <- cell_gene_scoring(seurat_object = astro_data, 
                                   genes_of_interest = c("Ggta1", "Gsn", "Osmr", 
                                                         "Vim", "Serpina3n", 
                                                         "Ctsb", "Gfap", 
                                                         "Csmd1", "C4b") , 
                                   normalised = T)
  
  # threshold <- 0.09  #testing
  threshold <- 0.07   #DAM
  astro_data$DAA_score <- gene_scores
  astro_data$DAA_cell <- gene_scores>threshold
  
  hist(x = gene_scores, 
       breaks = 50, 
       xlab = "DAM Score", 
       main = "Histogram of DAA scores of Astrocytes")
  abline(v = threshold, lty = 3)
  text(x = threshold+0.12, y = 5000, paste0("Threshold for DAA: ", threshold))
  
  colours <- scales::hue_pal()(2)
  
  DimPlot(object = astro_data,
          #label = TRUE,
          reduction = "umap", 
          group.by = "DAA_cell") + 
    scale_colour_manual(labels = c("Ordinary", "DAA"), 
                        values = colours)
  
  FeaturePlot(object = astro_data, 
              features = "DAA_score", 
              reduction = "umap", pt.size = 0.5)
  
  #proportion of age in the DAA cells -- pie chart
  {
    metadata <- astro_data@meta.data
    slices <- (table(subset(metadata, metadata$DAA_cell == TRUE)$age)/
      sum(table(subset(metadata, metadata$DAA_cell == TRUE)$age))) #* 100
    slices <- as.data.frame(slices)
    
    ggplot(data = slices) + 
      geom_bar(aes(x = "", y = Freq, fill = Var1), 
               width = 1, stat = "identity", 
               position = position_fill(reverse = T)
               ) + 
      coord_polar("y", start = 0) + 
      theme_void() + 
      theme(
        axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        panel.border = element_blank(),
        panel.grid=element_blank(),
        axis.ticks = element_blank(),
        plot.title=element_text(size=14, face="bold")
        ) + 
      geom_text(aes(x = 1, y = cumsum(Freq) - Freq/2,
                    label = percent(Freq)), size = 5) + 
      labs(title = paste0("Threshold Score: ", 
                          format(x = threshold, digits = 3),
                          "; DAA Pop: ", 
                          100 - as.numeric(gsub(pattern = "%", 
                                                replacement = "", 
                                                x = names(threshold), 
                                                ignore.case = T)), 
                          "%"
                          ), 
           fill = "Ages"
           )
  }
  
  #proportion of DAA cells in each age_treatment group -- column dot plot
  {
    metadata <- astro_data@meta.data
    plot_data <- table(metadata$sample, 
                       metadata$DAA_cell, 
                       metadata$age_treatment)
    
    for(i in seq(1, dim(plot_data)[3])){
      plot_data[,,i] <- (plot_data[,,i]/rowSums(plot_data[,,i]))*100
    }
    
    plot_data <- melt(plot_data)
    plot_data <- plot_data[complete.cases(plot_data),]
    plot_data <- plot_data[plot_data$Var2 == TRUE,]
    plot_data$sample <- plot_data$Var1
    plot_data$treatment <- gsub(pattern = "^.*m\\.?", 
                                replacement = "", 
                                x = plot_data$Var3, 
                                ignore.case = T)
    plot_data$age <- gsub(pattern = "\\..*$", replacement = "", 
                          x = plot_data$Var3, ignore.case = T)
    plot_data$Var1 <- paste0(plot_data$age, "_", plot_data$Var1)
    plot_data$age_treatment <- paste0(plot_data$age, "_", plot_data$treatment)
    
    col_plot <- 
      ggplot(data = plot_data, 
             aes(y = value, 
                 x = age)) + 
      geom_dotplot(aes(fill = treatment), 
                   binaxis = "y", 
                   position = "dodge", 
                   stackdir = "center", 
                   dotsize = 1.5) + 
      stat_summary(aes(fill = treatment), 
                   fun.data=mean_sdl, fun.args = list(mult=1),
                   geom="errorbar", color="black", width=0.2, 
                   position = position_dodge(0.9)) +
      stat_summary(aes(fill = treatment), 
                   fun.y=mean, geom="point", color="black", size = 0.8, 
                   position = position_dodge(0.9)) + 
      labs(fill = "Treatment",
           y = "Proportion of DAA cells",
           x = "Age group"
           ) + 
      expand_limits(y = 0) + 
      theme_light(base_size = 15) + 
      theme(panel.grid.minor = element_blank(), 
            legend.position = "bottom")
    
    #if(display_stats){
    stat.test.within.age <- plot_data %>%
      group_by(age) %>%
      t_test(formula = value ~ treatment) %>%
      adjust_pvalue(method = "bonferroni") %>%
      add_significance("p.adj")
    
    stat.test.within.age <- stat.test.within.age %>% 
      add_xy_position(x = "age", dodge = 0.8)
    
    for(i in stat.test.within.age$age){
      stat.test.within.age[stat.test.within.age$age == i,"y.position"] <- 
        max(plot_data[plot_data$age == i, "value"]) + 
        max(max(plot_data[plot_data$age == i, "value"])*0.1, 0.5)
    }
    
    col_plot <- col_plot + 
      stat_pvalue_manual(data = stat.test.within.age, 
                         label = "p.adj", tip.length = 0) + 
      scale_y_continuous(expand = expansion(mult = c(0,0.1)))
    
    stat.test.bw.age <- plot_data %>%
      t_test(formula = value ~ age) %>%
      adjust_pvalue(method = "bonferroni")
    
    stat.test.bw.age <- stat.test.bw.age %>% 
      add_xy_position(x = "age")
    
    stat.test.bw.age$y.position <- 
        max(stat.test.within.age$y.position) + 
        max(max(stat.test.within.age$y.position)*0.1, 0.5)
    
    col_plot <- col_plot + 
      stat_pvalue_manual(stat.test.bw.age, label = "p.adj", 
                         tip.length = 0.02, step.increase = 0.5) + 
      scale_y_continuous(expand = expansion(mult = c(0.05,0.1)))
    #}
    
    wrap_plots(list(col_plot, col_plot, col_plot), guides = "collect") & 
      theme(legend.position = "bottom")
    
  }
  
}
#pca on proportions using samples
{
  astro_data <- readRDS(file = paste0(loc_superclusters,
                                      "astro/astro_data.rds"))
  
  exprs_data_per_smpl <- AverageExpression(object = astro_data, 
                                           assays = "SCT", 
                                           group.by = "label", verbose = T)
  exprs_data_per_smpl <- as.data.frame(t(exprs_data_per_smpl[[1]]))
  
  data_pca <- as.data.frame(prcomp(x = exprs_data_per_smpl)$x)
  pca_vars <- summary(prcomp(x = exprs_data_per_smpl))$importance[2,]
  
  data_pca$Age <- gsub(pattern = "_.*$", replacement = "",
                   x = rownames(data_pca), ignore.case = T)
  data_pca$Age <- as.factor(data_pca$Age)
  
  data_pca$Label <- rownames(data_pca)
  data_pca$Treatment <- "PHP.GFAP-GFP"
  data_pca$Treatment[gsub(pattern = "^.*_", 
                          replacement = "", 
                          x = rownames(data_pca), 
                          ignore.case = T) %in% 
                       c("S02", "S04", "S06")] <- "PHP.GFAP-IL2"
  data_pca$Treatment <- as.factor(data_pca$Treatment)
  
  data_pca$Sample <- gsub(pattern = "^.*_", replacement = "", 
                          x = data_pca$Label, ignore.case = T)
  
  data_pca$age_treatment <- paste0(data_pca$Age, "_", data_pca$Treatment)
  
  centroids <- 
    aggregate(cbind(data_pca$PC1, data_pca$PC2)~data_pca$age_treatment, 
              data_pca, mean)
  colnames(centroids) <- c("age_treatment", "mean.PC1", "mean.PC2")
  
  data_pca <- merge(data_pca, centroids, by = "age_treatment")
  
  ggplot(data = data_pca) + 
    geom_hline(yintercept = 0, linetype = "dotted") + 
    geom_vline(xintercept = 0, linetype = "dotted") + 
    geom_segment(aes(x = mean.PC1, y = mean.PC2, 
                     xend = PC1, yend = PC2, colour = Treatment), 
                 linetype = "dashed") + 
    geom_point(size = 5, 
               aes(x = PC1, y = PC2, colour = Treatment, shape = Age)) + 
    geom_point(aes(x = mean.PC1, y = mean.PC2), shape = 8) + 
    geom_text_repel(aes(x = PC1, y = PC2, 
                        label = Sample, colour = Treatment)) + 
    xlab(label = paste0("PC1 (", percent(pca_vars["PC1"]), 
                        " of total variance)")) + 
    ylab(label = paste0("PC2 (", percent(pca_vars["PC2"]), 
                        " of total variance)")) + 
    xlim(c(-(max(abs(data_pca$PC1), abs(data_pca$PC2))),
           max(abs(data_pca$PC1), abs(data_pca$PC2)))) + 
    ylim(c(-(max(abs(data_pca$PC1), abs(data_pca$PC2))),
           max(abs(data_pca$PC1), abs(data_pca$PC2)))) + 
    coord_fixed(ratio = 1) + 
    theme_classic(base_size = 14)
  
  # # Determine the proportion of variance of each component
  # # Proportion of variance equals (PC stdev^2) / (sum all PCs stdev^2)
  # pca_res <- prcomp(x = exprs_data_per_smpl)
  # propvariances <- ((pca_res$sdev ^ 2) / (sum(pca_res$sdev ^ 2))) * 100
  # 
  # # check rotations to see which genes are contributing to which PC
  # pca_res$rotation
  
}



```


#importing mousebrain.org data
```{r}
#importing via loom and constructing the mousebrain seurat object for astrocytes
{
  # library(loomR)
  mousebrain <- loomR::connect(filename = paste0(loc_ext_data, 
                                                 "mousebrain/l5_all.loom"), 
                               mode = "r+", 
                               skip.validate = F)
  
  #using Class == "Astrocytes" to also get radial glia-like cells which 
  #table(mousebrain$col.attrs$Class[], mousebrain$col.attrs$TaxonomyRank4[])
  unique(mousebrain$col.attrs$Class[])
  astro_indices <- which(mousebrain$col.attrs$Class[] == "Astrocytes")
  
  #reconstructing the metadata
  {
    mousebrain_meta <- 
      as.data.frame(
      cbind(mousebrain$col.attrs$CellID[astro_indices], 
            mousebrain$col.attrs$Age[astro_indices], 
            mousebrain$col.attrs$Class[astro_indices], 
            mousebrain$col.attrs$Subclass[astro_indices], 
            mousebrain$col.attrs$ClassProbability_Astrocyte[astro_indices], 
            mousebrain$col.attrs$`ClassProbability_Bergmann-glia`[astro_indices], 
            mousebrain$col.attrs$ClusterName[astro_indices], 
            mousebrain$col.attrs$Clusters[astro_indices], 
            mousebrain$col.attrs$Description[astro_indices], 
            mousebrain$col.attrs$Developmental_compartment[astro_indices], 
            mousebrain$col.attrs$Location_based_on[astro_indices], 
            mousebrain$col.attrs$Probable_location[astro_indices], 
            mousebrain$col.attrs$Region[astro_indices], 
            mousebrain$col.attrs$SampleID[astro_indices], 
            mousebrain$col.attrs$Sex[astro_indices], 
            mousebrain$col.attrs$Taxonomy_group[astro_indices], 
            mousebrain$col.attrs$Tissue[astro_indices], 
            mousebrain$col.attrs$`_X`[astro_indices], 
            mousebrain$col.attrs$`_Y`[astro_indices]
            )
      )
    colnames(mousebrain_meta) <- c("CellID", "Age", "Class", "Subclass", 
                                   "ClassProbability_Astrocyte", 
                                   "ClassProbability_Bergmann-glia", 
                                   "ClusterName", "Clusters", "Description", 
                                   "Developmental_compartment", 
                                   "Location_based_on", "Probable_location", 
                                   "Region", "SampleID", "Sex", "Taxonomy_group", 
                                   "Tissue", "_X", "_Y")
    
    mousebrain_meta$ClassProbability_Astrocyte <- 
      as.numeric(mousebrain_meta$ClassProbability_Astrocyte)
    mousebrain_meta$`ClassProbability_Bergmann-glia` <- 
      as.numeric(mousebrain_meta$`ClassProbability_Bergmann-glia`)
    mousebrain_meta$Clusters <- as.numeric(mousebrain_meta$Clusters)
    mousebrain_meta$`_X` <- as.numeric(mousebrain_meta$`_X`)
    mousebrain_meta$`_Y` <- as.numeric(mousebrain_meta$`_Y`)
    
    #checking duplicates to see if they are simple clones or mislabelled unique cells
    #unique(mousebrain_meta$CellID[duplicated(mousebrain_meta$CellID)])
    # [1] "10X51_2_AGGATGTGTGCA-" "10X49_4_ATGCACTCTAGG-" "10X49_5_TTTGCTGGAAAT-" "10X49_5_GAAACTGTGCTA-"
    # [5] "10X49_5_TACTCTTCTACT-" "10X52_2_CTTACTGTCGTA-" "10X52_1_CCGTTGCTGAAC-"
    temp <- mousebrain_meta[mousebrain_meta$CellID == "10X51_2_AGGATGTGTGCA-",]
    temp2 <- mousebrain_meta[mousebrain_meta$CellID == "10X49_4_ATGCACTCTAGG-",]
    temp3 <- mousebrain_meta[mousebrain_meta$CellID == "10X49_5_TTTGCTGGAAAT-",]
    temp4 <- mousebrain_meta[mousebrain_meta$CellID == "10X49_5_GAAACTGTGCTA-",]
    temp5 <- mousebrain_meta[mousebrain_meta$CellID == "10X49_5_TACTCTTCTACT-",]
    temp6 <- mousebrain_meta[mousebrain_meta$CellID == "10X52_2_CTTACTGTCGTA-",]
    temp7 <- mousebrain_meta[mousebrain_meta$CellID == "10X52_1_CCGTTGCTGAAC-",]
    #all appear to be mislabelled 
    
    mousebrain_meta$CellID[duplicated(mousebrain_meta$CellID)] <- 
      paste0(mousebrain_meta$CellID[duplicated(mousebrain_meta$CellID)], "d")
    
    rownames(mousebrain_meta) <- mousebrain_meta$CellID
    
  }
  
  #reconstructing the counts matrix
  {
    mousebrain_counts <- mousebrain[["matrix"]][astro_indices,]
    rownames(mousebrain_counts) <- mousebrain_meta$CellID
    colnames(mousebrain_counts) <- mousebrain$row.attrs$Gene[]
    
    #transposing so that genes are rows and cells as columns for Seurat
    mousebrain_counts <- t(mousebrain_counts)
  }
  
  #creating the seurat object for integration
  mousebrain_seurat <- CreateSeuratObject(counts = mousebrain_counts, 
                                          project = "mousebrain.org", 
                                          meta.data = mousebrain_meta)
  ##-- >> means added or modified to follow the seurat SCT integration guide
  mousebrain_seurat <- SCTransform(object = mousebrain_seurat, 
                                   return.only.var.genes = F)
  
  saveRDS(object = mousebrain_seurat,
          file = paste0(loc_other_datasets,
                        "mousebrain/mousebrain_seurat_astro.rds"),
          compress = T)
  
  # mousebrain_seurat <-
  #   readRDS(file = paste0(loc_other_datasets,
  #                       "mousebrain/mousebrain_seurat_astro.rds"))
  
}
#importing the astrocyte seurat object and running integration
{
  astro_data <- readRDS(file = paste0(loc_superclusters,
                                      "astro/astro_data.rds"))
  ##--
  # mousebrain_seurat <- FindVariableFeatures(object = mousebrain_seurat, 
  #                                           selection.method = "vst")
  ##--
  integration_feats <- 
    SelectIntegrationFeatures(object.list = list(astro_data, mousebrain_seurat), 
                              nfeatures = 3000)
  ##--
  integrated_data <- 
    PrepSCTIntegration(object.list = list(astro_data, mousebrain_seurat), 
                       anchor.features = integration_feats, verbose = T)
  
  ##--
  integrated_data <- 
    FindIntegrationAnchors(object.list = integrated_data, #object.list = list(astro_data, mousebrain_seurat),
                           normalization.method = "SCT", 
                           anchor.features = integration_feats, verbose = T)
  
  integrated_data <- IntegrateData(anchorset = integrated_data, 
                                   normalization.method = "SCT", verbose = T)
  
  integrated_data@meta.data$astro_clusters <- 
    integrated_data@meta.data$seurat_clusters
  integrated_data@meta.data$source_dataset <- integrated_data@meta.data$sample
  integrated_data@meta.data$source_dataset[
    !is.na(integrated_data@meta.data$sample)] <- "Source data"
  integrated_data@meta.data$source_dataset[
    is.na(integrated_data@meta.data$sample)] <- "MouseBrain.org"
  ##--
  #integrated_data <- ScaleData(integrated_data)
  
  integrated_data <- RunPCA(integrated_data, npcs = 100)
  
  ElbowPlot(object = integrated_data, ndims = 100)
  
  integrated_data <- RunUMAP(integrated_data, reduction = "pca", dims = 1:70)
  integrated_data <- FindNeighbors(integrated_data, 
                                   reduction = "umap", 
                                   dims = 1:2, 
                                   nn.eps = 0)
  integrated_data <- FindClusters(integrated_data, resolution = 0.03)
  
  saveRDS(object = integrated_data,
          file = paste0(loc_other_datasets,
                        "mousebrain/integrated_data_astro.rds"),
          compress = T)
  # integrated_data <- readRDS(file = paste0(loc_other_datasets,
  #                                          "mousebrain/integrated_data_astro.rds"))
  
  integrated_data@meta.data$Description <- 
    str_wrap(string = integrated_data@meta.data$Description, width = 25)
  
  p1 <- DimPlot(integrated_data, reduction = "umap", 
                group.by = "source_dataset") + 
    labs(title = "") & coord_fixed()
  p1$data$source_dataset <- factor(x = p1$data$source_dataset, 
                                   levels = c("Source data", "MouseBrain.org"))
  p2 <- DimPlot(integrated_data, reduction = "umap", label = T)  & coord_fixed()
  p3 <- DimPlot(integrated_data, reduction = "umap", 
                #group.by = "astro_clusters", label = T) + 
                group.by = "seurat_clusters_renamed", label = F) + 
    labs(title = "Source data") & coord_fixed() 
  p4 <- DimPlot(integrated_data, reduction = "umap", 
                group.by = "Description", label = F) + 
    labs(title = "MouseBrain.org") & coord_fixed() 
  
  (p3|p4)
  
}


```


#monocle2 analysis
```{r}
#running monocle2 analysis for micro
{
  #loading microglia data
  micro_data <- readRDS(file = paste0(loc_superclusters,
                                      "micro/micro_data.rds"))
  
  #running monocle2
  micro_mono2 <- as(object = as.matrix(micro_data[["SCT"]]@data), 
                    Class = "sparseMatrix")
  pheno_data <- data.frame(micro_data@meta.data)
  pheno_data <- new(Class = "AnnotatedDataFrame", pheno_data)
  feat_data <- data.frame(gene_short_name = rownames(micro_mono2), 
                          row.names = rownames(micro_mono2))
  feat_data <- new(Class = "AnnotatedDataFrame", feat_data)
  
  micro_mono2 <- 
    monocle::newCellDataSet(micro_mono2, 
                            phenoData = pheno_data, 
                            featureData = feat_data, 
                            lowerDetectionLimit = 0.5, 
                            expressionFamily = VGAM::negbinomial.size())
  
  micro_mono2 <- DESeq2::estimateSizeFactors(micro_mono2)
  
  micro_mono2 <- DESeq2::estimateDispersions(micro_mono2)
  disp_table <- monocle:::dispersionTable(micro_mono2)
  #default mean_expression >= 0.5
  # 0.2 also worked with disp >= 1 but results in out of memory even at 500GB
  ordering_genes <- subset(disp_table, 
                           # mean_expression >= 0.01 & 
                           #   dispersion_empirical >= 0.08 * 
                           mean_expression >= 0.01 & 
                             dispersion_empirical >= 0.2 * 
                             dispersion_fit)$gene_id
  print(paste0("Number of ordering genes: ", length(ordering_genes)))
  #ordering_genes <- c("Apoe", "Lpl", "Cst7", "Axl", "Itgax", "Spp1", "Ccl6", "Csf1")
  
  micro_mono2 <- monocle::setOrderingFilter(micro_mono2, 
  						ordering_genes)
  
  #default auto_param_selection = T
  micro_mono2 <- monocle::reduceDimension(cds = micro_mono2, 
                                          norm_method = "none", 
                                          pseudo_expr = 0, 
                                          reduction_method = "DDRTree", 
                                          relative_exprs = F, 
                                          scaling = F, 
                                          max_components = 20)
  
  #old: micro_reclustered_monocle2.rds ; new: micro_data_monocle2.rds
  # saveRDS(object = micro_mono2, 
  # 	file = "/bi/group/liston/samar/AAV_IL2/02_seurat_results/micro_data_monocle2.rds", 
  # 	compress = T)
  
  micro_mono2 <- monocle::orderCells(cds = micro_mono2)
  
  saveRDS(object = micro_mono2,
          file = paste0(loc_monocle, 
                        "micro/micro_data_monocle2.rds"), 
  	compress = T)
  
}
#visualisations
{
  
  micro_data_monocle2 <- readRDS(file = paste0(loc_monocle,
                                               "micro/micro_data_monocle2.rds"))
  
  monocle::plot_complex_cell_trajectory(cds = micro_data_monocle2, 
                                        color_by = "Pseudotime", 
                                        cell_size = 0.5) + 
    facet_wrap(~age+treatment, ncol = 2) + NoLegend()
  
  monocle::plot_complex_cell_trajectory(cds = micro_data_monocle2, 
                                        color_by = "State", 
                                        cell_size = 0.5) + 
    facet_wrap(~State, ncol = 2) + NoLegend()
  
  #get proportions of each state
  {
    proportions <- table(micro_data_monocle2$age_treatment, 
                         micro_data_monocle2$State)
    proportions <- (proportions/rowSums(proportions))*100
    proportions <- as.data.frame.matrix(proportions)
  }
  #appears to already be sorted by age (highest young in State 1, 
  # highest old in State 7, which is the largest leaf node)
  
  monocle::plot_cell_trajectory(cds = micro_data_monocle2, 
                                        color_by = "age", 
                                        cell_size = 0.5) 
  
  #pca on proportions
  {
    proportions <- table(micro_data_monocle2$age_treatment,
                         micro_data_monocle2$State)
    proportions <- (proportions/rowSums(proportions))*100
    proportions <- as.data.frame.matrix(proportions)
    
    plot_data <- proportions
    
    data_pca <- as.data.frame(prcomp(x = plot_data)$x)
    pca_vars <- summary(prcomp(x = plot_data))$importance[2,]
    
    data_pca$Age <- gsub(pattern = "\\..*$", replacement = "",
                         x = rownames(data_pca), ignore.case = T)
    data_pca$Age <- as.factor(data_pca$Age)
    
    data_pca$Treatment <- gsub(pattern = "^.*m\\.?", replacement = "",
                               x = rownames(data_pca), ignore.case = T)
    data_pca$Treatment <- as.factor(data_pca$Treatment)
    
    data_pca$Sample <- rownames(data_pca)
    
    ggplot(data = data_pca, 
           aes(x = PC1, y = PC2, colour = Treatment, shape = Age)) + 
      geom_point(size = 5) + 
      xlab(label = paste0("PC1 (", percent(pca_vars["PC1"]), 
                          " of total variance)")) + 
      ylab(label = paste0("PC2 (", percent(pca_vars["PC2"]), 
                          " of total variance)")) + 
      xlim(c(-(max(abs(data_pca$PC1), abs(data_pca$PC2))),
             max(abs(data_pca$PC1), abs(data_pca$PC2)))) + 
      ylim(c(-(max(abs(data_pca$PC1), abs(data_pca$PC2))),
             max(abs(data_pca$PC1), abs(data_pca$PC2)))) + 
      coord_fixed(ratio = 1) + 
      theme_classic(base_size = 14) 
    
    # Determine the proportion of variance of each component
    # Proportion of variance equals (PC stdev^2) / (sum all PCs stdev^2)
    pca_res <- prcomp(x = plot_data)
    propvariances <- ((pca_res$sdev ^ 2) / (sum(pca_res$sdev ^ 2))) * 100
    
    # check rotations to see which monocle2 states are contributing to which PC
    pca_res$rotation
    
  }
  
  #pca on proportions using samples
  {
    proportions <- table(micro_data_monocle2$label, 
                         micro_data_monocle2$State)
    proportions <- (proportions/rowSums(proportions))*100
    proportions <- as.data.frame.matrix(proportions)
    
    plot_data <- proportions
    
    data_pca <- as.data.frame(prcomp(x = plot_data, rank. = 4)$x)
    pca_vars <- summary(prcomp(x = plot_data))$importance[2,]
    
    data_pca$Age <- gsub(pattern = "_.*$", replacement = "",
                     x = rownames(data_pca), ignore.case = T)
    data_pca$Age <- as.factor(data_pca$Age)
    
    data_pca$Label <- rownames(data_pca)
    data_pca <- unique(merge(data_pca, 
                             micro_data@meta.data[,c("treatment", "label")], 
                             by.x = "Label", by.y = "label", 
                             all.x = T, all.y = F))
    rownames(data_pca) <- data_pca$Label
    data_pca$Treatment <- data_pca$treatment
    data_pca$treatment <- NULL
    data_pca$Treatment <- as.factor(data_pca$Treatment)
    
    data_pca$Sample <- gsub(pattern = "^.*_", replacement = "", 
                            x = data_pca$Label, ignore.case = T)
    
    data_pca$age_treatment <- paste0(data_pca$Age, "_", data_pca$Treatment)
    
    centroids <- 
      aggregate(cbind(data_pca$PC1, data_pca$PC2)~data_pca$age_treatment, 
                data_pca, mean)
    colnames(centroids) <- c("age_treatment", "mean.PC1", "mean.PC2")
    
    data_pca <- merge(data_pca, centroids, by = "age_treatment")
    
    ggplot(data = data_pca) + 
      geom_segment(aes(x = mean.PC1, y = mean.PC2, 
                       xend = PC1, yend = PC2, colour = Treatment), 
                   linetype = "dashed") + 
      geom_point(size = 5, 
                 aes(x = PC1, y = PC2, colour = Treatment, shape = Age)) + 
      geom_point(aes(x = mean.PC1, y = mean.PC2), shape = 8) + 
      geom_text_repel(aes(x = PC1, y = PC2, 
                          label = Sample, colour = Treatment)) + 
      xlab(label = paste0("PC1 (", percent(pca_vars["PC1"]), 
                          " of total variance)")) + 
      ylab(label = paste0("PC2 (", percent(pca_vars["PC2"]), 
                          " of total variance)")) + 
      xlim(c(-(max(abs(data_pca$PC1), abs(data_pca$PC2))),
             max(abs(data_pca$PC1), abs(data_pca$PC2)))) + 
      ylim(c(-(max(abs(data_pca$PC1), abs(data_pca$PC2))),
             max(abs(data_pca$PC1), abs(data_pca$PC2)))) + 
      coord_fixed(ratio = 1) + 
      theme_classic(base_size = 14)
    
    # Determine the proportion of variance of each component
    # Proportion of variance equals (PC stdev^2) / (sum all PCs stdev^2)
    pca_res <- prcomp(x = plot_data)
    propvariances <- ((pca_res$sdev ^ 2) / (sum(pca_res$sdev ^ 2))) * 100
    
    # check rotations to see which monocle2 states are contributing to which PC
    pca_res$rotation
    
  }
  
}


#running monocle2 analysis for oligo
{
  oligo_data <- readRDS(file = paste0(loc_superclusters,
                                      "oligo/oligo_data.rds"))
  
  #running monocle2
  micro_mono2 <- as(object = as.matrix(oligo_data[["SCT"]]@data), 
                    Class = "sparseMatrix")
  pheno_data <- data.frame(oligo_data@meta.data)
  pheno_data <- new(Class = "AnnotatedDataFrame", pheno_data)
  feat_data <- data.frame(gene_short_name = rownames(micro_mono2), 
                          row.names = rownames(micro_mono2))
  feat_data <- new(Class = "AnnotatedDataFrame", feat_data)
  
  micro_mono2 <- 
    monocle::newCellDataSet(micro_mono2, 
                            phenoData = pheno_data, 
                            featureData = feat_data, 
                            lowerDetectionLimit = 0.5, 
                            expressionFamily = VGAM::negbinomial.size())
  
  micro_mono2 <- DESeq2::estimateSizeFactors(micro_mono2)
  
  micro_mono2 <- DESeq2::estimateDispersions(micro_mono2)
  disp_table <- monocle:::dispersionTable(micro_mono2)
  #default mean_expression >= 0.5
  # 0.2 also worked with disp >= 1 but results in out of memory even at 500GB
  ordering_genes <- subset(disp_table,
                           mean_expression >= 0.05 &
                             dispersion_empirical >= 0.1 *
                             dispersion_fit)$gene_id
  print(paste0("Number of ordering genes: ", length(ordering_genes)))
  
  micro_mono2 <- monocle::setOrderingFilter(micro_mono2, 
  						ordering_genes)
  
  #default auto_param_selection = T
  micro_mono2 <- monocle::reduceDimension(cds = micro_mono2, 
                                          norm_method = "none", 
                                          pseudo_expr = 0, 
                                          reduction_method = "DDRTree", 
                                          relative_exprs = F, 
                                          scaling = F, 
                                          max_components = 20, 
                                          ncenter = 30)
  
  micro_mono2 <- monocle::orderCells(cds = micro_mono2)
  
  saveRDS(object = micro_mono2,
          file = paste0(loc_monocle, 
                        "oligo/oligo_data_monocle2.rds"), 
  	compress = T)
}
#visualisations
{
  oligo_data_monocle2 <- readRDS(file = paste0(loc_monocle, 
                                               "oligo/oligo_data_monocle2.rds"))
  
  monocle::plot_cell_trajectory(cds = oligo_data_monocle2, 
                                        color_by = "age", 
                                        cell_size = 0.1) 
  
  monocle::plot_complex_cell_trajectory(cds = oligo_data_monocle2, 
                                        color_by = "Pseudotime", 
                                        cell_size = 0.1) + 
    facet_wrap(~age+treatment, ncol = 2) + NoLegend()
  
  monocle::plot_complex_cell_trajectory(cds = oligo_data_monocle2, 
                                        color_by = "age", 
                                        cell_size = 0.5) + 
    facet_wrap(~treatment, ncol = 2) + NoLegend() + 
    scale_color_manual(values = c("#00BA38", "#F564E3"), 
                       name = "age") + 
    ggrastr::rasterise(geom_point(size=0.1, alpha=0.1)) + 
    theme(text = element_text(size = 20), 
          axis.text = element_text(size = 16))
  
  monocle::plot_complex_cell_trajectory(cds = oligo_data_monocle2, 
                                        color_by = "State", 
                                        cell_size = 0.1) + 
    facet_wrap(~State, ncol = 2) + NoLegend()
  
  monocle::plot_complex_cell_trajectory(cds = oligo_data_monocle2, 
                                        color_by = "age", 
                                        cell_size = 0.1) + 
    facet_wrap(~age+treatment, ncol = 2) + NoLegend()
  
  #get proportions of each state
  {
    proportions <- table(oligo_data_monocle2$age_treatment, 
                         oligo_data_monocle2$State)
    proportions <- (proportions/rowSums(proportions))*100
  }
  
  oligo_data_monocle2 <- monocle::orderCells(cds = oligo_data_monocle2,
                                             root_state = c(6))
  
  monocle::plot_cell_trajectory(cds = oligo_data_monocle2, 
                                        color_by = "age", 
                                        cell_size = 0.1)
  
  monocle::plot_complex_cell_trajectory(cds = oligo_data_monocle2, 
                                        color_by = "Pseudotime", 
                                        cell_size = 0.1) + 
    facet_wrap(~age+treatment, ncol = 2) + NoLegend()
  
  monocle::plot_complex_cell_trajectory(cds = oligo_data_monocle2, 
                                        color_by = "State", 
                                        cell_size = 0.1) + 
    facet_wrap(~State, ncol = 2) + NoLegend()
  
  monocle::plot_complex_cell_trajectory(cds = oligo_data_monocle2, 
                                        color_by = "age", 
                                        cell_size = 0.1) + 
    facet_wrap(~age+treatment, ncol = 2) + NoLegend()
  
  saveRDS(object = oligo_data_monocle2, 
          file = paste0(loc_monocle, "oligo/oligo_data_monocle2_reordered.rds"), 
          compress = T)
  
  oligo_data_monocle2_reordered <- 
    readRDS(file = paste0(loc_monocle, 
                          "oligo/oligo_data_monocle2_reordered.rds"))
  
  #the below is using the original non-reordered object
  cell_order <- rownames(oligo_data@meta.data)
  meta_pseudo <- as(object = oligo_data_monocle2@phenoData, 
                    Class = "data.frame")
  meta_pseudo <- meta_pseudo[cell_order, "Pseudotime"]
  oligo_data$pseudotime <- meta_pseudo
  oligo_data@meta.data <- metadata
  FeaturePlot(object = oligo_data, features = "pseudotime", raster = T) + 
    viridis::scale_colour_viridis() + labs(title = "Pseudotime") + 
    xlab("Dimension 1") + ylab("Dimension 2") + 
    coord_fixed() + theme(text = element_text(size = 20), 
                          axis.text = element_text(size = 16))
  
  #pca on proportions
  {
    proportions <- table(oligo_data_monocle2_reordered$age_treatment, 
                         oligo_data_monocle2_reordered$State)
    proportions <- (proportions/rowSums(proportions))*100
    proportions <- as.data.frame.matrix(proportions)
    
    plot_data <- proportions
    
    data_pca <- as.data.frame(prcomp(x = plot_data)$x)
    pca_vars <- summary(prcomp(x = plot_data))$importance[2,]
    
    data_pca$Age <- gsub(pattern = "\\..*$", replacement = "", 
                         x = rownames(data_pca), ignore.case = T)
    data_pca$Age <- as.factor(data_pca$Age)
    
    data_pca$Treatment <- gsub(pattern = "^.*m\\.?", replacement = "", 
                               x = rownames(data_pca), ignore.case = T)
    data_pca$Treatment <- as.factor(data_pca$Treatment)
    
    data_pca$Sample <- rownames(data_pca)
    
    ggplot(data = data_pca, 
           aes(x = PC1, y = PC2, colour = Treatment, shape = Age)) + 
      geom_point(size = 5) + 
      xlab(label = paste0("PC1 (", percent(pca_vars["PC1"]), 
                          " of total variance)")) + 
      ylab(label = paste0("PC2 (", percent(pca_vars["PC2"]), 
                          " of total variance)")) + 
      xlim(c(-(max(abs(data_pca$PC1), abs(data_pca$PC2))),
             max(abs(data_pca$PC1), abs(data_pca$PC2)))) + 
      ylim(c(-(max(abs(data_pca$PC1), abs(data_pca$PC2))),
             max(abs(data_pca$PC1), abs(data_pca$PC2)))) + 
      coord_fixed(ratio = 1) + 
      theme_classic(base_size = 14) 
    
    # Determine the proportion of variance of each component
    # Proportion of variance equals (PC stdev^2) / (sum all PCs stdev^2)
    pca_res <- prcomp(x = plot_data)
    propvariances <- ((pca_res$sdev ^ 2) / (sum(pca_res$sdev ^ 2))) * 100
    
    # check rotations to see which monocle2 states are contributing to which PC
    pca_res$rotation
    
    }
  
}
#overlay S100b and other expression to find root state
{
  oligo_data_monocle2_ordered <- 
    readRDS(file = 
              paste0(loc_seurat_results, 
                     "/sct-normalised/monocle2/oligo/oligo_data_monocle2_ordered.rds"))
  
  oligo_data <- readRDS(file = paste0(loc_superclusters, 
                                        "oligo/oligo_data.rds"))
  
  cell_order <- rownames(oligo_data_monocle2_ordered@phenoData)
  
  immat_oligo_exprs <- as.matrix(oligo_data@assays$SCT@counts[
    c("S100b", "Klk6", "Hopx", "Pdgfra", "C1ql1"),])
  
  immat_oligo_exprs <- immat_oligo_exprs[,cell_order]
  
  oligo_data_monocle2_ordered$S100b <- immat_oligo_exprs["S100b",]
  oligo_data_monocle2_ordered$S100b_exprsd <- 
    oligo_data_monocle2_ordered$S100b > 
    summary(oligo_data_monocle2_ordered$S100b)["Mean"]
  oligo_data_monocle2_ordered$Klk6 <- immat_oligo_exprs["Klk6",]
  oligo_data_monocle2_ordered$Klk6_exprsd <- 
    oligo_data_monocle2_ordered$Klk6 > 
    summary(oligo_data_monocle2_ordered$Klk6)["Mean"]
  oligo_data_monocle2_ordered$Hopx <- immat_oligo_exprs["Hopx",]
  oligo_data_monocle2_ordered$Hopx_exprsd <- 
    oligo_data_monocle2_ordered$Hopx > 
    summary(oligo_data_monocle2_ordered$Hopx)["Mean"]
  oligo_data_monocle2_ordered$Pdgfra <- immat_oligo_exprs["Pdgfra",]
  oligo_data_monocle2_ordered$Pdgfra_exprsd <- 
    oligo_data_monocle2_ordered$Pdgfra > 
    summary(oligo_data_monocle2_ordered$Pdgfra)["Mean"]
  oligo_data_monocle2_ordered$C1ql1 <- immat_oligo_exprs["C1ql1",]
  oligo_data_monocle2_ordered$C1ql1_exprsd <- 
    oligo_data_monocle2_ordered$C1ql1 > 
    summary(oligo_data_monocle2_ordered$C1ql1)["Mean"]
  
  monocle::plot_complex_cell_trajectory(cds = oligo_data_monocle2_ordered, 
                                        color_by = "Pseudotime", 
                                        cell_size = 0.5, 
                                        show_branch_points = T) + 
    facet_wrap(~age+treatment, ncol = 2) + NoLegend()
  monocle::plot_complex_cell_trajectory(cds = oligo_data_monocle2_ordered, 
                                        color_by = "State", 
                                        cell_size = 0.5, 
                                        show_branch_points = T) + 
    facet_wrap(~State, ncol = 2) + NoLegend()
  
  monocle::plot_complex_cell_trajectory(cds = oligo_data_monocle2_ordered, 
                                        color_by = "S100b", 
                                        cell_size = 0.5, 
                                        show_branch_points = T) + 
    facet_wrap(~age+treatment, ncol = 2) + NoLegend()
  monocle::plot_complex_cell_trajectory(cds = oligo_data_monocle2_ordered, 
                                        color_by = "Klk6", 
                                        cell_size = 0.5, 
                                        show_branch_points = T) + 
    facet_wrap(~age+treatment, ncol = 2) + NoLegend()
  monocle::plot_complex_cell_trajectory(cds = oligo_data_monocle2_ordered, 
                                        color_by = "Hopx", 
                                        cell_size = 0.5, 
                                        show_branch_points = T) + 
    facet_wrap(~age+treatment, ncol = 2) + NoLegend()
  monocle::plot_complex_cell_trajectory(cds = oligo_data_monocle2_ordered, 
                                        color_by = "Pdgfra", 
                                        cell_size = 0.5, 
                                        show_branch_points = T) + 
    facet_wrap(~age+treatment, ncol = 2) + NoLegend()
  monocle::plot_complex_cell_trajectory(cds = oligo_data_monocle2_ordered, 
                                        color_by = "C1ql1", 
                                        cell_size = 0.5, 
                                        show_branch_points = T) + 
    facet_wrap(~age+treatment, ncol = 2) + NoLegend()
  
  monocle::plot_complex_cell_trajectory(cds = oligo_data_monocle2_ordered, 
                                        color_by = "seurat_clusters", 
                                        cell_size = 0.5, 
                                        show_branch_points = T) + 
    facet_wrap(~age+treatment, ncol = 2) + NoLegend()
  
  #S100b proportions
  oligo_proportions <- table(oligo_data_monocle2_ordered$State, 
                             oligo_data_monocle2_ordered$S100b_exprsd, 
                             oligo_data_monocle2_ordered$age, 
                             oligo_data_monocle2_ordered$treatment)
  
  oligo_proportions_list <- list()
  for(i in seq(1, dim(oligo_proportions)[4])){
    oligo_proportions_list[[i]] <- list()
    
    for(j in seq(1, dim(oligo_proportions)[3])){
      oligo_proportions_list[[i]][[j]] <- 
        t(t(oligo_proportions[,,j,i])/colSums(oligo_proportions[,,j,i]))*100
      
    }
    names(oligo_proportions_list[[i]]) <- c("04m", "12m", "18m", "24m")
  }
  names(oligo_proportions_list) <- c("eGFP", "IL2")
  
  
  #reordering for highest S100b>mean(S100b) expressing state in 04m
  oligo_data_monocle2_reordered <-
    monocle::orderCells(cds = oligo_data_monocle2_ordered, root_state = 3)
  saveRDS(
    object = oligo_data_monocle2_reordered,
    file =
      paste0(loc_seurat_results,
             "/sct-normalised/monocle2/oligo/oligo_data_monocle2_reordered.rds"),
    compress = T)
  # oligo_data_monocle2_reordered <- 
  #   readRDS(file = 
  #             paste0(loc_seurat_results, 
  #            "/sct-normalised/monocle2/oligo/oligo_data_monocle2_reordered.rds")
  #           )
  
  monocle::plot_complex_cell_trajectory(cds = oligo_data_monocle2_reordered, 
                                        color_by = "Pseudotime", 
                                        cell_size = 0.5, 
                                        show_branch_points = T) + 
    facet_wrap(~age+treatment, ncol = 2) + NoLegend()
  monocle::plot_complex_cell_trajectory(cds = oligo_data_monocle2_reordered, 
                                        color_by = "State", 
                                        cell_size = 0.5, 
                                        show_branch_points = T) + 
    facet_wrap(~State, ncol = 2) + NoLegend()
  monocle::plot_complex_cell_trajectory(cds = oligo_data_monocle2_reordered, 
                                        color_by = "S100b", 
                                        cell_size = 0.5, 
                                        show_branch_points = T) + 
    facet_wrap(~age+treatment, ncol = 2) + NoLegend()
  monocle::plot_complex_cell_trajectory(cds = oligo_data_monocle2_reordered, 
                                        color_by = "Klk6", 
                                        cell_size = 0.5, 
                                        show_branch_points = T) + 
    facet_wrap(~age+treatment, ncol = 2) + NoLegend()
  monocle::plot_complex_cell_trajectory(cds = oligo_data_monocle2_reordered, 
                                        color_by = "Hopx", 
                                        cell_size = 0.5, 
                                        show_branch_points = T) + 
    facet_wrap(~age+treatment, ncol = 2) + NoLegend()
  
  #cell proportions
  {
    proportions <- table(oligo_data_monocle2_reordered$age_treatment, 
                         oligo_data_monocle2_reordered$State)
    proportions <- (proportions/rowSums(proportions))*100
  }
  
  #S100b proportions
  oligo_proportions <- table(oligo_data_monocle2_reordered$State, 
                             oligo_data_monocle2_reordered$S100b_exprsd, 
                             oligo_data_monocle2_reordered$age, 
                             oligo_data_monocle2_reordered$treatment)
  oligo_proportions_list <- list()
  for(i in seq(1, dim(oligo_proportions)[4])){
    #print(i)
    oligo_proportions_list[[i]] <- list()
    
    for(j in seq(1, dim(oligo_proportions)[3])){
      #print(j)
      oligo_proportions_list[[i]][[j]] <- 
        t(t(oligo_proportions[,,j,i])/colSums(oligo_proportions[,,j,i]))*100
      
    }
    names(oligo_proportions_list[[i]]) <- c("04m", "12m", "18m", "24m")
  }
  names(oligo_proportions_list) <- c("eGFP", "IL2")
  
  #Klk6 proportions
  oligo_proportions <- table(oligo_data_monocle2_reordered$State, 
                             oligo_data_monocle2_reordered$Klk6_exprsd, 
                             oligo_data_monocle2_reordered$age, 
                             oligo_data_monocle2_reordered$treatment)
  oligo_proportions_list <- list()
  for(i in seq(1, dim(oligo_proportions)[4])){
    #print(i)
    oligo_proportions_list[[i]] <- list()
    
    for(j in seq(1, dim(oligo_proportions)[3])){
      #print(j)
      oligo_proportions_list[[i]][[j]] <- 
        t(t(oligo_proportions[,,j,i])/colSums(oligo_proportions[,,j,i]))*100
      
    }
    names(oligo_proportions_list[[i]]) <- c("04m", "12m", "18m", "24m")
  }
  names(oligo_proportions_list) <- c("eGFP", "IL2")
  
  #Hopx proportions
  oligo_proportions <- table(oligo_data_monocle2_reordered$State, 
                             oligo_data_monocle2_reordered$Hopx_exprsd, 
                             oligo_data_monocle2_reordered$age, 
                             oligo_data_monocle2_reordered$treatment)
  oligo_proportions_list <- list()
  for(i in seq(1, dim(oligo_proportions)[4])){
    #print(i)
    oligo_proportions_list[[i]] <- list()
    
    for(j in seq(1, dim(oligo_proportions)[3])){
      #print(j)
      oligo_proportions_list[[i]][[j]] <- 
        t(t(oligo_proportions[,,j,i])/colSums(oligo_proportions[,,j,i]))*100
      
    }
    names(oligo_proportions_list[[i]]) <- c("04m", "12m", "18m", "24m")
  }
  names(oligo_proportions_list) <- c("eGFP", "IL2")

}
#sum of deviances - by samples relative to 04m GFP average
{
  monocle_results <- 
    list(readRDS(file = paste0(loc_monocle, 
                               "oligo/oligo_data_monocle2_reordered.rds")))
  names(monocle_results) <- c("Combined")
  
  #calculating sum of deviances
  list_of_deviances <- list()
  for(i in seq(1, length(monocle_results))){
    
    result <- monocle_results[[i]]
    
    proportions <- table(result$label, 
                         result$State)
    proportions <- (proportions/rowSums(proportions))*100
    proportions <- as.data.frame.matrix(proportions)
    
    #bringing 04m GFP to the top
    proportions <- rbind(proportions[which(rownames(proportions) %in% 
                                             c("04m_S01", "04m_S03", "04m_S05")),], 
                         proportions[-which(rownames(proportions) %in% 
                                              c("04m_S01", "04m_S03", "04m_S05")),])
    
    #finding mean of 04m_GFP and rearranging rows
    proportions <- rbind(colMeans(
      proportions[which(rownames(proportions) %in%
                          c("04m_S01", "04m_S03", "04m_S05")),]),
      proportions[-(which(rownames(proportions) %in%
                            c("04m_S01", "04m_S03", "04m_S05"))),])
    prop_scores <- NULL
    
    if(nrow(proportions) > 1){
      for(j in seq(2, nrow(proportions))){
        proportions[j,] <- abs(proportions[1,] - proportions[j,])
        rownames(proportions)[j] <- 
          paste0(rownames(proportions)[j], "-04m_GFP_avg")
      }
    }
    proportions[1,] <- 0
    rownames(proportions)[1] <- "Reference"
    
    sum_of_deviances <- rowSums(proportions)/2
    list_of_deviances[[length(list_of_deviances)+1]] <- sum_of_deviances
    names(list_of_deviances)[length(list_of_deviances)] <-
      names(monocle_results)[i]
      
  }
  
  #creating grouped col plot
  {
    plot_data <- cbind(names(list_of_deviances)[1], 
                       names(list_of_deviances[[1]]), 
                       list_of_deviances[[1]])
    rownames(plot_data) <- NULL
    for(i in seq(2, length(monocle_results))){
      
      plot_data <- rbind(plot_data, 
                         cbind(names(list_of_deviances)[i], 
                               names(list_of_deviances[[i]]), 
                               list_of_deviances[[i]]))
      rownames(plot_data) <- NULL
    }
    
    colnames(plot_data) <- c("Cell.Type", 
                             "Label", 
                             "Sum.of.Deviances")
    plot_data <- as.data.frame(plot_data)
    plot_data$Sum.of.Deviances <- as.numeric(plot_data$Sum.of.Deviances)
    plot_data$Age <- gsub(pattern = "_.*$", replacement = "", 
                          x = plot_data$Label, ignore.case = T)
    plot_data$Treatment <- NA
    plot_data$Treatment[grepl(pattern = "_S0(2|4|6)-.*$", 
                              x = plot_data$Label, 
                              ignore.case = T)] <- "PHP.GFAP-IL2"
    plot_data$Treatment[grepl(pattern = "_S0(1|3|5)-.*$", 
                              x = plot_data$Label, 
                              ignore.case = T)] <- "PHP.GFAP-GFP"
    plot_data$Treatment[is.na(plot_data$Treatment)] <- "PHP.GFAP-GFP"
    plot_data$Age.and.Treatment <- paste0(plot_data$Age, 
                                          " ", 
                                          plot_data$Treatment)
    
    plot_data$Sum.of.Deviances[is.na(plot_data$Treatment)] <- NA
    plot_data$Sum.of.Deviances[is.na(plot_data$Treatment)] <- -Inf
    plot_data$Age.and.Treatment[is.na(plot_data$Treatment)] <- NA
    
    #removing zero entries
    plot_data <- plot_data[-(which(plot_data$Sum.of.Deviances == 0)),]
    
    col_plot <- 
      ggplot(data = plot_data, 
             aes(y = Sum.of.Deviances, 
                 x = Cell.Type)) + 
      geom_dotplot(aes(fill = Age.and.Treatment), 
                   binaxis = "y", 
                   position = "dodge", 
                   stackdir = "center", 
                   dotsize = 1.5) + 
      stat_summary(aes(fill = Age.and.Treatment), 
                   fun.data=mean_sdl, fun.args = list(mult=1),
                   geom="errorbar", color="black", width=0.2, 
                   position = position_dodge(0.9)) +
      stat_summary(aes(fill = Age.and.Treatment), 
                   fun.y=mean, geom="point", color="black", size = 0.8, 
                   position = position_dodge(0.9)) + 
      labs(fill = "Age and Treatment", 
           y = "Distance from 04m GFP", 
           x = "Astrocyte group"
           ) + 
      expand_limits(y = 0) + 
      theme_light(base_size = 15) + 
      theme(panel.grid.minor = element_blank(),
            axis.text.x = element_text(angle = 0, hjust = 0.5))
    
    stat.test.within.cell.type <- plot_data %>% 
      group_by(Cell.Type) %>% 
      t_test(formula = Sum.of.Deviances ~ Age.and.Treatment) %>% 
      adjust_pvalue(method = "bonferroni") %>% 
      add_significance("p.adj")
    
    stat.test.within.cell.type <- stat.test.within.cell.type %>% 
      add_xy_position(x = "Cell.Type", dodge = 0.8)
    
    stat.test.within.cell.type$y.position <- 
      stat.test.within.cell.type$y.position + 
      (stat.test.within.cell.type$y.position * 0.05)
    
    col_plot & xlab(NULL) & 
      stat_pvalue_manual(data = stat.test.within.cell.type, 
                         label = "p.adj", tip.length = 0.01, size = 4.5) & 
      theme(legend.position = "bottom", 
            text = element_text(size = 20), 
            axis.text = element_text(size = 16)) & 
      scale_fill_brewer(palette = "Accent")
    
  }
    
  
}


#running monocle2 analysis for astro
{
  astro_data_all <- readRDS(file = paste0(loc_superclusters,
                                      "astro/astro_data.rds"))
  
  for(i in levels(micro_data_all@meta.data$seurat_clusters)){
    
    print(paste0("Proessing for cluster: ", i))
    
    if(i == 2){
      micro_data <- subset(x = micro_data_all, 
                           subset = seurat_clusters %in% c(2,3))
    }
    else if(i == 3){
      next
    }
    else{
      micro_data <- subset(x = micro_data_all, 
                           subset = seurat_clusters %in% c(i))
    }
    
    #running monocle2
    micro_mono2 <- as(object = as.matrix(micro_data[["SCT"]]@data), 
                      Class = "sparseMatrix")
    pheno_data <- data.frame(micro_data@meta.data)
    pheno_data <- new(Class = "AnnotatedDataFrame", pheno_data)
    feat_data <- data.frame(gene_short_name = rownames(micro_mono2), 
                            row.names = rownames(micro_mono2))
    feat_data <- new(Class = "AnnotatedDataFrame", feat_data)
    
    micro_mono2 <- 
      monocle::newCellDataSet(micro_mono2, 
                              phenoData = pheno_data, 
                              featureData = feat_data, 
                              lowerDetectionLimit = 0.5, 
                              expressionFamily = VGAM::negbinomial.size())
    
    micro_mono2 <- DESeq2::estimateSizeFactors(micro_mono2)
    
    micro_mono2 <- DESeq2::estimateDispersions(micro_mono2)
    disp_table <- monocle:::dispersionTable(micro_mono2)
    summary(disp_table)
    #default mean_expression >= 0.5
    # 0.2 also worked with disp >= 1 but results in out of memory even at 500GB
    
    if(i %in% c(0,1)){
      ordering_genes <- subset(disp_table, 
                               mean_expression >= 0.05 & 
                                 dispersion_empirical >= 0.4 * 
                                 dispersion_fit)$gene_id
    }
    else if(i %in% c(2)){
      ordering_genes <- subset(disp_table, 
                               mean_expression >= 0.06 & 
                                 dispersion_empirical >= 0.5 * 
                                 dispersion_fit)$gene_id
    }
    else if(i %in% c(4)){
      ordering_genes <- subset(disp_table, 
                               mean_expression >= 0.08 & 
                                 dispersion_empirical >= 0.1 * 
                                 dispersion_fit)$gene_id
    }
    else if(i %in% c(5)){
      ordering_genes <- subset(disp_table, 
                               mean_expression >= 0.1 & 
                                 dispersion_empirical >= 0.1 * 
                                 dispersion_fit)$gene_id
    }
    else if(i %in% c(6)){
      ordering_genes <- subset(disp_table, 
                               mean_expression >= 0.1 & 
                                 dispersion_empirical >= 0.2 * 
                                 dispersion_fit)$gene_id
    }
    
    print(paste0("Number of ordering genes: ", length(ordering_genes)))
    
    micro_mono2 <- monocle::setOrderingFilter(micro_mono2, 
    						ordering_genes)
    
    #default auto_param_selection = T
    micro_mono2 <- monocle::reduceDimension(cds = micro_mono2, 
                                            norm_method = "none", 
                                            pseudo_expr = 0, 
                                            reduction_method = "DDRTree", 
                                            relative_exprs = F, 
                                            scaling = F, 
                                            max_components = 20)
    
    micro_mono2 <- monocle::orderCells(cds = micro_mono2)
    
    saveRDS(object = micro_mono2,
            file = paste0(loc_monocle, 
                          "astro/astro_data_monocle2_cluster", i, ".rds"), 
    	compress = T)
    
    
    }
}
#visualisations
{
  astro_data_monocle2 <- 
    readRDS(file = paste0(loc_monocle,
                          "astro/astro_data_monocle2_cluster_5.rds"))
  
  monocle::plot_complex_cell_trajectory(cds = astro_data_monocle2, 
                                        color_by = "Pseudotime", 
                                        cell_size = 0.5) + 
    facet_wrap(~age+treatment, ncol = 2) + NoLegend()
  
  monocle::plot_complex_cell_trajectory(cds = astro_data_monocle2, 
                                        color_by = "age", 
                                        cell_size = 0.5) + 
    facet_wrap(~treatment, ncol = 2) + NoLegend() + 
    scale_color_manual(values = c("#00BA38", "#F564E3"), 
                       name = "age") + 
    labs(title = "Non-telencephalon Astrocytes")
    # labs(title = "Telencephalon Astrocytes")
    # labs(title = "Bergmann Glia")
    # labs(title = "Cerebellar Astrocytes")
    # labs(title = "Olfactory Astrocytes")
    # labs(title = "Striatal Astrocytes")
  
  monocle::plot_complex_cell_trajectory(cds = astro_data_monocle2, 
                                        color_by = "State", 
                                        cell_size = 0.5) + 
    facet_wrap(~State, ncol = 2) + NoLegend()
  
  #get proportions of each state
  {
    proportions <- table(astro_data_monocle2$age_treatment, 
                         astro_data_monocle2$State)
    proportions <- (proportions/rowSums(proportions))*100
    proportions <- as.data.frame.matrix(proportions)
  }
  #appears to already be sorted by age (highest young in State 1, 
  # highest old in State 7, which is the largest leaf node)
  
  monocle::plot_cell_trajectory(cds = astro_data_monocle2, 
                                        color_by = "age", 
                                        cell_size = 0.5) 
  
  #for astro cluster 1, cluster2,3 , cluster4 and cluster 6
  astro_data_monocle2 <- monocle::orderCells(cds = astro_data_monocle2, 
                                             root_state = c(3))
  
  #for astro cluster 0
  astro_data_monocle2 <- monocle::orderCells(cds = astro_data_monocle2, 
                                             root_state = c(5))
  
  saveRDS(object = astro_data_monocle2, 
          file = paste0(loc_monocle,
                          "astro/astro_data_monocle2_cluster_6.rds"), 
          compress = T)
  
  # astro_data_monocle2 <-
  #   readRDS(file = paste0(loc_monocle,
  #                         "astro/astro_data_monocle2_cluster_0.rds"))
  
  #pca on proportions
  {
    proportions <- table(astro_data_monocle2$age_treatment, 
                         astro_data_monocle2$State)
    proportions <- (proportions/rowSums(proportions))*100
    proportions <- as.data.frame.matrix(proportions)
    
    plot_data <- proportions
    
    data_pca <- as.data.frame(prcomp(x = plot_data)$x)
    pca_vars <- summary(prcomp(x = plot_data))$importance[2,]
    
    data_pca$Age <- gsub(pattern = "\\..*$", replacement = "", 
                         x = rownames(data_pca), ignore.case = T)
    data_pca$Age <- as.factor(data_pca$Age)
    
    data_pca$Treatment <- gsub(pattern = "^.*m\\.?", replacement = "", 
                               x = rownames(data_pca), ignore.case = T)
    data_pca$Treatment <- as.factor(data_pca$Treatment)
    
    data_pca$Sample <- rownames(data_pca)
    
    ggplot(data = data_pca, 
           aes(x = PC1, y = PC2, colour = Treatment, shape = Age)) + 
      geom_point(size = 5) + 
      xlab(label = paste0("PC1 (", percent(pca_vars["PC1"]), 
                          " of total variance)")) + 
      ylab(label = paste0("PC2 (", percent(pca_vars["PC2"]), 
                          " of total variance)")) + 
      xlim(c(-(max(abs(data_pca$PC1), abs(data_pca$PC2))),
             max(abs(data_pca$PC1), abs(data_pca$PC2)))) + 
      ylim(c(-(max(abs(data_pca$PC1), abs(data_pca$PC2))),
             max(abs(data_pca$PC1), abs(data_pca$PC2)))) + 
      coord_fixed(ratio = 1) + 
      theme_classic(base_size = 14) 
    
    # Determine the proportion of variance of each component
    # Proportion of variance equals (PC stdev^2) / (sum all PCs stdev^2)
    pca_res <- prcomp(x = plot_data)
    propvariances <- ((pca_res$sdev ^ 2) / (sum(pca_res$sdev ^ 2))) * 100
    
    # check rotations to see which monocle2 states are contributing to which PC
    pca_res$rotation
    
    }
  
}
#sum of deviances - by samples relative to 04m GFP average
{
  monocle_results <- 
    list(readRDS(file = paste0(loc_monocle, 
                               "astro/astro_data_monocle2_cluster_0.rds")), 
         readRDS(file = paste0(loc_monocle, 
                               "astro/astro_data_monocle2_cluster_1.rds")), 
         readRDS(file = paste0(loc_monocle, 
                               "astro/astro_data_monocle2_cluster_2,3.rds")), 
         readRDS(file = paste0(loc_monocle, 
                               "astro/astro_data_monocle2_cluster_4.rds")), 
         readRDS(file = paste0(loc_monocle, 
                               "astro/astro_data_monocle2_cluster_5.rds")), 
         readRDS(file = paste0(loc_monocle, 
                               "astro/astro_data_monocle2_cluster_6.rds")))
  names(monocle_results) <- c("Non-telencephalon astrocytes", 
                              "Telencephalon astrocytes", 
                              "Bergmann glia", 
                              "Cerebellar astrocytes", 
                              "Olfactory astrocytes", 
                              "Striatal astrocytes")
  
  #calculating sum of deviances
  list_of_deviances <- list()
  for(i in seq(1, length(monocle_results))){
    result <- monocle_results[[i]]
    
    proportions <- table(result$label, 
                         result$State)
    proportions <- (proportions/rowSums(proportions))*100
    proportions <- as.data.frame.matrix(proportions)
    
    #bringing 04m GFP to the top
    proportions <- rbind(proportions[which(rownames(proportions) %in% 
                                             c("04m_S01", "04m_S03", "04m_S05")),], 
                         proportions[-which(rownames(proportions) %in% 
                                              c("04m_S01", "04m_S03", "04m_S05")),])
    
    #finding mean of 04m_GFP and rearranging rows
    proportions <- rbind(colMeans(
      proportions[which(rownames(proportions) %in%
                          c("04m_S01", "04m_S03", "04m_S05")),]),
      proportions[-(which(rownames(proportions) %in%
                            c("04m_S01", "04m_S03", "04m_S05"))),])
    prop_scores <- NULL
    
    if(nrow(proportions) > 1){
      for(j in seq(2, nrow(proportions))){
        proportions[j,] <- abs(proportions[1,] - proportions[j,])
        rownames(proportions)[j] <- 
          paste0(rownames(proportions)[j], "-04m_GFP_avg")
      }
    }
    proportions[1,] <- 0
    rownames(proportions)[1] <- "Reference"
    
    sum_of_deviances <- rowSums(proportions)/2
    list_of_deviances[[length(list_of_deviances)+1]] <- sum_of_deviances
    names(list_of_deviances)[length(list_of_deviances)] <-
      names(monocle_results)[i]
  }
  
  #creating grouped col plot
  {
    plot_data <- cbind(names(list_of_deviances)[1], 
                       names(list_of_deviances[[1]]), 
                       list_of_deviances[[1]])
    rownames(plot_data) <- NULL
    for(i in seq(2, length(monocle_results))){
      
      plot_data <- rbind(plot_data, 
                         cbind(names(list_of_deviances)[i], 
                               names(list_of_deviances[[i]]), 
                               list_of_deviances[[i]]))
      rownames(plot_data) <- NULL
    }
    
    colnames(plot_data) <- c("Cell.Type", 
                             "Label", 
                             "Sum.of.Deviances")
    plot_data <- as.data.frame(plot_data)
    plot_data$Sum.of.Deviances <- as.numeric(plot_data$Sum.of.Deviances)
    plot_data$Age <- gsub(pattern = "_.*$", replacement = "", 
                          x = plot_data$Label, ignore.case = T)
    plot_data$Treatment <- NA
    plot_data$Treatment[grepl(pattern = "_S0(2|4|6)-.*$", 
                              x = plot_data$Label, 
                              ignore.case = T)] <- "PHP.GFAP-IL2"
    plot_data$Treatment[grepl(pattern = "_S0(1|3|5)-.*$", 
                              x = plot_data$Label, 
                              ignore.case = T)] <- "PHP.GFAP-GFP"
    plot_data$Treatment[is.na(plot_data$Treatment)] <- "PHP.GFAP-GFP"
    plot_data$Age.and.Treatment <- paste0(plot_data$Age, 
                                          " ", 
                                          plot_data$Treatment)
    
    plot_data$Sum.of.Deviances[is.na(plot_data$Treatment)] <- NA
    plot_data$Sum.of.Deviances[is.na(plot_data$Treatment)] <- -Inf
    plot_data$Age.and.Treatment[is.na(plot_data$Treatment)] <- NA
    
    #removing zero entries
    plot_data <- plot_data[-(which(plot_data$Sum.of.Deviances == 0)),]
    
    col_plot <- 
      ggplot(data = plot_data, 
             aes(y = Sum.of.Deviances, 
                 x = Cell.Type)) + 
      geom_dotplot(aes(fill = Age.and.Treatment), 
                   binaxis = "y", 
                   position = "dodge", 
                   stackdir = "center", 
                   dotsize = 1.5) + 
      stat_summary(aes(fill = Age.and.Treatment), 
                   fun.data=mean_sdl, fun.args = list(mult=1),
                   geom="errorbar", color="black", width=0.2, 
                   position = position_dodge(0.9)) +
      stat_summary(aes(fill = Age.and.Treatment), 
                   fun.y=mean, geom="point", color="black", size = 0.8, 
                   position = position_dodge(0.9)) + 
      labs(fill = "Age and Treatment", 
           y = "Distance from 04m GFP", 
           x = "Astrocyte group"
           ) + 
      expand_limits(y = 0) + 
      theme_light(base_size = 15) + 
      theme(panel.grid.minor = element_blank(),
            axis.text.x = element_text(angle = 0, hjust = 0.5))
    
    stat.test.within.cell.type <- plot_data %>% 
      group_by(Cell.Type) %>% 
      t_test(formula = Sum.of.Deviances ~ Age.and.Treatment) %>% 
      adjust_pvalue(method = "bonferroni") %>% 
      add_significance("p.adj")
    
    stat.test.within.cell.type <- stat.test.within.cell.type %>% 
      add_xy_position(x = "Cell.Type", dodge = 0.8)
    
    stat.test.within.cell.type$y.position <- 
      stat.test.within.cell.type$y.position + 
      (stat.test.within.cell.type$y.position * 0.05)
    
    col_plot & xlab(NULL) & 
      stat_pvalue_manual(data = stat.test.within.cell.type, 
                         label = "p.adj", tip.length = 0.01, size = 4.5) & 
      theme(legend.position = "bottom", 
            text = element_text(size = 20), 
            axis.text = element_text(size = 16)) & 
      scale_fill_brewer(palette = "Accent")
    
  }
    
  
}


```


#differential expression analysis
```{r}
#finding markers for reclustered microglia
{
  micro_data <- readRDS(file = paste0(loc_superclusters,
                                      "micro/micro_data.rds"))
  
  #diff exprss GFP
  {
    seurat_object <- subset(micro_data, treatment %in% c("PHP.GFAP-GFP"))
    
    markers_list <- 
      get_diff_exprs_list(seurat_object = seurat_object, 
                          cluster_column = "age")
    
    saveRDS(object = markers_list,
            file = paste0(loc_diff_exprs_results,
                          "micro/micro_markers_list_gfp_nofilter.rds"),
            compress = T)
    
  }
  #diff exprss IL2
  {
    seurat_object <- subset(micro_data, treatment %in% c("PHP.GFAP-IL2"))
    
    markers_list <- 
      get_diff_exprs_list(seurat_object = seurat_object, 
                          cluster_column = "age")
    
    saveRDS(object = markers_list,
            file = paste0(loc_diff_exprs_results,
                          "micro/micro_markers_list_il2_nofilter.rds"),
            compress = T)
    
  }
  #diff exprs contrasting treatments grouped by age
  {
    seurat_object <- subset(micro_data, age %in% c("04m"))
    markers_list <- 
      get_diff_exprs_list(seurat_object = seurat_object, 
                          cluster_column = "treatment")
    
    saveRDS(object = markers_list,
            file = paste0(loc_diff_exprs_results,
                          "micro/micro_markers_list_04m_nofilter.rds"),
            compress = T)
    
    seurat_object <- subset(micro_data, age %in% c("24m"))
    markers_list <- 
      get_diff_exprs_list(seurat_object = seurat_object, 
                          cluster_column = "treatment")
    
    saveRDS(object = markers_list,
            file = paste0(loc_diff_exprs_results,
                          "micro/micro_markers_list_24m_nofilter.rds"),
            compress = T)
  }
  
  #2d exprs plot -- age and treatment contrast axes
  {
    markers_list_il2 <- readRDS(file = paste0(loc_diff_exprs_results,
                                      "micro/micro_markers_list_il2_nofilter.rds"))
    markers_list_gfp <- readRDS(file = paste0(loc_diff_exprs_results,
                                      "micro/micro_markers_list_gfp_nofilter.rds"))
    markers_list_24m <- readRDS(file = paste0(loc_diff_exprs_results,
                                      "micro/micro_markers_list_24m_nofilter.rds"))
    markers_list_04m <- readRDS(file = paste0(loc_diff_exprs_results,
                                      "micro/micro_markers_list_04m_nofilter.rds"))
    
    exprs2d <- 
      get_2d_diff_exprs_plot(FindMarkers_result_x = markers_list_gfp[[1]], 
                             FindMarkers_result_y = markers_list_24m[[2]], 
                             label_x = "\n04m GFP vs 24m GFP", 
                             label_y = "\n24m IL2 vs 24m GFP", 
                             main_title = NULL, 
                             label_quad1 = "", 
                             label_quad3 = "",
                             fold_change_thresh = 0.5, 
                             adj_p_val_thresh = 0.01, 
                             axisdh = 0.5, 
                             show_corr = T)
    exprs2d + rasterise(layer = geom_point(), dpi = 150)
    
    write.table(x = markers_list_il2[[2]],
                file = paste0(loc_diff_exprs_results,
                              "micro/IL2_24m_vs_IL2_04m.tab"),
                quote = F, sep = "\t", row.names = T, col.names = T)
    write.table(x = markers_list_gfp[[2]],
                file = paste0(loc_diff_exprs_results,
                              "micro/GFP_24m_vs_GFP_04m.tab"),
                quote = F, sep = "\t", row.names = T, col.names = T)
    write.table(x = markers_list_24m[[2]],
                file = paste0(loc_diff_exprs_results,
                              "micro/24m_IL2_vs_24m_GFP.tab"),
                quote = F, sep = "\t", row.names = T, col.names = T)
    write.table(x = markers_list_04m[[2]],
                file = paste0(loc_diff_exprs_results,
                              "micro/04m_IL2_vs_04m_GFP.tab"),
                quote = F, sep = "\t", row.names = T, col.names = T)
    
  }
  
}


#finding markers for reclustered oligodendrocytes 
{
  oligo_data <- readRDS(file = paste0(loc_superclusters,
                                      "oligo/oligo_data.rds"))
  
  #diff exprs GFP
  {
    seurat_object <- subset(oligo_data, treatment %in% c("PHP.GFAP-GFP"))
  
    markers_list <- 
      get_diff_exprs_list(seurat_object = seurat_object, 
                          cluster_column = "age")
    
    saveRDS(object = markers_list,
            file = paste0(loc_diff_exprs_results,
                          "oligo/oligo_markers_list_gfp_nofilter.rds"),
            compress = T)
    
  }
  #diff exprs IL2
  {
    seurat_object <- subset(oligo_data, treatment %in% c("PHP.GFAP-IL2"))
  
    markers_list <- 
      get_diff_exprs_list(seurat_object = seurat_object, 
                          cluster_column = "age")
    
    saveRDS(object = markers_list,
            file = paste0(loc_diff_exprs_results,
                          "oligo/oligo_markers_list_il2_nofilter.rds"),
            compress = T)
    
  }
  #diff exprs contrasting treatments grouped by age
  {
    seurat_object <- subset(oligo_data, age %in% c("04m"))
    markers_list <- 
      get_diff_exprs_list(seurat_object = seurat_object, 
                          cluster_column = "treatment")
    
    saveRDS(object = markers_list,
            file = paste0(loc_diff_exprs_results,
                          "oligo/oligo_markers_list_04m_nofilter.rds"),
            compress = T)
    
    seurat_object <- subset(oligo_data, age %in% c("24m"))
    markers_list <- 
      get_diff_exprs_list(seurat_object = seurat_object, 
                          cluster_column = "treatment")
    
    saveRDS(object = markers_list,
            file = paste0(loc_diff_exprs_results,
                          "oligo/oligo_markers_list_24m_nofilter.rds"),
            compress = T)
  }
  
  #2d exprs plot -- age and treatment contrast axes
  {
    markers_list_il2 <- readRDS(file = paste0(loc_diff_exprs_results,
                                      "oligo/oligo_markers_list_il2_nofilter.rds"))
    markers_list_gfp <- readRDS(file = paste0(loc_diff_exprs_results,
                                      "oligo/oligo_markers_list_gfp_nofilter.rds"))
    markers_list_24m <- readRDS(file = paste0(loc_diff_exprs_results,
                                      "oligo/oligo_markers_list_24m_nofilter.rds"))
    markers_list_04m <- readRDS(file = paste0(loc_diff_exprs_results,
                                      "oligo/oligo_markers_list_04m_nofilter.rds"))
    
    exprs2d <- 
      get_2d_diff_exprs_plot(FindMarkers_result_x = markers_list_gfp[[1]], 
                             FindMarkers_result_y = markers_list_24m[[2]], 
                             label_x = "\n04m GFP vs 24m GFP", 
                             label_y = "\n24m IL2 vs 24m GFP", 
                             main_title = NULL, 
                             label_quad1 = "", 
                             label_quad3 = "", 
                             fold_change_thresh = 0.5, 
                             adj_p_val_thresh = 0.01, 
                             axisdh = 0.5, 
                             show_corr = T)
    exprs2d + rasterise(layer = geom_point(), dpi = 150)
    
    write.table(x = markers_list_il2[[2]],
                file = paste0(loc_diff_exprs_results,
                              "oligo/IL2_24m_vs_IL2_04m.tab"),
                quote = F, sep = "\t", row.names = T, col.names = T)
    write.table(x = markers_list_gfp[[2]],
                file = paste0(loc_diff_exprs_results,
                              "oligo/GFP_24m_vs_GFP_04m.tab"),
                quote = F, sep = "\t", row.names = T, col.names = T)
    write.table(x = markers_list_24m[[2]],
                file = paste0(loc_diff_exprs_results,
                              "oligo/24m_IL2_vs_24m_GFP.tab"),
                quote = F, sep = "\t", row.names = T, col.names = T)
    write.table(x = markers_list_04m[[2]],
                file = paste0(loc_diff_exprs_results,
                              "oligo/04m_IL2_vs_04m_GFP.tab"),
                quote = F, sep = "\t", row.names = T, col.names = T)
    
  }
  
}


#finding markers for reclustered astrocytes 
{
  astro_data <- readRDS(file = paste0(loc_superclusters,
                                      "astro/astro_data.rds"))
  
  #diff exprs GFP
  {
    seurat_object <- subset(astro_data, treatment %in% c("PHP.GFAP-GFP"))
    
    markers_list <- 
      get_diff_exprs_list(seurat_object = seurat_object, 
                          cluster_column = "age")
    
    saveRDS(object = markers_list,
            file = paste0(loc_diff_exprs_results,
                          "astro/astro_markers_list_gfp_nofilter.rds"),
            compress = T)
    
  }
  #diff exprs IL2
  {
    seurat_object <- subset(astro_data, treatment %in% c("PHP.GFAP-IL2"))
    
    markers_list <- 
      get_diff_exprs_list(seurat_object = seurat_object, 
                          cluster_column = "age")
    
    saveRDS(object = markers_list,
            file = paste0(loc_diff_exprs_results,
                          "astro/astro_markers_list_il2_nofilter.rds"),
            compress = T)
    
  }
  #diff exprss by astrocyte subpopulation no filter, contrasting ages
  {
    for(subpop in sort(as.character(
      unique(astro_data$seurat_clusters_renamed)))){
      
      #diff exprs GFP
      {
        seurat_object <- subset(astro_data, 
                                treatment %in% c("PHP.GFAP-GFP") & 
                                  seurat_clusters_renamed %in% c(subpop))
        
        markers_list <- 
          get_diff_exprs_list(seurat_object = seurat_object, 
                              cluster_column = "age")
        
        saveRDS(object = markers_list,
                file = paste0(loc_diff_exprs_results,
                              "astro/subpops/", 
                              gsub(pattern = "(\n| )+", replacement = ".", 
                                   x = subpop, ignore.case = T), 
                              "_markers_list_gfp_nofilter.rds"),
                compress = T)
      }
      #diff exprs IL2
      {
        seurat_object <- subset(astro_data, 
                                treatment %in% c("PHP.GFAP-IL2") & 
                                  seurat_clusters_renamed %in% c(subpop))
        
        markers_list <- 
          get_diff_exprs_list(seurat_object = seurat_object, 
                              cluster_column = "age")
        
        saveRDS(object = markers_list,
                file = paste0(loc_diff_exprs_results,
                              "astro/subpops/", 
                              gsub(pattern = "(\n| )+", replacement = ".", 
                                   x = subpop, ignore.case = T), 
                              "_markers_list_il2_nofilter.rds"),
                compress = T)
      }
      
    }
    
  }
  #diff exprs contrasting treatments
  {
    seurat_object <- subset(astro_data, age %in% c("04m"))
    markers_list <- 
      get_diff_exprs_list(seurat_object = seurat_object, 
                          cluster_column = "treatment")
    
    saveRDS(object = markers_list,
            file = paste0(loc_diff_exprs_results,
                          "astro/astro_markers_list_04m_nofilter.rds"),
            compress = T)
    
    seurat_object <- subset(astro_data, age %in% c("24m"))
    markers_list <- 
      get_diff_exprs_list(seurat_object = seurat_object, 
                          cluster_column = "treatment")
    
    saveRDS(object = markers_list,
            file = paste0(loc_diff_exprs_results,
                          "astro/astro_markers_list_24m_nofilter.rds"),
            compress = T)
  }
  
  #2d exprs plot -- age and treatment contrast axes
  {
    markers_list_il2 <- readRDS(file = paste0(loc_diff_exprs_results,
                                      "astro/astro_markers_list_il2_nofilter.rds"))
    markers_list_gfp <- readRDS(file = paste0(loc_diff_exprs_results,
                                      "astro/astro_markers_list_gfp_nofilter.rds"))
    markers_list_24m <- readRDS(file = paste0(loc_diff_exprs_results,
                                      "astro/astro_markers_list_24m_nofilter.rds"))
    markers_list_04m <- readRDS(file = paste0(loc_diff_exprs_results,
                                      "astro/astro_markers_list_04m_nofilter.rds"))
    
    exprs2d <- 
      get_2d_diff_exprs_plot(FindMarkers_result_x = markers_list_gfp[[1]], 
                             FindMarkers_result_y = markers_list_24m[[2]], 
                             label_x = "\n04m GFP vs 24m GFP", 
                             label_y = "\n24m IL2 vs 24m GFP", 
                             main_title = NULL, 
                             label_quad1 = "", 
                             label_quad3 = "", 
                             fold_change_thresh = 0.5, 
                             adj_p_val_thresh = 0.01, 
                             axisdh = 0.5, 
                             show_corr = T)
    exprs2d + rasterise(layer = geom_point(), dpi = 150)
    
    write.table(x = markers_list_il2[[2]],
                file = paste0(loc_diff_exprs_results,
                              "astro/IL2_24m_vs_IL2_04m.tab"),
                quote = F, sep = "\t", row.names = T, col.names = T)
    write.table(x = markers_list_gfp[[2]],
                file = paste0(loc_diff_exprs_results,
                              "astro/GFP_24m_vs_GFP_04m.tab"),
                quote = F, sep = "\t", row.names = T, col.names = T)
    write.table(x = markers_list_24m[[2]],
                file = paste0(loc_diff_exprs_results,
                              "astro/24m_IL2_vs_24m_GFP.tab"),
                quote = F, sep = "\t", row.names = T, col.names = T)
    write.table(x = markers_list_04m[[2]],
                file = paste0(loc_diff_exprs_results,
                              "astro/04m_IL2_vs_04m_GFP.tab"),
                quote = F, sep = "\t", row.names = T, col.names = T)
    
  }
  
}


```


#GSEA analysis using GAGE and PathView
```{r}

#reading in the seurat objects and kegg_genesets
{
  micro_data <- readRDS(file = paste0(loc_superclusters,
                                      "micro/micro_data.rds"))
  oligo_data <- readRDS(file = paste0(loc_superclusters,
                                      "oligo/oligo_data.rds"))
  astro_data <- readRDS(file = paste0(loc_superclusters,
                                      "astro/astro_data.rds"))
  
  #mapping gene symbols to entrez/ncbi ids (gage only uses these)
  {
    genes <- data.frame(genes_symbol = rownames(micro_data[["SCT"]]@data))
    genes <- as.data.frame(id2eg(ids = genes$genes_symbol, 
                                 category = "SYMBOL", 
                                 org = "mouse", 
                                 na.rm = F))
    genes$ENTREZID[genes$ENTREZID == "NA"] <- NA
    
    #no duplicates found -- in case of duplicates, check tissue Tregs scripts 
    # for removal code
  }
  
  #loading/downloading kegg genesets
  {
    kegg_genesets <- kegg.gsets(species = "mouse",
                                id.type = "entrez",
                                check.new = T)
    saveRDS(object = kegg_genesets,
            file = paste0(loc_pathview_results, "kegg_genesets.rds"),
            compress = T)
    # kegg_genesets <- readRDS(file = paste0(loc_pathview_results,
    #                                        "kegg_genesets.rds"))
    
  }
  
  
}


#running gage analysis for micro old vs young in GFP
{
  #loading metadata
  metadata <- micro_data@meta.data
  
  #getting cell names and finding column indices
  young_cells <- rownames(subset(metadata,
                               metadata$age_treatment == "04m.PHP.GFAP-GFP"))
  old_cells <- rownames(subset(metadata,
                               metadata$age_treatment == "24m.PHP.GFAP-GFP"))
  young_cells <- which(colnames(micro_data[["SCT"]]@data) %in% young_cells)
  old_cells <- which(colnames(micro_data[["SCT"]]@data) %in% old_cells)
  
  #extracting and formatting log counts matrix to minimise ram footprint
  counts_matrix <- micro_data[["SCT"]]@data[,c(young_cells, old_cells)]
  counts_matrix <- as.data.frame(counts_matrix)
  gc()
  
  #mapping ids
  # do check that the order is the same
  # all(unique(rownames(counts_matrix) == genes$SYMBOL))
  counts_matrix$ENTREZID <- genes$ENTREZID
  counts_matrix <- counts_matrix[complete.cases(counts_matrix),]
  rownames(counts_matrix) <- counts_matrix$ENTREZID
  counts_matrix$ENTREZID <- NULL
  
  #running gage
  gage_results <- gage(exprs = counts_matrix, 
                       gsets = kegg_genesets$kg.sets, 
                       ref = (1:length(young_cells)), 
                       samp = ((length(young_cells)+1) : ncol(counts_matrix)), 
                       same.dir = F, 
                       compare = "as.group", 
                       set.size = c(15, 1000)
                       )
  gc()
  
  saveRDS(object = gage_results, 
          file = paste0(loc_pathview_results, "micro/gage_results_GFP.rds"), 
          compress = T)
  # gage_results <- readRDS(file = paste0(loc_pathview_results, 
  #                                       "micro/gage_results_GFP.rds"))
  
}
#running gage analysis for micro old vs young in IL2
{
  #loading metadata
  metadata <- micro_data@meta.data
  
  #getting cell names and finding column indices
  young_cells <- rownames(subset(metadata,
                               metadata$age_treatment == "04m.PHP.GFAP-IL2"))
  old_cells <- rownames(subset(metadata,
                               metadata$age_treatment == "24m.PHP.GFAP-IL2"))
  young_cells <- which(colnames(micro_data[["SCT"]]@data) %in% young_cells)
  old_cells <- which(colnames(micro_data[["SCT"]]@data) %in% old_cells)
  
  #extracting and formatting log counts matrix to minimise ram footprint
  counts_matrix <- micro_data[["SCT"]]@data[,c(young_cells, old_cells)]
  counts_matrix <- as.data.frame(counts_matrix)
  gc()
  
  #mapping ids
  # do check that the order is the same
  # all(unique(rownames(counts_matrix) == genes$SYMBOL))
  counts_matrix$ENTREZID <- genes$ENTREZID
  counts_matrix <- counts_matrix[complete.cases(counts_matrix),]
  rownames(counts_matrix) <- counts_matrix$ENTREZID
  counts_matrix$ENTREZID <- NULL
  
  #running gage
  gage_results <- gage(exprs = counts_matrix, 
                       gsets = kegg_genesets$kg.sets, 
                       ref = (1:length(young_cells)), 
                       samp = ((length(young_cells)+1) : ncol(counts_matrix)), 
                       same.dir = F, 
                       compare = "as.group", 
                       set.size = c(15, 1000)
                       )
  gc()
  
  saveRDS(object = gage_results, 
          file = paste0(loc_pathview_results, "micro/gage_results_IL2.rds"), 
          compress = T)
  
}
#running gage analysis for micro IL2 vs GFP in young
{
  #loading metadata
  metadata <- micro_data@meta.data
  
  #getting cell names and finding column indices
  gfp_cells <- rownames(subset(metadata,
                               metadata$age_treatment == "04m.PHP.GFAP-GFP"))
  il2_cells <- rownames(subset(metadata,
                               metadata$age_treatment == "04m.PHP.GFAP-IL2"))
  gfp_cells <- which(colnames(micro_data[["SCT"]]@data) %in% gfp_cells)
  il2_cells <- which(colnames(micro_data[["SCT"]]@data) %in% il2_cells)
  
  #extracting and formatting log counts matrix to minimise ram footprint
  counts_matrix <- micro_data[["SCT"]]@data[,c(gfp_cells, il2_cells)]
  counts_matrix <- as.data.frame(counts_matrix)
  gc()
  
  #mapping ids
  # do check that the order is the same
  # all(unique(rownames(counts_matrix) == genes$SYMBOL))
  counts_matrix$ENTREZID <- genes$ENTREZID
  counts_matrix <- counts_matrix[complete.cases(counts_matrix),]
  rownames(counts_matrix) <- counts_matrix$ENTREZID
  counts_matrix$ENTREZID <- NULL
  
  #running gage
  gage_results <- gage(exprs = counts_matrix, 
                       gsets = kegg_genesets$kg.sets, 
                       ref = (1:length(gfp_cells)), 
                       samp = ((length(gfp_cells)+1) : ncol(counts_matrix)), 
                       same.dir = F, 
                       compare = "as.group", 
                       set.size = c(15, 1000)
                       )
  gc()
  
  saveRDS(object = gage_results, 
          file = paste0(loc_pathview_results, "micro/gage_results_04m.rds"), 
          compress = T)
  
  # gage_results <- readRDS(file = paste0(loc_pathview_results, 
  #                                       "micro/gage_results_04m.rds"))
  
}
#running gage analysis for micro IL2 vs GFP in old
{
  #loading metadata
  metadata <- micro_data@meta.data
  
  #getting cell names and finding column indices
  gfp_cells <- rownames(subset(metadata,
                               metadata$age_treatment == "24m.PHP.GFAP-GFP"))
  il2_cells <- rownames(subset(metadata,
                               metadata$age_treatment == "24m.PHP.GFAP-IL2"))
  gfp_cells <- which(colnames(micro_data[["SCT"]]@data) %in% gfp_cells)
  il2_cells <- which(colnames(micro_data[["SCT"]]@data) %in% il2_cells)
  
  #extracting and formatting log counts matrix to minimise ram footprint
  counts_matrix <- micro_data[["SCT"]]@data[,c(gfp_cells, il2_cells)]
  counts_matrix <- as.data.frame(counts_matrix)
  gc()
  
  #mapping ids
  # do check that the order is the same
  # all(unique(rownames(counts_matrix) == genes$SYMBOL))
  counts_matrix$ENTREZID <- genes$ENTREZID
  counts_matrix <- counts_matrix[complete.cases(counts_matrix),]
  rownames(counts_matrix) <- counts_matrix$ENTREZID
  counts_matrix$ENTREZID <- NULL
  
  #running gage
  gage_results <- gage(exprs = counts_matrix, 
                       gsets = kegg_genesets$kg.sets, 
                       ref = (1:length(gfp_cells)), 
                       samp = ((length(gfp_cells)+1) : ncol(counts_matrix)), 
                       same.dir = F, 
                       compare = "as.group", 
                       set.size = c(15, 1000)
                       )
  gc()
  
  saveRDS(object = gage_results, 
          file = paste0(loc_pathview_results, "micro/gage_results_24m.rds"), 
          compress = T)
  
  # gage_results <- readRDS(file = paste0(loc_pathview_results, 
  #                                       "micro/gage_results_04m.rds"))
  
}
#running gagePipe and gageComp on all micro
{
  #loading metadata
  metadata <- micro_data@meta.data
  
  #creating the customised kegg.gs object
  kegg.gs <- kegg_genesets$kg.sets
  
  #getting cell names and finding column indices
  {
    old_gfp_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "24m.PHP.GFAP-GFP"))
    old_il2_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "24m.PHP.GFAP-IL2"))
    young_gfp_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "04m.PHP.GFAP-GFP"))
    young_il2_cells <- rownames(subset(metadata, 
                                 metadata$age_treatment == "04m.PHP.GFAP-IL2"))
    
    old_gfp_cells <- which(colnames(micro_data[["SCT"]]@data) %in% 
                             old_gfp_cells)
    old_il2_cells <- which(colnames(micro_data[["SCT"]]@data) %in% 
                             old_il2_cells)
    young_gfp_cells <- which(colnames(micro_data[["SCT"]]@data) %in% 
                               young_gfp_cells)
    young_il2_cells <- which(colnames(micro_data[["SCT"]]@data) %in% 
                               young_il2_cells)
  }
  
  #creating sample and reference lists
  sampList <- list(old_gfp_cells, old_il2_cells, young_il2_cells, old_il2_cells)
  refList <- list(young_gfp_cells, young_il2_cells, young_gfp_cells, 
                   old_gfp_cells)
  sampnames <- c("c_GFP.24m.vs.GFP.04m", "c_IL2.24m.vs.IL2.04m", 
                 "c_04m.IL2.vs.04m.GFP", "c_24m.IL2.vs.24m.GFP")
  counts_data <- as.data.frame(micro_data@assays$SCT@data)
  
  #mapping ids
  # do check that the order is the same
  # all(unique(rownames(counts_data) == genes$SYMBOL))
  counts_data$ENTREZID <- genes$ENTREZID
  counts_data <- counts_data[complete.cases(counts_data),]
  rownames(counts_data) <- counts_data$ENTREZID
  counts_data$ENTREZID <- NULL
  
  #saving work.dir and restoring it later
  work.dir <- getwd()
  setwd(paste0(loc_pathview_results, "micro/gage_combined/"))
  
  #running gage pipe
  gagePipe(arraydata = counts_data, 
           dataname = "micro_", trim.at = F, 
           sampnames = sampnames, 
           gsname = "kegg.gs", 
           ref.list = refList, 
           samp.list = sampList, 
           comp.list = "as.group", 
           heatmap = T)
  
  #load("micro_.gage.RData")
  
  #running gage comparison
  gageComp(sampnames = sampnames, 
           dataname = "micro_", 
           gsname = "kegg.gs")
  
  #restoring work.dir
  setwd(work.dir)
  
  #garbage collection
  gc()
  
}
#running pathview analysis on treatment and age results
{
  #loading gage results
  {
    gage_results_gfp <- readRDS(file = paste0(loc_pathview_results, 
                                              "micro/gage_results_GFP.rds"))
    
    gage_results_il2 <- readRDS(file = paste0(loc_pathview_results, 
                                              "micro/gage_results_IL2.rds"))
    
    gage_results_04m <- readRDS(file = paste0(loc_pathview_results, 
                                              "micro/gage_results_04m.rds"))
    
    gage_results_24m <- readRDS(file = paste0(loc_pathview_results, 
                                              "micro/gage_results_24m.rds"))
  }
  
  #loading differential expression results
  {
    #loading GFP results
    diff_exprs_results_gfp <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "micro/micro_markers_list_gfp_nofilter.rds"))[[2]]
    diff_exprs_results_gfp <-
      diff_exprs_results_gfp[diff_exprs_results_gfp$p_val_adj < 0.1, "avg_log2FC",
                         drop = F]
    
    #loading IL2 results
    diff_exprs_results_il2 <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "micro/micro_markers_list_il2_nofilter.rds"))[[2]]
    diff_exprs_results_il2 <-
      diff_exprs_results_il2[diff_exprs_results_il2$p_val_adj < 0.1, "avg_log2FC",
                         drop = F]
    
    #loading 04m results
    diff_exprs_results_04m <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "micro/micro_markers_list_04m_nofilter.rds"))[[2]]
    diff_exprs_results_04m <-
      diff_exprs_results_04m[diff_exprs_results_04m$p_val_adj < 0.1, "avg_log2FC",
                         drop = F]
    
    #loading 24m results
    diff_exprs_results_24m <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "micro/micro_markers_list_24m_nofilter.rds"))[[2]]
    diff_exprs_results_24m <-
      diff_exprs_results_24m[diff_exprs_results_24m$p_val_adj < 0.1, "avg_log2FC",
                         drop = F]
    
    genes_to_select <- unique(c(rownames(diff_exprs_results_gfp), 
                                rownames(diff_exprs_results_il2), 
                                rownames(diff_exprs_results_04m), 
                                rownames(diff_exprs_results_24m)))
    
    #reloading differential expression results with genes significant in at 
    # least one contrast
    {
      #loading GFP results
      diff_exprs_results_gfp <- 
        readRDS(file = paste0(loc_diff_exprs_results, 
                              "micro/micro_markers_list_gfp_nofilter.rds"))[[2]]
      diff_exprs_results_gfp <-
        diff_exprs_results_gfp[genes_to_select, "avg_log2FC",
                           drop = F]
      
      #loading IL2 results
      diff_exprs_results_il2 <- 
        readRDS(file = paste0(loc_diff_exprs_results, 
                              "micro/micro_markers_list_il2_nofilter.rds"))[[2]]
      diff_exprs_results_il2 <-
        diff_exprs_results_il2[genes_to_select, "avg_log2FC",
                           drop = F]
      
      #loading 04m results
      diff_exprs_results_04m <- 
        readRDS(file = paste0(loc_diff_exprs_results, 
                              "micro/micro_markers_list_04m_nofilter.rds"))[[2]]
      diff_exprs_results_04m <-
        diff_exprs_results_04m[genes_to_select, "avg_log2FC",
                           drop = F]
      
      #loading 24m results
      diff_exprs_results_24m <- 
        readRDS(file = paste0(loc_diff_exprs_results, 
                              "micro/micro_markers_list_24m_nofilter.rds"))[[2]]
      diff_exprs_results_24m <-
        diff_exprs_results_24m[genes_to_select, "avg_log2FC",
                           drop = F]
    }
    
  }
  
  #merging diff exprs data
  diff_exprs_results <- merge(x = diff_exprs_results_gfp, 
                              y = diff_exprs_results_il2, 
                              by = "row.names", 
                              all = T)
  diff_exprs_results_2 <- merge(x = diff_exprs_results_04m, 
                              y = diff_exprs_results_24m, 
                              by = "row.names", 
                              all = T)
  diff_exprs_results <- merge(x = diff_exprs_results, 
                              y = diff_exprs_results_2, 
                              by = "Row.names", 
                              all = T)
  rownames(diff_exprs_results) <- diff_exprs_results$Row.names
  diff_exprs_results$Row.names <- NULL
  diff_exprs_results_2 <- NULL
  colnames(diff_exprs_results) <- paste0("Contrast ", 
                                         1:ncol(diff_exprs_results))
  
  #to exclude those extreme cases where a gene was not detected in a contrast
  diff_exprs_results <- diff_exprs_results[complete.cases(diff_exprs_results),]
  
  #setting up pathview
  data(gene.idtype.bods)
  work.dir <- getwd()
  path <- paste0(loc_pathview_results, "micro/pathview_treatment_and_age/")
  dir.create(path = path, recursive = T)
  setwd(path)
  
  #writing out the gage results
  write.table(x = gage_results_gfp$greater, 
                  file = "gage_results_GFP.tab", 
                  quote = F, 
                  sep = "\t", 
                  row.names = T)
  write.table(x = gage_results_il2$greater, 
                  file = "gage_results_IL2.tab", 
                  quote = F, 
                  sep = "\t", 
                  row.names = T)
  write.table(x = gage_results_04m$greater, 
                  file = "gage_results_04m.tab", 
                  quote = F, 
                  sep = "\t", 
                  row.names = T)
  write.table(x = gage_results_24m$greater, 
                  file = "gage_results_24m.tab", 
                  quote = F, 
                  sep = "\t", 
                  row.names = T)
  
  #extracting pathway ids
  pathway_ids <- 
    c(rownames(gage_results_gfp$greater), 
      rownames(gage_results_il2$greater), 
      rownames(gage_results_04m$greater), 
      rownames(gage_results_24m$greater))
  pathway_ids <- gsub(pattern = " .*$", 
                      replacement = "", 
                      x = pathway_ids, 
                      ignore.case = T)
  pathway_ids <- unique(pathway_ids)
  #adding missing pathway ids
  pathway_ids <- c(pathway_ids, "mmu05022", "mmu04370", "mmu04012", "mmu04014", 
                   "mmu04211", "mmu04657", "mmu01200", "mmu04310", "mmu04024", 
                   "mmu04213", "mmu04371", "mmu04145", "mmu04150", "mmu04015", 
                   "mmu04722", "mmu04068", "mmu04210", "mmu04213", "mmu04218", 
                   "mmu04136")
  pathway_ids <- unique(pathway_ids)
  
  #running pathview in batches due to an undocumented limitation of a maximum
  # number of KEGG ids that can be queried in a single call. Not sure what value
  # and whether the limitation is in pathview or on KEGG's end
  pathway_ids <- split(pathway_ids, ceiling(seq_along(pathway_ids)/20))
  for(batch in pathway_ids){
    pathview(gene.data = diff_exprs_results, 
             cpd.data = NULL, 
             #pathway.id = pathway_ids, 
             pathway.id = batch, 
             species = "mouse", 
             kegg.dir = path, 
             gene.idtype = "SYMBOL", 
             low = list(gene = "#2a7fff", cpd = "blue"), 
             mid = list(gene = "white", cpd = "gray"), 
             high = list(gene = "#ff5555", cpd = "yellow"), 
             limit = list(gene = 0.5, cpd = 1), 
             bins = list(gene = 10, cpd = 10), 
             kegg.native = T, 
             na.col = "gray", 
             same.layer = F)
    }
  
  #resetting work dir
  setwd(work.dir)
  
}
#creating the manual summarised comparison -- ref contrast -- figure 6
{
  gene_sets <- kegg_genesets$kg.sets
  
  for(i in seq(1, length(gene_sets))){
    gene_sets[[i]] <- genes$SYMBOL[which(genes$ENTREZID %in% gene_sets[[i]])]
  }
  
  #loading differential expression results
  {
    #loading GFP results
    diff_exprs_results_gfp <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "micro/micro_markers_list_gfp_nofilter.rds"))[[2]]
    diff_exprs_results_gfp <-
      diff_exprs_results_gfp[diff_exprs_results_gfp$p_val_adj < 0.1, "avg_log2FC",
                         drop = F]
    
    #loading IL2 results
    diff_exprs_results_il2 <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "micro/micro_markers_list_il2_nofilter.rds"))[[2]]
    diff_exprs_results_il2 <-
      diff_exprs_results_il2[diff_exprs_results_il2$p_val_adj < 0.1, "avg_log2FC",
                         drop = F]
    
    #loading 04m results
    diff_exprs_results_04m <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "micro/micro_markers_list_04m_nofilter.rds"))[[2]]
    diff_exprs_results_04m <-
      diff_exprs_results_04m[diff_exprs_results_04m$p_val_adj < 0.1, "avg_log2FC",
                         drop = F]
    
    #loading 24m results
    diff_exprs_results_24m <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "micro/micro_markers_list_24m_nofilter.rds"))[[2]]
    diff_exprs_results_24m <-
      diff_exprs_results_24m[diff_exprs_results_24m$p_val_adj < 0.1, "avg_log2FC",
                         drop = F]
    
    genes_to_select <- unique(c(rownames(diff_exprs_results_gfp), 
                                rownames(diff_exprs_results_il2), 
                                rownames(diff_exprs_results_04m), 
                                rownames(diff_exprs_results_24m)))
    
    #reloading differential expression results with genes significant in at 
    # least one contrast
    {
      #loading GFP results
      diff_exprs_results_gfp <- 
        readRDS(file = paste0(loc_diff_exprs_results, 
                              "micro/micro_markers_list_gfp_nofilter.rds"))[[2]]
      diff_exprs_results_gfp <-
        diff_exprs_results_gfp[genes_to_select, "avg_log2FC",
                           drop = F]
      
      #loading IL2 results
      diff_exprs_results_il2 <- 
        readRDS(file = paste0(loc_diff_exprs_results, 
                              "micro/micro_markers_list_il2_nofilter.rds"))[[2]]
      diff_exprs_results_il2 <-
        diff_exprs_results_il2[genes_to_select, "avg_log2FC",
                           drop = F]
      
      #loading 04m results
      diff_exprs_results_04m <- 
        readRDS(file = paste0(loc_diff_exprs_results, 
                              "micro/micro_markers_list_04m_nofilter.rds"))[[2]]
      diff_exprs_results_04m <-
        diff_exprs_results_04m[genes_to_select, "avg_log2FC",
                           drop = F]
      
      #loading 24m results
      diff_exprs_results_24m <- 
        readRDS(file = paste0(loc_diff_exprs_results, 
                              "micro/micro_markers_list_24m_nofilter.rds"))[[2]]
      diff_exprs_results_24m <-
        diff_exprs_results_24m[genes_to_select, "avg_log2FC",
                           drop = F]
    }
    
  }
  
  #formatting data frame of expression directions
  avg_sign_df <- data.frame(A = double(), B = double(), 
                            C = double(), D = double())
  for(i in seq(1, length(gene_sets))){
    #importing contrast data
    cont_ref <- diff_exprs_results_gfp[gene_sets[[i]],]
    cont_a <- diff_exprs_results_il2[gene_sets[[i]],]
    cont_b <- diff_exprs_results_04m[gene_sets[[i]],]
    cont_c <- diff_exprs_results_24m[gene_sets[[i]],]
    
    #checking which indices are negative in ref contrast
    flip_index <- which(cont_ref < 0)
    
    #if even a single negative value is found
    if(length(flip_index) > 0){
      cont_ref[flip_index] <- -(cont_ref[flip_index])
      cont_a[flip_index] <- -(cont_a[flip_index])
      cont_b[flip_index] <- -(cont_b[flip_index])
      cont_c[flip_index] <- -(cont_c[flip_index])
    }
    
    #formatting entry
    entry <- 
      c(mean(sign(cont_ref), na.rm = T), 
        mean(sign(cont_a), na.rm = T), 
        mean(sign(cont_b), na.rm = T), 
        mean(sign(cont_c), na.rm = T)
        )
    
    #if nothing was found in any contrast (no qualifying genes in any contrast
    # for a given pathway)
    if(all(is.na(entry))){
      next
    }
    
    #appending entry
    avg_sign_df <- rbind(avg_sign_df, entry)
    rownames(avg_sign_df)[nrow(avg_sign_df)] <- names(gene_sets)[i]
  }
  colnames(avg_sign_df) <- c("GFP 24m vs GFP 04m (ref)", "IL2 24m vs IL2 04m", 
                             "04m IL2 vs 04m GFP", "24m IL2 vs 24m GFP")
  
  #writing out full results
  write.table(x = avg_sign_df %>% rownames_to_column(var = "Pathway"), 
              file = paste0(loc_04mv24m_results, 
                            "/00_figures/Figure6/", 
                            "micro_direction_data.tab"), 
              quote = F, sep = "\t", row.names = F, col.names = T)
  
  #formatting word wrapping
  rownames(avg_sign_df) <- str_wrap(string = rownames(avg_sign_df), width = 40)
  
  #reordering data frame for heatmap
  ordering <- rowMeans(avg_sign_df[complete.cases(avg_sign_df),], na.rm = T)
  ordering <- ordering[order(ordering, decreasing = T)]
  avg_sign_df <- 
    rbind(avg_sign_df[rownames(avg_sign_df) %in% names(ordering),], 
          avg_sign_df[!rownames(avg_sign_df) %in% names(ordering),])
  ordering <- rev(rownames(avg_sign_df))
  
  #importing pre-selected pathways
  selected_mmu <- c("mmu03013", "mmu04066", "mmu04071", "mmu04151", "mmu04218", 
                    "mmu04620", "mmu04664", "mmu04722", "mmu05010", "mmu05012", 
                    "mmu00190", "mmu03010", "mmu03040", "mmu03050", "mmu04012", 
                    "mmu04014", "mmu04137", "mmu04140", "mmu04141", "mmu04142", 
                    "mmu04144", "mmu04150", "mmu04211", "mmu04657", "mmu04660", 
                    "mmu04810")
  avg_sign_df <- avg_sign_df[gsub(pattern = " .*$", 
                                  replacement = "", 
                                  x = rownames(avg_sign_df), 
                                  ignore.case = T) %in% selected_mmu,]
  
  #converting to heatmap
  avg_sign_df <- avg_sign_df %>% 
    rownames_to_column() %>% 
    gather(colname, value, -rowname)
  
  avg_sign_df$rowname <- factor(x = avg_sign_df$rowname, 
                                levels = ordering, ordered = T)
  avg_sign_df$colname <- factor(x = avg_sign_df$colname, 
                                levels = c("GFP 24m vs GFP 04m (ref)", 
                                           "IL2 24m vs IL2 04m", 
                                           "04m IL2 vs 04m GFP", 
                                           "24m IL2 vs 24m GFP"), 
                                ordered = T)
  
  #inverting values for colour scale in ggplot2
  avg_sign_df$value <- -(avg_sign_df$value)
  
  #plot
  ggplot(avg_sign_df, 
         aes(x = colname, 
             y = rowname, 
             fill = value)) + 
    geom_tile() + 
    scale_fill_gradient2(low = ("#3DBEDB"), #munsell::mnsl("5B 7/8")
                         mid = "white", 
                         high = ("#CAAE42"), #munsell::mnsl("5Y 7/8") 
                         breaks = c(-1, -0.5, 0, 0.5, 1), 
                         labels = c("Closer to\naged control", "", "", "", 
                                    "Closer to\nyoung control"), 
                         limits = c(-1, 1)) + 
    xlab(NULL) + ylab(NULL) + labs(fill = NULL) + 
    theme_light() + 
    theme(axis.text.x = element_text(size = 13, angle = 25, 
                                     hjust = 0.9, vjust = 0.9), 
          text = element_text(size = 16), 
          axis.text.y =  element_text(size = 14))
  
}


#running gage analysis for oligo old vs young in GFP
{
  #loading metadata
  metadata <- oligo_data@meta.data
  
  #getting cell names and finding column indices
  young_cells <- rownames(subset(metadata,
                               metadata$age_treatment == "04m.PHP.GFAP-GFP"))
  old_cells <- rownames(subset(metadata,
                               metadata$age_treatment == "24m.PHP.GFAP-GFP"))
  young_cells <- which(colnames(oligo_data[["SCT"]]@data) %in% young_cells)
  old_cells <- which(colnames(oligo_data[["SCT"]]@data) %in% old_cells)
  
  #extracting and formatting log counts matrix to minimise ram footprint
  counts_matrix <- oligo_data[["SCT"]]@data[,c(young_cells, old_cells)]
  counts_matrix <- as.data.frame(counts_matrix)
  gc()
  
  #mapping ids
  # do check that the order is the same
  # all(unique(rownames(counts_matrix) == genes$SYMBOL))
  counts_matrix$ENTREZID <- genes$ENTREZID
  counts_matrix <- counts_matrix[complete.cases(counts_matrix),]
  rownames(counts_matrix) <- counts_matrix$ENTREZID
  counts_matrix$ENTREZID <- NULL
  
  #running gage
  gage_results <- gage(exprs = counts_matrix, 
                       gsets = kegg_genesets$kg.sets, 
                       ref = (1:length(young_cells)), 
                       samp = ((length(young_cells)+1) : ncol(counts_matrix)), 
                       same.dir = F, 
                       compare = "as.group", 
                       set.size = c(15, 1000)
                       )
  gc()
  
  saveRDS(object = gage_results, 
          file = paste0(loc_pathview_results, "oligo/gage_results_GFP.rds"), 
          compress = T)
  
}
#running gage analysis for oligo old vs young in IL2
{
  #loading metadata
  metadata <- oligo_data@meta.data
  
  #getting cell names and finding column indices
  young_cells <- rownames(subset(metadata,
                               metadata$age_treatment == "04m.PHP.GFAP-IL2"))
  old_cells <- rownames(subset(metadata,
                               metadata$age_treatment == "24m.PHP.GFAP-IL2"))
  young_cells <- which(colnames(oligo_data[["SCT"]]@data) %in% young_cells)
  old_cells <- which(colnames(oligo_data[["SCT"]]@data) %in% old_cells)
  
  #extracting and formatting log counts matrix to minimise ram footprint
  counts_matrix <- oligo_data[["SCT"]]@data[,c(young_cells, old_cells)]
  counts_matrix <- as.data.frame(counts_matrix)
  gc()
  
  #mapping ids
  # do check that the order is the same
  # all(unique(rownames(counts_matrix) == genes$SYMBOL))
  counts_matrix$ENTREZID <- genes$ENTREZID
  counts_matrix <- counts_matrix[complete.cases(counts_matrix),]
  rownames(counts_matrix) <- counts_matrix$ENTREZID
  counts_matrix$ENTREZID <- NULL
  
  #running gage
  gage_results <- gage(exprs = counts_matrix, 
                       gsets = kegg_genesets$kg.sets, 
                       ref = (1:length(young_cells)), 
                       samp = ((length(young_cells)+1) : ncol(counts_matrix)), 
                       same.dir = F, 
                       compare = "as.group", 
                       set.size = c(15, 1000)
                       )
  gc()
  
  saveRDS(object = gage_results, 
          file = paste0(loc_pathview_results, "oligo/gage_results_IL2.rds"), 
          compress = T)
  
}
#running gage analysis for oligo IL2 vs GFP in young
{
  #loading metadata
  metadata <- oligo_data@meta.data
  
  #getting cell names and finding column indices
  gfp_cells <- rownames(subset(metadata,
                               metadata$age_treatment == "04m.PHP.GFAP-GFP"))
  il2_cells <- rownames(subset(metadata,
                               metadata$age_treatment == "04m.PHP.GFAP-IL2"))
  gfp_cells <- which(colnames(oligo_data[["SCT"]]@data) %in% gfp_cells)
  il2_cells <- which(colnames(oligo_data[["SCT"]]@data) %in% il2_cells)
  
  #extracting and formatting log counts matrix to minimise ram footprint
  counts_matrix <- oligo_data[["SCT"]]@data[,c(gfp_cells, il2_cells)]
  counts_matrix <- as.data.frame(counts_matrix)
  gc()
  
  #mapping ids
  # do check that the order is the same
  # all(unique(rownames(counts_matrix) == genes$SYMBOL))
  counts_matrix$ENTREZID <- genes$ENTREZID
  counts_matrix <- counts_matrix[complete.cases(counts_matrix),]
  rownames(counts_matrix) <- counts_matrix$ENTREZID
  counts_matrix$ENTREZID <- NULL
  
  #running gage
  gage_results <- gage(exprs = counts_matrix, 
                       gsets = kegg_genesets$kg.sets, 
                       ref = (1:length(gfp_cells)), 
                       samp = ((length(gfp_cells)+1) : ncol(counts_matrix)), 
                       same.dir = F, 
                       compare = "as.group", 
                       set.size = c(15, 1000)
                       )
  gc()
  
  saveRDS(object = gage_results, 
          file = paste0(loc_pathview_results, "oligo/gage_results_04m.rds"), 
          compress = T)
  
}
#running gage analysis for oligo IL2 vs GFP in old
{
  #loading metadata
  metadata <- oligo_data@meta.data
  
  #getting cell names and finding column indices
  gfp_cells <- rownames(subset(metadata,
                               metadata$age_treatment == "24m.PHP.GFAP-GFP"))
  il2_cells <- rownames(subset(metadata,
                               metadata$age_treatment == "24m.PHP.GFAP-IL2"))
  gfp_cells <- which(colnames(oligo_data[["SCT"]]@data) %in% gfp_cells)
  il2_cells <- which(colnames(oligo_data[["SCT"]]@data) %in% il2_cells)
  
  #extracting and formatting log counts matrix to minimise ram footprint
  counts_matrix <- oligo_data[["SCT"]]@data[,c(gfp_cells, il2_cells)]
  counts_matrix <- as.data.frame(counts_matrix)
  gc()
  
  #mapping ids
  # do check that the order is the same
  # all(unique(rownames(counts_matrix) == genes$SYMBOL))
  counts_matrix$ENTREZID <- genes$ENTREZID
  counts_matrix <- counts_matrix[complete.cases(counts_matrix),]
  rownames(counts_matrix) <- counts_matrix$ENTREZID
  counts_matrix$ENTREZID <- NULL
  
  #running gage
  gage_results <- gage(exprs = counts_matrix, 
                       gsets = kegg_genesets$kg.sets, 
                       ref = (1:length(gfp_cells)), 
                       samp = ((length(gfp_cells)+1) : ncol(counts_matrix)), 
                       same.dir = F, 
                       compare = "as.group", 
                       set.size = c(15, 1000)
                       )
  gc()
  
  saveRDS(object = gage_results, 
          file = paste0(loc_pathview_results, "oligo/gage_results_24m.rds"), 
          compress = T)
  
}
#running gagePipe and gageComp on all micro
{
  #loading metadata
  metadata <- oligo_data@meta.data
  
  #creating the customised kegg.gs object
  kegg.gs <- kegg_genesets$kg.sets
  
  #getting cell names and finding column indices
  {
    old_gfp_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "24m.PHP.GFAP-GFP"))
    old_il2_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "24m.PHP.GFAP-IL2"))
    young_gfp_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "04m.PHP.GFAP-GFP"))
    young_il2_cells <- rownames(subset(metadata, 
                                 metadata$age_treatment == "04m.PHP.GFAP-IL2"))
    
    old_gfp_cells <- which(colnames(oligo_data[["SCT"]]@data) %in% 
                             old_gfp_cells)
    old_il2_cells <- which(colnames(oligo_data[["SCT"]]@data) %in% 
                             old_il2_cells)
    young_gfp_cells <- which(colnames(oligo_data[["SCT"]]@data) %in% 
                               young_gfp_cells)
    young_il2_cells <- which(colnames(oligo_data[["SCT"]]@data) %in% 
                               young_il2_cells)
  }
  
  #creating sample and reference lists
  sampList <- list(old_gfp_cells, old_il2_cells, young_il2_cells, old_il2_cells)
  refList <- list(young_gfp_cells, young_il2_cells, young_gfp_cells, 
                   old_gfp_cells)
  sampnames <- c("c_GFP.24m.vs.GFP.04m", "c_IL2.24m.vs.IL2.04m", 
                 "c_04m.IL2.vs.04m.GFP", "c_24m.IL2.vs.24m.GFP")
  counts_data <- as.data.frame(oligo_data@assays$SCT@data)
  
  #mapping ids
  # do check that the order is the same
  # all(unique(rownames(counts_data) == genes$SYMBOL))
  counts_data$ENTREZID <- genes$ENTREZID
  counts_data <- counts_data[complete.cases(counts_data),]
  rownames(counts_data) <- counts_data$ENTREZID
  counts_data$ENTREZID <- NULL
  
  #saving work.dir and restoring it later
  work.dir <- getwd()
  setwd(paste0(loc_pathview_results, "oligo/gage_combined/"))
  
  #running gage pipe
  gagePipe(arraydata = counts_data, 
           dataname = "oligo_", trim.at = F, 
           sampnames = sampnames, 
           gsname = "kegg.gs", 
           ref.list = refList, 
           samp.list = sampList, 
           comp.list = "as.group", 
           heatmap = T)
  
  #running gage comparison
  gageComp(sampnames = sampnames, 
           dataname = "oligo_", 
           gsname = "kegg.gs")
  
  #restoring work.dir
  setwd(work.dir)
  
  #garbage collection
  gc()
  
}
#running pathview analysis on treatment and age results
{
  #loading gage results
  {
    gage_results_gfp <- readRDS(file = paste0(loc_pathview_results, 
                                              "oligo/gage_results_GFP.rds"))
    
    gage_results_il2 <- readRDS(file = paste0(loc_pathview_results, 
                                              "oligo/gage_results_IL2.rds"))
    
    gage_results_04m <- readRDS(file = paste0(loc_pathview_results, 
                                              "oligo/gage_results_04m.rds"))
    
    gage_results_24m <- readRDS(file = paste0(loc_pathview_results, 
                                              "oligo/gage_results_24m.rds"))
  }
  
  #loading differential expression results
  {
    #loading GFP results
    diff_exprs_results_gfp <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "oligo/oligo_markers_list_gfp_nofilter.rds"))[[2]]
    diff_exprs_results_gfp <-
      diff_exprs_results_gfp[diff_exprs_results_gfp$p_val_adj < 0.1, "avg_log2FC",
                         drop = F]
    
    #loading IL2 results
    diff_exprs_results_il2 <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "oligo/oligo_markers_list_il2_nofilter.rds"))[[2]]
    diff_exprs_results_il2 <-
      diff_exprs_results_il2[diff_exprs_results_il2$p_val_adj < 0.1, "avg_log2FC",
                         drop = F]
    
    #loading 04m results
    diff_exprs_results_04m <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "oligo/oligo_markers_list_04m_nofilter.rds"))[[2]]
    diff_exprs_results_04m <-
      diff_exprs_results_04m[diff_exprs_results_04m$p_val_adj < 0.1, "avg_log2FC",
                         drop = F]
    
    #loading 24m results
    diff_exprs_results_24m <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "oligo/oligo_markers_list_24m_nofilter.rds"))[[2]]
    diff_exprs_results_24m <-
      diff_exprs_results_24m[diff_exprs_results_24m$p_val_adj < 0.1, "avg_log2FC",
                         drop = F]
    
    genes_to_select <- unique(c(rownames(diff_exprs_results_gfp), 
                                rownames(diff_exprs_results_il2), 
                                rownames(diff_exprs_results_04m), 
                                rownames(diff_exprs_results_24m)))
    
    #reloading differential expression results with genes significant in at 
    # least one contrast
    {
      #loading GFP results
      diff_exprs_results_gfp <- 
        readRDS(file = paste0(loc_diff_exprs_results, 
                              "oligo/oligo_markers_list_gfp_nofilter.rds"))[[2]]
      diff_exprs_results_gfp <-
        diff_exprs_results_gfp[genes_to_select, "avg_log2FC",
                           drop = F]
      
      #loading IL2 results
      diff_exprs_results_il2 <- 
        readRDS(file = paste0(loc_diff_exprs_results, 
                              "oligo/oligo_markers_list_il2_nofilter.rds"))[[2]]
      diff_exprs_results_il2 <-
        diff_exprs_results_il2[genes_to_select, "avg_log2FC",
                           drop = F]
      
      #loading 04m results
      diff_exprs_results_04m <- 
        readRDS(file = paste0(loc_diff_exprs_results, 
                              "oligo/oligo_markers_list_04m_nofilter.rds"))[[2]]
      diff_exprs_results_04m <-
        diff_exprs_results_04m[genes_to_select, "avg_log2FC",
                           drop = F]
      
      #loading 24m results
      diff_exprs_results_24m <- 
        readRDS(file = paste0(loc_diff_exprs_results, 
                              "oligo/oligo_markers_list_24m_nofilter.rds"))[[2]]
      diff_exprs_results_24m <-
        diff_exprs_results_24m[genes_to_select, "avg_log2FC",
                           drop = F]
    }
    
  }
  
  #merging diff exprs data
  diff_exprs_results <- merge(x = diff_exprs_results_gfp, 
                              y = diff_exprs_results_il2, 
                              by = "row.names", 
                              all = T)
  diff_exprs_results_2 <- merge(x = diff_exprs_results_04m, 
                              y = diff_exprs_results_24m, 
                              by = "row.names", 
                              all = T)
  diff_exprs_results <- merge(x = diff_exprs_results, 
                              y = diff_exprs_results_2, 
                              by = "Row.names", 
                              all = T)
  rownames(diff_exprs_results) <- diff_exprs_results$Row.names
  diff_exprs_results$Row.names <- NULL
  diff_exprs_results_2 <- NULL
  colnames(diff_exprs_results) <- paste0("Contrast ", 
                                         1:ncol(diff_exprs_results))
  
  #to exclude those extreme cases where a gene was not detected in a contrast
  diff_exprs_results <- diff_exprs_results[complete.cases(diff_exprs_results),]
  
  #setting up pathview
  data(gene.idtype.bods)
  work.dir <- getwd()
  path <- paste0(loc_pathview_results, "oligo/pathview_treatment_and_age/")
  dir.create(path = path, recursive = T)
  setwd(path)
  
  #writing out the gage results
  write.table(x = gage_results_gfp$greater, 
                  file = "gage_results_GFP.tab", 
                  quote = F, 
                  sep = "\t", 
                  row.names = T)
  write.table(x = gage_results_il2$greater, 
                  file = "gage_results_IL2.tab", 
                  quote = F, 
                  sep = "\t", 
                  row.names = T)
  write.table(x = gage_results_04m$greater, 
                  file = "gage_results_04m.tab", 
                  quote = F, 
                  sep = "\t", 
                  row.names = T)
  write.table(x = gage_results_24m$greater, 
                  file = "gage_results_24m.tab", 
                  quote = F, 
                  sep = "\t", 
                  row.names = T)
  
  #extracting pathway ids
  pathway_ids <- 
    c(rownames(gage_results_gfp$greater), 
      rownames(gage_results_il2$greater), 
      rownames(gage_results_04m$greater), 
      rownames(gage_results_24m$greater))
  pathway_ids <- gsub(pattern = " .*$", 
                      replacement = "", 
                      x = pathway_ids, 
                      ignore.case = T)
  pathway_ids <- unique(pathway_ids)
  #adding missing pathways
  pathway_ids <- c(pathway_ids, "mmu05022", "mmu04370", "mmu04012", "mmu04014", 
                   "mmu04211", "mmu04657", "mmu01200", "mmu04310", "mmu04024", 
                   "mmu04213", "mmu04371", "mmu04145", "mmu04150", "mmu04015", 
                   "mmu04722", "mmu04068", "mmu04210", "mmu04218", "mmu04136")
  
  #running pathview in batches due to an undocumented limitation of a maximum
  # number of KEGG ids that can be queried in a single call. Not sure what value
  # and whether the limitation is in pathview or on KEGG's end
  pathway_ids <- split(pathway_ids, ceiling(seq_along(pathway_ids)/20))
  for(batch in pathway_ids){
    pathview(gene.data = diff_exprs_results, 
             cpd.data = NULL, 
             pathway.id = batch, 
             species = "mouse", 
             kegg.dir = path, 
             gene.idtype = "SYMBOL", 
             low = list(gene = "#2a7fff", cpd = "blue"), 
             mid = list(gene = "white", cpd = "gray"), 
             high = list(gene = "#ff5555", cpd = "yellow"), 
             limit = list(gene = 0.5, cpd = 1), 
             bins = list(gene = 10, cpd = 10), 
             kegg.native = T, 
             na.col = "gray", 
             same.layer = F)
    }
  
  #resetting work dir
  setwd(work.dir)
  
}
#creating the manual summarised comparison -- ref contrast -- figure 6
{
  gene_sets <- kegg_genesets$kg.sets
  
  for(i in seq(1, length(gene_sets))){
    gene_sets[[i]] <- genes$SYMBOL[which(genes$ENTREZID %in% gene_sets[[i]])]
  }
  
  #loading differential expression results
  {
    #loading GFP results
    diff_exprs_results_gfp <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "oligo/oligo_markers_list_gfp_nofilter.rds"))[[2]]
    diff_exprs_results_gfp <-
      diff_exprs_results_gfp[diff_exprs_results_gfp$p_val_adj < 0.1, "avg_log2FC",
                         drop = F]
    
    #loading IL2 results
    diff_exprs_results_il2 <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "oligo/oligo_markers_list_il2_nofilter.rds"))[[2]]
    diff_exprs_results_il2 <-
      diff_exprs_results_il2[diff_exprs_results_il2$p_val_adj < 0.1, "avg_log2FC",
                         drop = F]
    
    #loading 04m results
    diff_exprs_results_04m <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "oligo/oligo_markers_list_04m_nofilter.rds"))[[2]]
    diff_exprs_results_04m <-
      diff_exprs_results_04m[diff_exprs_results_04m$p_val_adj < 0.1, "avg_log2FC",
                         drop = F]
    
    #loading 24m results
    diff_exprs_results_24m <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "oligo/oligo_markers_list_24m_nofilter.rds"))[[2]]
    diff_exprs_results_24m <-
      diff_exprs_results_24m[diff_exprs_results_24m$p_val_adj < 0.1, "avg_log2FC",
                         drop = F]
    
    genes_to_select <- unique(c(rownames(diff_exprs_results_gfp), 
                                rownames(diff_exprs_results_il2), 
                                rownames(diff_exprs_results_04m), 
                                rownames(diff_exprs_results_24m)))
    
    #reloading differential expression results with genes significant in at 
    # least one contrast
    {
      #loading GFP results
      diff_exprs_results_gfp <- 
        readRDS(file = paste0(loc_diff_exprs_results, 
                              "oligo/oligo_markers_list_gfp_nofilter.rds"))[[2]]
      diff_exprs_results_gfp <-
        diff_exprs_results_gfp[genes_to_select, "avg_log2FC",
                           drop = F]
      
      #loading IL2 results
      diff_exprs_results_il2 <- 
        readRDS(file = paste0(loc_diff_exprs_results, 
                              "oligo/oligo_markers_list_il2_nofilter.rds"))[[2]]
      diff_exprs_results_il2 <-
        diff_exprs_results_il2[genes_to_select, "avg_log2FC",
                           drop = F]
      
      #loading 04m results
      diff_exprs_results_04m <- 
        readRDS(file = paste0(loc_diff_exprs_results, 
                              "oligo/oligo_markers_list_04m_nofilter.rds"))[[2]]
      diff_exprs_results_04m <-
        diff_exprs_results_04m[genes_to_select, "avg_log2FC",
                           drop = F]
      
      #loading 24m results
      diff_exprs_results_24m <- 
        readRDS(file = paste0(loc_diff_exprs_results, 
                              "oligo/oligo_markers_list_24m_nofilter.rds"))[[2]]
      diff_exprs_results_24m <-
        diff_exprs_results_24m[genes_to_select, "avg_log2FC",
                           drop = F]
    }
    
  }
  
  #formatting data frame of expression directions
  avg_sign_df <- data.frame(A = double(), B = double(), 
                            C = double(), D = double())
  for(i in seq(1, length(gene_sets))){
    #importing contrast data
    cont_ref <- diff_exprs_results_gfp[gene_sets[[i]],]
    cont_a <- diff_exprs_results_il2[gene_sets[[i]],]
    cont_b <- diff_exprs_results_04m[gene_sets[[i]],]
    cont_c <- diff_exprs_results_24m[gene_sets[[i]],]
    
    #checking which indices are negative in ref contrast
    flip_index <- which(cont_ref < 0)
    
    #if even a single negative value is found
    if(length(flip_index) > 0){
      cont_ref[flip_index] <- -(cont_ref[flip_index])
      cont_a[flip_index] <- -(cont_a[flip_index])
      cont_b[flip_index] <- -(cont_b[flip_index])
      cont_c[flip_index] <- -(cont_c[flip_index])
    }
    
    #formatting entry
    entry <- 
      c(mean(sign(cont_ref), na.rm = T), 
        mean(sign(cont_a), na.rm = T), 
        mean(sign(cont_b), na.rm = T), 
        mean(sign(cont_c), na.rm = T)
        )
    
    #if nothing was found in any contrast (no qualifying genes in any contrast
    # for a given pathway)
    if(all(is.na(entry))){
      next
    }
    
    #appending entry
    avg_sign_df <- rbind(avg_sign_df, entry)
    rownames(avg_sign_df)[nrow(avg_sign_df)] <- names(gene_sets)[i]
  }
  colnames(avg_sign_df) <- c("GFP 24m vs GFP 04m (ref)", "IL2 24m vs IL2 04m", 
                             "04m IL2 vs 04m GFP", "24m IL2 vs 24m GFP")
  
  #writing out full results
  write.table(x = avg_sign_df %>% rownames_to_column(var = "Pathway"), 
              file = paste0(loc_04mv24m_results, 
                            "/00_figures/Figure6/", 
                            "oligo_direction_data.tab"), 
              quote = F, sep = "\t", row.names = F, col.names = T)
  
  #formatting word wrapping
  rownames(avg_sign_df) <- str_wrap(string = rownames(avg_sign_df), width = 40)
  
  #reordering data frame for heatmap
  ordering <- rowMeans(avg_sign_df[complete.cases(avg_sign_df),], na.rm = T)
  ordering <- ordering[order(ordering, decreasing = T)]
  avg_sign_df <- 
    rbind(avg_sign_df[rownames(avg_sign_df) %in% names(ordering),], 
          avg_sign_df[!rownames(avg_sign_df) %in% names(ordering),])
  ordering <- rev(rownames(avg_sign_df))
  
  #importing pre-selected pathways
  selected_mmu <- c("mmu05010", "mmu04071", "mmu00190", "mmu03010", "mmu03040", 
                    "mmu03050", "mmu04137", "mmu04140", "mmu04141", "mmu04144", 
                    "mmu04810")
  avg_sign_df <- avg_sign_df[gsub(pattern = " .*$", 
                                  replacement = "", 
                                  x = rownames(avg_sign_df), 
                                  ignore.case = T) %in% selected_mmu,]
  
  #converting to heatmap
  avg_sign_df <- avg_sign_df %>% 
    rownames_to_column() %>% 
    gather(colname, value, -rowname)
  
  avg_sign_df$rowname <- factor(x = avg_sign_df$rowname, 
                                levels = ordering, ordered = T)
  avg_sign_df$colname <- factor(x = avg_sign_df$colname, 
                                levels = c("GFP 24m vs GFP 04m (ref)", 
                                           "IL2 24m vs IL2 04m", 
                                           "04m IL2 vs 04m GFP", 
                                           "24m IL2 vs 24m GFP"), 
                                ordered = T)
  
  #inverting values for colour scale in ggplot2
  avg_sign_df$value <- -(avg_sign_df$value)
  
  #plot
  ggplot(avg_sign_df, 
         aes(x = colname, 
             y = rowname, 
             fill = value)) + 
    geom_tile() + 
    scale_fill_gradient2(low = ("#3DBEDB"), #munsell::mnsl("5B 7/8")
                         mid = "white", 
                         high = ("#CAAE42"), #munsell::mnsl("5Y 7/8") 
                         breaks = c(-1, -0.5, 0, 0.5, 1), 
                         labels = c("Closer to\naged control", "", "", "", 
                                    "Closer to\nyoung control"), 
                         limits = c(-1, 1)) + 
    xlab(NULL) + ylab(NULL) + labs(fill = NULL) + 
    theme_light() + 
    theme(axis.text.x = element_text(size = 13, angle = 25, 
                                     hjust = 0.9, vjust = 0.9), 
          text = element_text(size = 16), 
          axis.text.y =  element_text(size = 14))
  
}


#running analyses for astrocyte all and subpops old vs young 
{
  # sort(as.character(
  #     unique(astro_data$seurat_clusters_renamed)))
  # [1] "Bergmann Glia"                 "Cerebellar\nAstrocytes"        "Non-telencephalon\nAstrocytes"
  # [4] "Olfactory\nAstrocytes"         "Striatal\nAstrocytes"          "Telencephalon\nAstrocytes"  
  
  #loading differential expression results for each subpop and selecting genes
  subpops <- paste0("astro/subpops/", 
                    c("Bergmann.Glia", "Cerebellar.Astrocytes", 
                      "Non-telencephalon.Astrocytes", "Olfactory.Astrocytes", 
                      "Striatal.Astrocytes", "Telencephalon.Astrocytes", 
                      "All.Astrocytes"))
  #initialising genes_to_select for the loop to get a unified list for all astro
  genes_to_select <- character()
  for(subpop in subpops){
    #loading differential expression results
    #loading GFP results
    diff_exprs_results_gfp <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            subpop, 
                            "_markers_list_gfp_nofilter.rds"))[[2]]
    diff_exprs_results_gfp <-
      diff_exprs_results_gfp[diff_exprs_results_gfp$p_val_adj < 0.1, "avg_log2FC",
                         drop = F]
    
    #loading IL2 results
    diff_exprs_results_il2 <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            subpop, 
                            "_markers_list_il2_nofilter.rds"))[[2]]
    diff_exprs_results_il2 <-
      diff_exprs_results_il2[diff_exprs_results_il2$p_val_adj < 0.1, "avg_log2FC",
                         drop = F]
    
    #loading 04m results
    diff_exprs_results_04m <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            subpop, 
                            "_markers_list_04m_nofilter.rds"))[[2]]
    diff_exprs_results_04m <-
      diff_exprs_results_04m[diff_exprs_results_04m$p_val_adj < 0.1, "avg_log2FC",
                         drop = F]
    
    #loading 24m results
    diff_exprs_results_24m <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            subpop, 
                            "_markers_list_24m_nofilter.rds"))[[2]]
    diff_exprs_results_24m <-
      diff_exprs_results_24m[diff_exprs_results_24m$p_val_adj < 0.1, "avg_log2FC",
                         drop = F] 
    
    genes_to_select <- unique(c(genes_to_select, 
                                rownames(diff_exprs_results_gfp), 
                                rownames(diff_exprs_results_il2), 
                                rownames(diff_exprs_results_04m), 
                                rownames(diff_exprs_results_24m)))
  }
  genes_to_select <- unique(genes_to_select)
  
  ## all astrocytes
  astro_data <- readRDS(file = paste0(loc_superclusters,
                                      "astro/astro_data.rds"))
  #running gage analysis for old vs young in GFP
  {
    #loading metadata
    metadata <- astro_data@meta.data
    
    #getting cell names and finding column indices
    young_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "04m.PHP.GFAP-GFP"))
    old_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "24m.PHP.GFAP-GFP"))
    young_cells <- which(colnames(astro_data[["SCT"]]@data) %in% young_cells)
    old_cells <- which(colnames(astro_data[["SCT"]]@data) %in% old_cells)
    
    #extracting and formatting log counts matrix to minimise ram footprint
    counts_matrix <- astro_data[["SCT"]]@data[,c(young_cells, old_cells)]
    counts_matrix <- as.data.frame(counts_matrix)
    gc()
    
    #mapping ids
    # do check that the order is the same
    # all(unique(rownames(counts_matrix) == genes$SYMBOL))
    counts_matrix$ENTREZID <- genes$ENTREZID
    counts_matrix <- counts_matrix[complete.cases(counts_matrix),]
    rownames(counts_matrix) <- counts_matrix$ENTREZID
    counts_matrix$ENTREZID <- NULL
    
    #running gage
    gage_results <- gage(exprs = counts_matrix, 
                         gsets = kegg_genesets$kg.sets, 
                         ref = (1:length(young_cells)), 
                         samp = ((length(young_cells)+1) : ncol(counts_matrix)), 
                         same.dir = F, 
                         compare = "as.group", 
                         set.size = c(15, 1000)
                         )
    gc()
    
    saveRDS(object = gage_results, 
            file = paste0(loc_pathview_results, "astro/gage_results_all_GFP.rds"), 
            compress = T)
    
  }
  #running gage analysis for old vs young in IL2
  {
    #loading metadata
    metadata <- astro_data@meta.data
    
    #getting cell names and finding column indices
    young_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "04m.PHP.GFAP-IL2"))
    old_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "24m.PHP.GFAP-IL2"))
    young_cells <- which(colnames(astro_data[["SCT"]]@data) %in% young_cells)
    old_cells <- which(colnames(astro_data[["SCT"]]@data) %in% old_cells)
    
    #extracting and formatting log counts matrix to minimise ram footprint
    counts_matrix <- astro_data[["SCT"]]@data[,c(young_cells, old_cells)]
    counts_matrix <- as.data.frame(counts_matrix)
    gc()
    
    #mapping ids
    # do check that the order is the same
    # all(unique(rownames(counts_matrix) == genes$SYMBOL))
    counts_matrix$ENTREZID <- genes$ENTREZID
    counts_matrix <- counts_matrix[complete.cases(counts_matrix),]
    rownames(counts_matrix) <- counts_matrix$ENTREZID
    counts_matrix$ENTREZID <- NULL
    
    #running gage
    gage_results <- gage(exprs = counts_matrix, 
                         gsets = kegg_genesets$kg.sets, 
                         ref = (1:length(young_cells)), 
                         samp = ((length(young_cells)+1) : ncol(counts_matrix)), 
                         same.dir = F, 
                         compare = "as.group", 
                         set.size = c(15, 1000)
                         )
    gc()
    
    saveRDS(object = gage_results, 
            file = paste0(loc_pathview_results, "astro/gage_results_all_IL2.rds"), 
            compress = T)
    
  }
  #running gage analysis for IL2 vs GFP in young
  {
    #loading metadata
    metadata <- astro_data@meta.data
    
    #getting cell names and finding column indices
    young_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "04m.PHP.GFAP-GFP"))
    old_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "04m.PHP.GFAP-IL2"))
    young_cells <- which(colnames(astro_data[["SCT"]]@data) %in% young_cells)
    old_cells <- which(colnames(astro_data[["SCT"]]@data) %in% old_cells)
    
    #extracting and formatting log counts matrix to minimise ram footprint
    counts_matrix <- astro_data[["SCT"]]@data[,c(young_cells, old_cells)]
    counts_matrix <- as.data.frame(counts_matrix)
    gc()
    
    #mapping ids
    # do check that the order is the same
    # all(unique(rownames(counts_matrix) == genes$SYMBOL))
    counts_matrix$ENTREZID <- genes$ENTREZID
    counts_matrix <- counts_matrix[complete.cases(counts_matrix),]
    rownames(counts_matrix) <- counts_matrix$ENTREZID
    counts_matrix$ENTREZID <- NULL
    
    #running gage
    gage_results <- gage(exprs = counts_matrix, 
                         gsets = kegg_genesets$kg.sets, 
                         ref = (1:length(young_cells)), 
                         samp = ((length(young_cells)+1) : ncol(counts_matrix)), 
                         same.dir = F, 
                         compare = "as.group", 
                         set.size = c(15, 1000)
                         )
    gc()
    
    saveRDS(object = gage_results, 
            file = paste0(loc_pathview_results, "astro/gage_results_all_04m.rds"), 
            compress = T)
    
  }
  #running gage analysis for IL2 vs GFP in old
  {
    #loading metadata
    metadata <- astro_data@meta.data
    
    #getting cell names and finding column indices
    young_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "24m.PHP.GFAP-GFP"))
    old_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "24m.PHP.GFAP-IL2"))
    young_cells <- which(colnames(astro_data[["SCT"]]@data) %in% young_cells)
    old_cells <- which(colnames(astro_data[["SCT"]]@data) %in% old_cells)
    
    #extracting and formatting log counts matrix to minimise ram footprint
    counts_matrix <- astro_data[["SCT"]]@data[,c(young_cells, old_cells)]
    counts_matrix <- as.data.frame(counts_matrix)
    gc()
    
    #mapping ids
    # do check that the order is the same
    # all(unique(rownames(counts_matrix) == genes$SYMBOL))
    counts_matrix$ENTREZID <- genes$ENTREZID
    counts_matrix <- counts_matrix[complete.cases(counts_matrix),]
    rownames(counts_matrix) <- counts_matrix$ENTREZID
    counts_matrix$ENTREZID <- NULL
    
    #running gage
    gage_results <- gage(exprs = counts_matrix, 
                         gsets = kegg_genesets$kg.sets, 
                         ref = (1:length(young_cells)), 
                         samp = ((length(young_cells)+1) : ncol(counts_matrix)), 
                         same.dir = F, 
                         compare = "as.group", 
                         set.size = c(15, 1000)
                         )
    gc()
    
    saveRDS(object = gage_results, 
            file = paste0(loc_pathview_results, "astro/gage_results_all_24m.rds"), 
            compress = T)
    
  }
  #running pathview analysis on treatment results
  {
    #loading GFP results
    gage_results_gfp <- readRDS(file = paste0(loc_pathview_results, 
                                              "astro/gage_results_all_GFP.rds"))
    diff_exprs_results_gfp <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "astro/subpops/All.Astrocytes_markers_list_gfp_nofilter.rds"))[[2]]
    diff_exprs_results_gfp <- 
      diff_exprs_results_gfp[genes_to_select, "avg_log2FC", 
                         drop = F]
    
    #loading IL2 results
    gage_results_il2 <- readRDS(file = paste0(loc_pathview_results, 
                                              "astro/gage_results_all_IL2.rds"))
    diff_exprs_results_il2 <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "astro/subpops/All.Astrocytes_markers_list_il2_nofilter.rds"))[[2]]
    diff_exprs_results_il2 <- 
      diff_exprs_results_il2[genes_to_select, "avg_log2FC", 
                         drop = F]
    
    #loading young results
    gage_results_04m <- readRDS(file = paste0(loc_pathview_results, 
                                              "astro/gage_results_all_04m.rds"))
    diff_exprs_results_04m <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "astro/subpops/All.Astrocytes_markers_list_04m_nofilter.rds"))[[2]]
    diff_exprs_results_04m <- 
      diff_exprs_results_04m[genes_to_select, "avg_log2FC", 
                         drop = F]
    
    #loading old results
    gage_results_24m <- readRDS(file = paste0(loc_pathview_results, 
                                              "astro/gage_results_all_24m.rds"))
    diff_exprs_results_24m <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "astro/subpops/All.Astrocytes_markers_list_24m_nofilter.rds"))[[2]]
    diff_exprs_results_24m <- 
      diff_exprs_results_24m[genes_to_select, "avg_log2FC", 
                         drop = F]
    
    #merging diff exprs data
    diff_exprs_results <- merge(x = diff_exprs_results_gfp, 
                                y = diff_exprs_results_il2, 
                                by = "row.names", 
                                all = T)
    diff_exprs_results_2 <- merge(x = diff_exprs_results_04m, 
                                y = diff_exprs_results_24m, 
                                by = "row.names", 
                                all = T)
    diff_exprs_results <- merge(x = diff_exprs_results, 
                                y = diff_exprs_results_2, 
                                by = "Row.names", 
                                all = T)
    rownames(diff_exprs_results) <- diff_exprs_results$Row.names
    diff_exprs_results$Row.names <- NULL
    diff_exprs_results_2 <- NULL
    colnames(diff_exprs_results) <- paste0("Contrast ", 
                                           1:ncol(diff_exprs_results))
    
    #to exclude those extreme cases where a gene was not detected in a contrast
    diff_exprs_results <- diff_exprs_results[complete.cases(diff_exprs_results),]
    
    #setting up pathview
    data(gene.idtype.bods)
    work.dir <- getwd()
    path <- paste0(loc_pathview_results, "astro/all_pathview_treatment/")
    dir.create(path = path, recursive = T)
    setwd(path)
    
    #writing out the gage results
    write.table(x = gage_results_gfp$greater, 
                    file = "gage_results_GFP.tab", 
                    quote = F, 
                    sep = "\t", 
                    row.names = T)
    write.table(x = gage_results_il2$greater, 
                    file = "gage_results_IL2.tab", 
                    quote = F, 
                    sep = "\t", 
                    row.names = T)
    write.table(x = gage_results_04m$greater, 
                    file = "gage_results_04m.tab", 
                    quote = F, 
                    sep = "\t", 
                    row.names = T)
    write.table(x = gage_results_24m$greater, 
                    file = "gage_results_24m.tab", 
                    quote = F, 
                    sep = "\t", 
                    row.names = T)
    
    #getting pathway ids from the gage results
    pathway_ids <- 
      c(rownames(gage_results_gfp$greater), 
        rownames(gage_results_il2$greater), 
        rownames(gage_results_04m$greater), 
        rownames(gage_results_24m$greater))
    pathway_ids <- gsub(pattern = " .*$", 
                        replacement = "", 
                        x = pathway_ids, 
                        ignore.case = T)
    pathway_ids <- unique(pathway_ids)
    #adding missing pathways
    pathway_ids <- c(pathway_ids, "mmu05022", "mmu04370", "mmu04012", "mmu04014", 
                     "mmu04211", "mmu04657", "mmu01200", "mmu04310", "mmu04024", 
                     "mmu04213", "mmu04371", "mmu04145", "mmu04150", "mmu04015", 
                     "mmu04722", "mmu04068", "mmu04210", "mmu04218", "mmu04136")
    
    #running pathview in batches due to an undocumented limitation of a maximum
    # number of KEGG ids that can be queried in a single call. Not sure what value
    # and whether the limitation is in pathview or on KEGG's end
    pathway_ids <- split(pathway_ids, ceiling(seq_along(pathway_ids)/20))
    for(batch in pathway_ids){
      pathview(gene.data = diff_exprs_results, 
               cpd.data = NULL, 
               pathway.id = batch, 
               species = "mouse", 
               kegg.dir = path, 
               gene.idtype = "SYMBOL", 
               low = list(gene = "#2a7fff", cpd = "blue"), 
               mid = list(gene = "white", cpd = "gray"), 
               high = list(gene = "#ff5555", cpd = "yellow"), 
               limit = list(gene = 0.5, cpd = 1), 
               bins = list(gene = 10, cpd = 10), 
               kegg.native = T, 
               na.col = "gray", 
               same.layer = F)
      }
    
    #resetting work dir
    setwd(work.dir)
    
  }
  
  
  
  ## non-telencephalon
  astro_data <- readRDS(file = paste0(loc_superclusters,
                                      "astro/astro_data.rds"))
  astro_data <- 
    subset(astro_data, 
           seurat_clusters_renamed == "Non-telencephalon\nAstrocytes")
  #running gage analysis for old vs young in GFP
  {
    #loading metadata
    metadata <- astro_data@meta.data
    
    #getting cell names and finding column indices
    young_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "04m.PHP.GFAP-GFP"))
    old_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "24m.PHP.GFAP-GFP"))
    young_cells <- which(colnames(astro_data[["SCT"]]@data) %in% young_cells)
    old_cells <- which(colnames(astro_data[["SCT"]]@data) %in% old_cells)
    
    #extracting and formatting log counts matrix to minimise ram footprint
    counts_matrix <- astro_data[["SCT"]]@data[,c(young_cells, old_cells)]
    counts_matrix <- as.data.frame(counts_matrix)
    gc()
    
    #mapping ids
    # do check that the order is the same
    # all(unique(rownames(counts_matrix) == genes$SYMBOL))
    counts_matrix$ENTREZID <- genes$ENTREZID
    counts_matrix <- counts_matrix[complete.cases(counts_matrix),]
    rownames(counts_matrix) <- counts_matrix$ENTREZID
    counts_matrix$ENTREZID <- NULL
    
    #running gage
    gage_results <- gage(exprs = counts_matrix, 
                         gsets = kegg_genesets$kg.sets, 
                         ref = (1:length(young_cells)), 
                         samp = ((length(young_cells)+1) : ncol(counts_matrix)), 
                         same.dir = F, 
                         compare = "as.group", 
                         set.size = c(15, 1000)
                         )
    gc()
    
    saveRDS(object = gage_results, 
            file = paste0(loc_pathview_results, "astro/gage_results_non-telencephalon_GFP.rds"), 
            compress = T)
    
  }
  #running gage analysis for old vs young in IL2
  {
    #loading metadata
    metadata <- astro_data@meta.data
    
    #getting cell names and finding column indices
    young_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "04m.PHP.GFAP-IL2"))
    old_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "24m.PHP.GFAP-IL2"))
    young_cells <- which(colnames(astro_data[["SCT"]]@data) %in% young_cells)
    old_cells <- which(colnames(astro_data[["SCT"]]@data) %in% old_cells)
    
    #extracting and formatting log counts matrix to minimise ram footprint
    counts_matrix <- astro_data[["SCT"]]@data[,c(young_cells, old_cells)]
    counts_matrix <- as.data.frame(counts_matrix)
    gc()
    
    #mapping ids
    # do check that the order is the same
    # all(unique(rownames(counts_matrix) == genes$SYMBOL))
    counts_matrix$ENTREZID <- genes$ENTREZID
    counts_matrix <- counts_matrix[complete.cases(counts_matrix),]
    rownames(counts_matrix) <- counts_matrix$ENTREZID
    counts_matrix$ENTREZID <- NULL
    
    #running gage
    gage_results <- gage(exprs = counts_matrix, 
                         gsets = kegg_genesets$kg.sets, 
                         ref = (1:length(young_cells)), 
                         samp = ((length(young_cells)+1) : ncol(counts_matrix)), 
                         same.dir = F, 
                         compare = "as.group", 
                         set.size = c(15, 1000)
                         )
    gc()
    
    saveRDS(object = gage_results, 
            file = paste0(loc_pathview_results, "astro/gage_results_non-telencephalon_IL2.rds"), 
            compress = T)
    
  }
  #running gage analysis for IL2 vs GFP in young
  {
    #loading metadata
    metadata <- astro_data@meta.data
    
    #getting cell names and finding column indices
    young_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "04m.PHP.GFAP-GFP"))
    old_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "04m.PHP.GFAP-IL2"))
    young_cells <- which(colnames(astro_data[["SCT"]]@data) %in% young_cells)
    old_cells <- which(colnames(astro_data[["SCT"]]@data) %in% old_cells)
    
    #extracting and formatting log counts matrix to minimise ram footprint
    counts_matrix <- astro_data[["SCT"]]@data[,c(young_cells, old_cells)]
    counts_matrix <- as.data.frame(counts_matrix)
    gc()
    
    #mapping ids
    # do check that the order is the same
    # all(unique(rownames(counts_matrix) == genes$SYMBOL))
    counts_matrix$ENTREZID <- genes$ENTREZID
    counts_matrix <- counts_matrix[complete.cases(counts_matrix),]
    rownames(counts_matrix) <- counts_matrix$ENTREZID
    counts_matrix$ENTREZID <- NULL
    
    #running gage
    gage_results <- gage(exprs = counts_matrix, 
                         gsets = kegg_genesets$kg.sets, 
                         ref = (1:length(young_cells)), 
                         samp = ((length(young_cells)+1) : ncol(counts_matrix)), 
                         same.dir = F, 
                         compare = "as.group", 
                         set.size = c(15, 1000)
                         )
    gc()
    
    saveRDS(object = gage_results, 
            file = paste0(loc_pathview_results, "astro/gage_results_non-telencephalon_04m.rds"), 
            compress = T)
    
  }
  #running gage analysis for IL2 vs GFP in old
  {
    #loading metadata
    metadata <- astro_data@meta.data
    
    #getting cell names and finding column indices
    young_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "24m.PHP.GFAP-GFP"))
    old_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "24m.PHP.GFAP-IL2"))
    young_cells <- which(colnames(astro_data[["SCT"]]@data) %in% young_cells)
    old_cells <- which(colnames(astro_data[["SCT"]]@data) %in% old_cells)
    
    #extracting and formatting log counts matrix to minimise ram footprint
    counts_matrix <- astro_data[["SCT"]]@data[,c(young_cells, old_cells)]
    counts_matrix <- as.data.frame(counts_matrix)
    gc()
    
    #mapping ids
    # do check that the order is the same
    # all(unique(rownames(counts_matrix) == genes$SYMBOL))
    counts_matrix$ENTREZID <- genes$ENTREZID
    counts_matrix <- counts_matrix[complete.cases(counts_matrix),]
    rownames(counts_matrix) <- counts_matrix$ENTREZID
    counts_matrix$ENTREZID <- NULL
    
    #running gage
    gage_results <- gage(exprs = counts_matrix, 
                         gsets = kegg_genesets$kg.sets, 
                         ref = (1:length(young_cells)), 
                         samp = ((length(young_cells)+1) : ncol(counts_matrix)), 
                         same.dir = F, 
                         compare = "as.group", 
                         set.size = c(15, 1000)
                         )
    gc()
    
    saveRDS(object = gage_results, 
            file = paste0(loc_pathview_results, "astro/gage_results_non-telencephalon_24m.rds"), 
            compress = T)
    
  }
  #running pathview analysis on treatment results
  {
    #loading GFP results
    gage_results_gfp <- readRDS(file = paste0(loc_pathview_results, 
                                              "astro/gage_results_non-telencephalon_GFP.rds"))
    diff_exprs_results_gfp <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "astro/subpops/Non-telencephalon.Astrocytes_markers_list_gfp_nofilter.rds"))[[2]]
    diff_exprs_results_gfp <- 
      diff_exprs_results_gfp[genes_to_select, "avg_log2FC", 
                         drop = F]
    
    #loading IL2 results
    gage_results_il2 <- readRDS(file = paste0(loc_pathview_results, 
                                              "astro/gage_results_non-telencephalon_IL2.rds"))
    diff_exprs_results_il2 <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "astro/subpops/Non-telencephalon.Astrocytes_markers_list_il2_nofilter.rds"))[[2]]
    diff_exprs_results_il2 <- 
      diff_exprs_results_il2[genes_to_select, "avg_log2FC", 
                         drop = F]
    
    #loading young results
    gage_results_04m <- readRDS(file = paste0(loc_pathview_results, 
                                              "astro/gage_results_non-telencephalon_04m.rds"))
    diff_exprs_results_04m <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "astro/subpops/Non-telencephalon.Astrocytes_markers_list_04m_nofilter.rds"))[[2]]
    diff_exprs_results_04m <- 
      diff_exprs_results_04m[genes_to_select, "avg_log2FC", 
                         drop = F]
    
    #loading old results
    gage_results_24m <- readRDS(file = paste0(loc_pathview_results, 
                                              "astro/gage_results_non-telencephalon_24m.rds"))
    diff_exprs_results_24m <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "astro/subpops/Non-telencephalon.Astrocytes_markers_list_24m_nofilter.rds"))[[2]]
    diff_exprs_results_24m <- 
      diff_exprs_results_24m[genes_to_select, "avg_log2FC", 
                         drop = F]
    
    #merging diff exprs data
    diff_exprs_results <- merge(x = diff_exprs_results_gfp, 
                                y = diff_exprs_results_il2, 
                                by = "row.names", 
                                all = T)
    diff_exprs_results_2 <- merge(x = diff_exprs_results_04m, 
                                y = diff_exprs_results_24m, 
                                by = "row.names", 
                                all = T)
    diff_exprs_results <- merge(x = diff_exprs_results, 
                                y = diff_exprs_results_2, 
                                by = "Row.names", 
                                all = T)
    rownames(diff_exprs_results) <- diff_exprs_results$Row.names
    diff_exprs_results$Row.names <- NULL
    diff_exprs_results_2 <- NULL
    colnames(diff_exprs_results) <- paste0("Contrast ", 
                                           1:ncol(diff_exprs_results))
    
    #to exclude those extreme cases where a gene was not detected in a contrast
    diff_exprs_results <- diff_exprs_results[complete.cases(diff_exprs_results),]
    
    #setting up pathview
    data(gene.idtype.bods)
    work.dir <- getwd()
    path <- paste0(loc_pathview_results, "astro/non-telencephalon_pathview_treatment/")
    dir.create(path = path, recursive = T)
    setwd(path)
    
    #writing out the gage results
    write.table(x = gage_results_gfp$greater, 
                    file = "gage_results_GFP.tab", 
                    quote = F, 
                    sep = "\t", 
                    row.names = T)
    write.table(x = gage_results_il2$greater, 
                    file = "gage_results_IL2.tab", 
                    quote = F, 
                    sep = "\t", 
                    row.names = T)
    write.table(x = gage_results_04m$greater, 
                    file = "gage_results_04m.tab", 
                    quote = F, 
                    sep = "\t", 
                    row.names = T)
    write.table(x = gage_results_24m$greater, 
                    file = "gage_results_24m.tab", 
                    quote = F, 
                    sep = "\t", 
                    row.names = T)
    
    #getting pathway ids from the gage results
    pathway_ids <- 
      c(rownames(gage_results_gfp$greater), 
        rownames(gage_results_il2$greater), 
        rownames(gage_results_04m$greater), 
        rownames(gage_results_24m$greater))
    pathway_ids <- gsub(pattern = " .*$", 
                        replacement = "", 
                        x = pathway_ids, 
                        ignore.case = T)
    pathway_ids <- unique(pathway_ids)
    #adding missing pathways
    pathway_ids <- c(pathway_ids, "mmu05022", "mmu04370", "mmu04012", "mmu04014", 
                     "mmu04211", "mmu04657", "mmu01200", "mmu04310", "mmu04024", 
                     "mmu04213", "mmu04371", "mmu04145", "mmu04150", "mmu04015", 
                     "mmu04722", "mmu04068", "mmu04210", "mmu04218", "mmu04136")
    
    #running pathview in batches due to an undocumented limitation of a maximum
    # number of KEGG ids that can be queried in a single call. Not sure what value
    # and whether the limitation is in pathview or on KEGG's end
    pathway_ids <- split(pathway_ids, ceiling(seq_along(pathway_ids)/20))
    for(batch in pathway_ids){
      pathview(gene.data = diff_exprs_results, 
               cpd.data = NULL, 
               pathway.id = batch, 
               species = "mouse", 
               kegg.dir = path, 
               gene.idtype = "SYMBOL", 
               low = list(gene = "#2a7fff", cpd = "blue"), 
               mid = list(gene = "white", cpd = "gray"), 
               high = list(gene = "#ff5555", cpd = "yellow"), 
               limit = list(gene = 0.5, cpd = 1), 
               bins = list(gene = 10, cpd = 10), 
               kegg.native = T, 
               na.col = "gray", 
               same.layer = F)
      }
    
    #resetting work dir
    setwd(work.dir)
    
  }
  
  
  ## telencephalon
  astro_data <- readRDS(file = paste0(loc_superclusters,
                                      "astro/astro_data.rds"))
  astro_data <- 
    subset(astro_data, 
           seurat_clusters_renamed == "Telencephalon\nAstrocytes")
  #running gage analysis for old vs young in GFP
  {
    #loading metadata
    metadata <- astro_data@meta.data
    
    #getting cell names and finding column indices
    young_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "04m.PHP.GFAP-GFP"))
    old_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "24m.PHP.GFAP-GFP"))
    young_cells <- which(colnames(astro_data[["SCT"]]@data) %in% young_cells)
    old_cells <- which(colnames(astro_data[["SCT"]]@data) %in% old_cells)
    
    #extracting and formatting log counts matrix to minimise ram footprint
    counts_matrix <- astro_data[["SCT"]]@data[,c(young_cells, old_cells)]
    counts_matrix <- as.data.frame(counts_matrix)
    gc()
    
    #mapping ids
    # do check that the order is the same
    # all(unique(rownames(counts_matrix) == genes$SYMBOL))
    counts_matrix$ENTREZID <- genes$ENTREZID
    counts_matrix <- counts_matrix[complete.cases(counts_matrix),]
    rownames(counts_matrix) <- counts_matrix$ENTREZID
    counts_matrix$ENTREZID <- NULL
    
    #running gage
    gage_results <- gage(exprs = counts_matrix, 
                         gsets = kegg_genesets$kg.sets, 
                         ref = (1:length(young_cells)), 
                         samp = ((length(young_cells)+1) : ncol(counts_matrix)), 
                         same.dir = F, 
                         compare = "as.group", 
                         set.size = c(15, 1000)
                         )
    gc()
    
    saveRDS(object = gage_results, 
            file = paste0(loc_pathview_results, "astro/gage_results_telencephalon_GFP.rds"), 
            compress = T)
    
  }
  #running gage analysis for old vs young in IL2
  {
    #loading metadata
    metadata <- astro_data@meta.data
    
    #getting cell names and finding column indices
    young_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "04m.PHP.GFAP-IL2"))
    old_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "24m.PHP.GFAP-IL2"))
    young_cells <- which(colnames(astro_data[["SCT"]]@data) %in% young_cells)
    old_cells <- which(colnames(astro_data[["SCT"]]@data) %in% old_cells)
    
    #extracting and formatting log counts matrix to minimise ram footprint
    counts_matrix <- astro_data[["SCT"]]@data[,c(young_cells, old_cells)]
    counts_matrix <- as.data.frame(counts_matrix)
    gc()
    
    #mapping ids
    # do check that the order is the same
    # all(unique(rownames(counts_matrix) == genes$SYMBOL))
    counts_matrix$ENTREZID <- genes$ENTREZID
    counts_matrix <- counts_matrix[complete.cases(counts_matrix),]
    rownames(counts_matrix) <- counts_matrix$ENTREZID
    counts_matrix$ENTREZID <- NULL
    
    #running gage
    gage_results <- gage(exprs = counts_matrix, 
                         gsets = kegg_genesets$kg.sets, 
                         ref = (1:length(young_cells)), 
                         samp = ((length(young_cells)+1) : ncol(counts_matrix)), 
                         same.dir = F, 
                         compare = "as.group", 
                         set.size = c(15, 1000)
                         )
    gc()
    
    saveRDS(object = gage_results, 
            file = paste0(loc_pathview_results, "astro/gage_results_telencephalon_IL2.rds"), 
            compress = T)
    
  }
  #running gage analysis for IL2 vs GFP in young
  {
    #loading metadata
    metadata <- astro_data@meta.data
    
    #getting cell names and finding column indices
    young_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "04m.PHP.GFAP-GFP"))
    old_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "04m.PHP.GFAP-IL2"))
    young_cells <- which(colnames(astro_data[["SCT"]]@data) %in% young_cells)
    old_cells <- which(colnames(astro_data[["SCT"]]@data) %in% old_cells)
    
    #extracting and formatting log counts matrix to minimise ram footprint
    counts_matrix <- astro_data[["SCT"]]@data[,c(young_cells, old_cells)]
    counts_matrix <- as.data.frame(counts_matrix)
    gc()
    
    #mapping ids
    # do check that the order is the same
    # all(unique(rownames(counts_matrix) == genes$SYMBOL))
    counts_matrix$ENTREZID <- genes$ENTREZID
    counts_matrix <- counts_matrix[complete.cases(counts_matrix),]
    rownames(counts_matrix) <- counts_matrix$ENTREZID
    counts_matrix$ENTREZID <- NULL
    
    #running gage
    gage_results <- gage(exprs = counts_matrix, 
                         gsets = kegg_genesets$kg.sets, 
                         ref = (1:length(young_cells)), 
                         samp = ((length(young_cells)+1) : ncol(counts_matrix)), 
                         same.dir = F, 
                         compare = "as.group", 
                         set.size = c(15, 1000)
                         )
    gc()
    
    saveRDS(object = gage_results, 
            file = paste0(loc_pathview_results, "astro/gage_results_telencephalon_04m.rds"), 
            compress = T)
    
  }
  #running gage analysis for IL2 vs GFP in old
  {
    #loading metadata
    metadata <- astro_data@meta.data
    
    #getting cell names and finding column indices
    young_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "24m.PHP.GFAP-GFP"))
    old_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "24m.PHP.GFAP-IL2"))
    young_cells <- which(colnames(astro_data[["SCT"]]@data) %in% young_cells)
    old_cells <- which(colnames(astro_data[["SCT"]]@data) %in% old_cells)
    
    #extracting and formatting log counts matrix to minimise ram footprint
    counts_matrix <- astro_data[["SCT"]]@data[,c(young_cells, old_cells)]
    counts_matrix <- as.data.frame(counts_matrix)
    gc()
    
    #mapping ids
    # do check that the order is the same
    # all(unique(rownames(counts_matrix) == genes$SYMBOL))
    counts_matrix$ENTREZID <- genes$ENTREZID
    counts_matrix <- counts_matrix[complete.cases(counts_matrix),]
    rownames(counts_matrix) <- counts_matrix$ENTREZID
    counts_matrix$ENTREZID <- NULL
    
    #running gage
    gage_results <- gage(exprs = counts_matrix, 
                         gsets = kegg_genesets$kg.sets, 
                         ref = (1:length(young_cells)), 
                         samp = ((length(young_cells)+1) : ncol(counts_matrix)), 
                         same.dir = F, 
                         compare = "as.group", 
                         set.size = c(15, 1000)
                         )
    gc()
    
    saveRDS(object = gage_results, 
            file = paste0(loc_pathview_results, "astro/gage_results_telencephalon_24m.rds"), 
            compress = T)
    
  }
  #running pathview analysis on treatment results
  {
    #loading GFP results
    gage_results_gfp <- readRDS(file = paste0(loc_pathview_results, 
                                              "astro/gage_results_telencephalon_GFP.rds"))
    diff_exprs_results_gfp <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "astro/subpops/Telencephalon.Astrocytes_markers_list_gfp_nofilter.rds"))[[2]]
    diff_exprs_results_gfp <- 
      diff_exprs_results_gfp[genes_to_select, "avg_log2FC", 
                         drop = F]
    
    #loading IL2 results
    gage_results_il2 <- readRDS(file = paste0(loc_pathview_results, 
                                              "astro/gage_results_telencephalon_IL2.rds"))
    diff_exprs_results_il2 <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "astro/subpops/Telencephalon.Astrocytes_markers_list_il2_nofilter.rds"))[[2]]
    diff_exprs_results_il2 <- 
      diff_exprs_results_il2[genes_to_select, "avg_log2FC", 
                         drop = F]
    
    #loading young results
    gage_results_04m <- readRDS(file = paste0(loc_pathview_results, 
                                              "astro/gage_results_telencephalon_04m.rds"))
    diff_exprs_results_04m <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "astro/subpops/Telencephalon.Astrocytes_markers_list_04m_nofilter.rds"))[[2]]
    diff_exprs_results_04m <- 
      diff_exprs_results_04m[genes_to_select, "avg_log2FC", 
                         drop = F]
    
    #loading old results
    gage_results_24m <- readRDS(file = paste0(loc_pathview_results, 
                                              "astro/gage_results_telencephalon_24m.rds"))
    diff_exprs_results_24m <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "astro/subpops/Telencephalon.Astrocytes_markers_list_24m_nofilter.rds"))[[2]]
    diff_exprs_results_24m <- 
      diff_exprs_results_24m[genes_to_select, "avg_log2FC", 
                         drop = F]
    
    #merging diff exprs data
    diff_exprs_results <- merge(x = diff_exprs_results_gfp, 
                                y = diff_exprs_results_il2, 
                                by = "row.names", 
                                all = T)
    diff_exprs_results_2 <- merge(x = diff_exprs_results_04m, 
                                y = diff_exprs_results_24m, 
                                by = "row.names", 
                                all = T)
    diff_exprs_results <- merge(x = diff_exprs_results, 
                                y = diff_exprs_results_2, 
                                by = "Row.names", 
                                all = T)
    rownames(diff_exprs_results) <- diff_exprs_results$Row.names
    diff_exprs_results$Row.names <- NULL
    diff_exprs_results_2 <- NULL
    colnames(diff_exprs_results) <- paste0("Contrast ", 
                                           1:ncol(diff_exprs_results))
    
    #to exclude those extreme cases where a gene was not detected in a contrast
    diff_exprs_results <- diff_exprs_results[complete.cases(diff_exprs_results),]
    
    #setting up pathview
    data(gene.idtype.bods)
    work.dir <- getwd()
    path <- paste0(loc_pathview_results, "astro/telencephalon_pathview_treatment/")
    dir.create(path = path, recursive = T)
    setwd(path)
    
    #writing out the gage results
    write.table(x = gage_results_gfp$greater, 
                    file = "gage_results_GFP.tab", 
                    quote = F, 
                    sep = "\t", 
                    row.names = T)
    write.table(x = gage_results_il2$greater, 
                    file = "gage_results_IL2.tab", 
                    quote = F, 
                    sep = "\t", 
                    row.names = T)
    write.table(x = gage_results_04m$greater, 
                    file = "gage_results_04m.tab", 
                    quote = F, 
                    sep = "\t", 
                    row.names = T)
    write.table(x = gage_results_24m$greater, 
                    file = "gage_results_24m.tab", 
                    quote = F, 
                    sep = "\t", 
                    row.names = T)
    
    #getting pathway ids from the gage results
    pathway_ids <- 
      c(rownames(gage_results_gfp$greater), 
        rownames(gage_results_il2$greater), 
        rownames(gage_results_04m$greater), 
        rownames(gage_results_24m$greater))
    pathway_ids <- gsub(pattern = " .*$", 
                        replacement = "", 
                        x = pathway_ids, 
                        ignore.case = T)
    pathway_ids <- unique(pathway_ids)
    #adding missing pathways
    pathway_ids <- c(pathway_ids, "mmu05022", "mmu04370", "mmu04012", "mmu04014", 
                     "mmu04211", "mmu04657", "mmu01200", "mmu04310", "mmu04024", 
                     "mmu04213", "mmu04371", "mmu04145", "mmu04150", "mmu04015", 
                     "mmu04722", "mmu04068", "mmu04210", "mmu04218", "mmu04136")
    
    #running pathview in batches due to an undocumented limitation of a maximum
    # number of KEGG ids that can be queried in a single call. Not sure what value
    # and whether the limitation is in pathview or on KEGG's end
    pathway_ids <- split(pathway_ids, ceiling(seq_along(pathway_ids)/20))
    for(batch in pathway_ids){
      pathview(gene.data = diff_exprs_results, 
               cpd.data = NULL, 
               pathway.id = batch, 
               species = "mouse", 
               kegg.dir = path, 
               gene.idtype = "SYMBOL", 
               low = list(gene = "#2a7fff", cpd = "blue"), 
               mid = list(gene = "white", cpd = "gray"), 
               high = list(gene = "#ff5555", cpd = "yellow"), 
               limit = list(gene = 0.5, cpd = 1), 
               bins = list(gene = 10, cpd = 10), 
               kegg.native = T, 
               na.col = "gray", 
               same.layer = F)
      }
    
    #resetting work dir
    setwd(work.dir)
    
  }
  
  
  ## bergmann
  astro_data <- readRDS(file = paste0(loc_superclusters,
                                      "astro/astro_data.rds"))
  astro_data <- 
    subset(astro_data, 
           seurat_clusters_renamed == "Bergmann Glia")
  #running gage analysis for old vs young in GFP
  {
    #loading metadata
    metadata <- astro_data@meta.data
    
    #getting cell names and finding column indices
    young_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "04m.PHP.GFAP-GFP"))
    old_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "24m.PHP.GFAP-GFP"))
    young_cells <- which(colnames(astro_data[["SCT"]]@data) %in% young_cells)
    old_cells <- which(colnames(astro_data[["SCT"]]@data) %in% old_cells)
    
    #extracting and formatting log counts matrix to minimise ram footprint
    counts_matrix <- astro_data[["SCT"]]@data[,c(young_cells, old_cells)]
    counts_matrix <- as.data.frame(counts_matrix)
    gc()
    
    #mapping ids
    # do check that the order is the same
    # all(unique(rownames(counts_matrix) == genes$SYMBOL))
    counts_matrix$ENTREZID <- genes$ENTREZID
    counts_matrix <- counts_matrix[complete.cases(counts_matrix),]
    rownames(counts_matrix) <- counts_matrix$ENTREZID
    counts_matrix$ENTREZID <- NULL
    
    #running gage
    gage_results <- gage(exprs = counts_matrix, 
                         gsets = kegg_genesets$kg.sets, 
                         ref = (1:length(young_cells)), 
                         samp = ((length(young_cells)+1) : ncol(counts_matrix)), 
                         same.dir = F, 
                         compare = "as.group", 
                         set.size = c(15, 1000)
                         )
    gc()
    
    saveRDS(object = gage_results, 
            file = paste0(loc_pathview_results, "astro/gage_results_bergmann_GFP.rds"), 
            compress = T)
    
  }
  #running gage analysis for old vs young in IL2
  {
    #loading metadata
    metadata <- astro_data@meta.data
    
    #getting cell names and finding column indices
    young_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "04m.PHP.GFAP-IL2"))
    old_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "24m.PHP.GFAP-IL2"))
    young_cells <- which(colnames(astro_data[["SCT"]]@data) %in% young_cells)
    old_cells <- which(colnames(astro_data[["SCT"]]@data) %in% old_cells)
    
    #extracting and formatting log counts matrix to minimise ram footprint
    counts_matrix <- astro_data[["SCT"]]@data[,c(young_cells, old_cells)]
    counts_matrix <- as.data.frame(counts_matrix)
    gc()
    
    #mapping ids
    # do check that the order is the same
    # all(unique(rownames(counts_matrix) == genes$SYMBOL))
    counts_matrix$ENTREZID <- genes$ENTREZID
    counts_matrix <- counts_matrix[complete.cases(counts_matrix),]
    rownames(counts_matrix) <- counts_matrix$ENTREZID
    counts_matrix$ENTREZID <- NULL
    
    #running gage
    gage_results <- gage(exprs = counts_matrix, 
                         gsets = kegg_genesets$kg.sets, 
                         ref = (1:length(young_cells)), 
                         samp = ((length(young_cells)+1) : ncol(counts_matrix)), 
                         same.dir = F, 
                         compare = "as.group", 
                         set.size = c(15, 1000)
                         )
    gc()
    
    saveRDS(object = gage_results, 
            file = paste0(loc_pathview_results, "astro/gage_results_bergmann_IL2.rds"), 
            compress = T)
    
  }
  #running gage analysis for IL2 vs GFP in young
  {
    #loading metadata
    metadata <- astro_data@meta.data
    
    #getting cell names and finding column indices
    young_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "04m.PHP.GFAP-GFP"))
    old_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "04m.PHP.GFAP-IL2"))
    young_cells <- which(colnames(astro_data[["SCT"]]@data) %in% young_cells)
    old_cells <- which(colnames(astro_data[["SCT"]]@data) %in% old_cells)
    
    #extracting and formatting log counts matrix to minimise ram footprint
    counts_matrix <- astro_data[["SCT"]]@data[,c(young_cells, old_cells)]
    counts_matrix <- as.data.frame(counts_matrix)
    gc()
    
    #mapping ids
    # do check that the order is the same
    # all(unique(rownames(counts_matrix) == genes$SYMBOL))
    counts_matrix$ENTREZID <- genes$ENTREZID
    counts_matrix <- counts_matrix[complete.cases(counts_matrix),]
    rownames(counts_matrix) <- counts_matrix$ENTREZID
    counts_matrix$ENTREZID <- NULL
    
    #running gage
    gage_results <- gage(exprs = counts_matrix, 
                         gsets = kegg_genesets$kg.sets, 
                         ref = (1:length(young_cells)), 
                         samp = ((length(young_cells)+1) : ncol(counts_matrix)), 
                         same.dir = F, 
                         compare = "as.group", 
                         set.size = c(15, 1000)
                         )
    gc()
    
    saveRDS(object = gage_results, 
            file = paste0(loc_pathview_results, "astro/gage_results_bergmann_04m.rds"), 
            compress = T)
    
  }
  #running gage analysis for IL2 vs GFP in old
  {
    #loading metadata
    metadata <- astro_data@meta.data
    
    #getting cell names and finding column indices
    young_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "24m.PHP.GFAP-GFP"))
    old_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "24m.PHP.GFAP-IL2"))
    young_cells <- which(colnames(astro_data[["SCT"]]@data) %in% young_cells)
    old_cells <- which(colnames(astro_data[["SCT"]]@data) %in% old_cells)
    
    #extracting and formatting log counts matrix to minimise ram footprint
    counts_matrix <- astro_data[["SCT"]]@data[,c(young_cells, old_cells)]
    counts_matrix <- as.data.frame(counts_matrix)
    gc()
    
    #mapping ids
    # do check that the order is the same
    # all(unique(rownames(counts_matrix) == genes$SYMBOL))
    counts_matrix$ENTREZID <- genes$ENTREZID
    counts_matrix <- counts_matrix[complete.cases(counts_matrix),]
    rownames(counts_matrix) <- counts_matrix$ENTREZID
    counts_matrix$ENTREZID <- NULL
    
    #running gage
    gage_results <- gage(exprs = counts_matrix, 
                         gsets = kegg_genesets$kg.sets, 
                         ref = (1:length(young_cells)), 
                         samp = ((length(young_cells)+1) : ncol(counts_matrix)), 
                         same.dir = F, 
                         compare = "as.group", 
                         set.size = c(15, 1000)
                         )
    gc()
    
    saveRDS(object = gage_results, 
            file = paste0(loc_pathview_results, "astro/gage_results_bergmann_24m.rds"), 
            compress = T)
    
  }
  #running pathview analysis on treatment results
  {
    #loading GFP results
    gage_results_gfp <- readRDS(file = paste0(loc_pathview_results, 
                                              "astro/gage_results_bergmann_GFP.rds"))
    diff_exprs_results_gfp <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "astro/subpops/Bergmann.Glia_markers_list_gfp_nofilter.rds"))[[2]]
    diff_exprs_results_gfp <- 
      diff_exprs_results_gfp[genes_to_select, "avg_log2FC", 
                         drop = F]
    
    #loading IL2 results
    gage_results_il2 <- readRDS(file = paste0(loc_pathview_results, 
                                              "astro/gage_results_bergmann_IL2.rds"))
    diff_exprs_results_il2 <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "astro/subpops/Bergmann.Glia_markers_list_il2_nofilter.rds"))[[2]]
    diff_exprs_results_il2 <- 
      diff_exprs_results_il2[genes_to_select, "avg_log2FC", 
                         drop = F]
    
    #loading young results
    gage_results_04m <- readRDS(file = paste0(loc_pathview_results, 
                                              "astro/gage_results_bergmann_04m.rds"))
    diff_exprs_results_04m <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "astro/subpops/Bergmann.Glia_markers_list_04m_nofilter.rds"))[[2]]
    diff_exprs_results_04m <- 
      diff_exprs_results_04m[genes_to_select, "avg_log2FC", 
                         drop = F]
    
    #loading old results
    gage_results_24m <- readRDS(file = paste0(loc_pathview_results, 
                                              "astro/gage_results_bergmann_24m.rds"))
    diff_exprs_results_24m <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "astro/subpops/Bergmann.Glia_markers_list_24m_nofilter.rds"))[[2]]
    diff_exprs_results_24m <- 
      diff_exprs_results_24m[genes_to_select, "avg_log2FC", 
                         drop = F]
    
    #merging diff exprs data
    diff_exprs_results <- merge(x = diff_exprs_results_gfp, 
                                y = diff_exprs_results_il2, 
                                by = "row.names", 
                                all = T)
    diff_exprs_results_2 <- merge(x = diff_exprs_results_04m, 
                                y = diff_exprs_results_24m, 
                                by = "row.names", 
                                all = T)
    diff_exprs_results <- merge(x = diff_exprs_results, 
                                y = diff_exprs_results_2, 
                                by = "Row.names", 
                                all = T)
    rownames(diff_exprs_results) <- diff_exprs_results$Row.names
    diff_exprs_results$Row.names <- NULL
    diff_exprs_results_2 <- NULL
    colnames(diff_exprs_results) <- paste0("Contrast ", 
                                           1:ncol(diff_exprs_results))
    
    #to exclude those extreme cases where a gene was not detected in a contrast
    diff_exprs_results <- diff_exprs_results[complete.cases(diff_exprs_results),]
    
    #setting up pathview
    data(gene.idtype.bods)
    work.dir <- getwd()
    path <- paste0(loc_pathview_results, "astro/bergmann_pathview_treatment/")
    dir.create(path = path, recursive = T)
    setwd(path)
    
    #writing out the gage results
    write.table(x = gage_results_gfp$greater, 
                    file = "gage_results_GFP.tab", 
                    quote = F, 
                    sep = "\t", 
                    row.names = T)
    write.table(x = gage_results_il2$greater, 
                    file = "gage_results_IL2.tab", 
                    quote = F, 
                    sep = "\t", 
                    row.names = T)
    write.table(x = gage_results_04m$greater, 
                    file = "gage_results_04m.tab", 
                    quote = F, 
                    sep = "\t", 
                    row.names = T)
    write.table(x = gage_results_24m$greater, 
                    file = "gage_results_24m.tab", 
                    quote = F, 
                    sep = "\t", 
                    row.names = T)
    
    #getting pathway ids from the gage results
    pathway_ids <- 
      c(rownames(gage_results_gfp$greater), 
        rownames(gage_results_il2$greater), 
        rownames(gage_results_04m$greater), 
        rownames(gage_results_24m$greater))
    pathway_ids <- gsub(pattern = " .*$", 
                        replacement = "", 
                        x = pathway_ids, 
                        ignore.case = T)
    pathway_ids <- unique(pathway_ids)
    #adding missing pathways
    pathway_ids <- c(pathway_ids, "mmu05022", "mmu04370", "mmu04012", "mmu04014", 
                     "mmu04211", "mmu04657", "mmu01200", "mmu04310", "mmu04024", 
                     "mmu04213", "mmu04371", "mmu04145", "mmu04150", "mmu04015", 
                     "mmu04722", "mmu04068", "mmu04210", "mmu04218", "mmu04136")
    
    #running pathview in batches due to an undocumented limitation of a maximum
    # number of KEGG ids that can be queried in a single call. Not sure what value
    # and whether the limitation is in pathview or on KEGG's end
    pathway_ids <- split(pathway_ids, ceiling(seq_along(pathway_ids)/20))
    for(batch in pathway_ids){
      pathview(gene.data = diff_exprs_results, 
               cpd.data = NULL, 
               pathway.id = batch, 
               species = "mouse", 
               kegg.dir = path, 
               gene.idtype = "SYMBOL", 
               low = list(gene = "#2a7fff", cpd = "blue"), 
               mid = list(gene = "white", cpd = "gray"), 
               high = list(gene = "#ff5555", cpd = "yellow"), 
               limit = list(gene = 0.5, cpd = 1), 
               bins = list(gene = 10, cpd = 10), 
               kegg.native = T, 
               na.col = "gray", 
               same.layer = F)
      }
    
    #resetting work dir
    setwd(work.dir)
    
  }
  
  
  ## cerebellar
  astro_data <- readRDS(file = paste0(loc_superclusters,
                                      "astro/astro_data.rds"))
  astro_data <- 
    subset(astro_data, 
           seurat_clusters_renamed == "Cerebellar\nAstrocytes")
  #running gage analysis for old vs young in GFP
  {
    #loading metadata
    metadata <- astro_data@meta.data
    
    #getting cell names and finding column indices
    young_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "04m.PHP.GFAP-GFP"))
    old_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "24m.PHP.GFAP-GFP"))
    young_cells <- which(colnames(astro_data[["SCT"]]@data) %in% young_cells)
    old_cells <- which(colnames(astro_data[["SCT"]]@data) %in% old_cells)
    
    #extracting and formatting log counts matrix to minimise ram footprint
    counts_matrix <- astro_data[["SCT"]]@data[,c(young_cells, old_cells)]
    counts_matrix <- as.data.frame(counts_matrix)
    gc()
    
    #mapping ids
    # do check that the order is the same
    # all(unique(rownames(counts_matrix) == genes$SYMBOL))
    counts_matrix$ENTREZID <- genes$ENTREZID
    counts_matrix <- counts_matrix[complete.cases(counts_matrix),]
    rownames(counts_matrix) <- counts_matrix$ENTREZID
    counts_matrix$ENTREZID <- NULL
    
    #running gage
    gage_results <- gage(exprs = counts_matrix, 
                         gsets = kegg_genesets$kg.sets, 
                         ref = (1:length(young_cells)), 
                         samp = ((length(young_cells)+1) : ncol(counts_matrix)), 
                         same.dir = F, 
                         compare = "as.group", 
                         set.size = c(15, 1000)
                         )
    gc()
    
    saveRDS(object = gage_results, 
            file = paste0(loc_pathview_results, "astro/gage_results_cerebellar_GFP.rds"), 
            compress = T)
    
  }
  #running gage analysis for old vs young in IL2
  {
    #loading metadata
    metadata <- astro_data@meta.data
    
    #getting cell names and finding column indices
    young_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "04m.PHP.GFAP-IL2"))
    old_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "24m.PHP.GFAP-IL2"))
    young_cells <- which(colnames(astro_data[["SCT"]]@data) %in% young_cells)
    old_cells <- which(colnames(astro_data[["SCT"]]@data) %in% old_cells)
    
    #extracting and formatting log counts matrix to minimise ram footprint
    counts_matrix <- astro_data[["SCT"]]@data[,c(young_cells, old_cells)]
    counts_matrix <- as.data.frame(counts_matrix)
    gc()
    
    #mapping ids
    # do check that the order is the same
    # all(unique(rownames(counts_matrix) == genes$SYMBOL))
    counts_matrix$ENTREZID <- genes$ENTREZID
    counts_matrix <- counts_matrix[complete.cases(counts_matrix),]
    rownames(counts_matrix) <- counts_matrix$ENTREZID
    counts_matrix$ENTREZID <- NULL
    
    #running gage
    gage_results <- gage(exprs = counts_matrix, 
                         gsets = kegg_genesets$kg.sets, 
                         ref = (1:length(young_cells)), 
                         samp = ((length(young_cells)+1) : ncol(counts_matrix)), 
                         same.dir = F, 
                         compare = "as.group", 
                         set.size = c(15, 1000)
                         )
    gc()
    
    saveRDS(object = gage_results, 
            file = paste0(loc_pathview_results, "astro/gage_results_cerebellar_IL2.rds"), 
            compress = T)
    
  }
  #running gage analysis for IL2 vs GFP in young
  {
    #loading metadata
    metadata <- astro_data@meta.data
    
    #getting cell names and finding column indices
    young_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "04m.PHP.GFAP-GFP"))
    old_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "04m.PHP.GFAP-IL2"))
    young_cells <- which(colnames(astro_data[["SCT"]]@data) %in% young_cells)
    old_cells <- which(colnames(astro_data[["SCT"]]@data) %in% old_cells)
    
    #extracting and formatting log counts matrix to minimise ram footprint
    counts_matrix <- astro_data[["SCT"]]@data[,c(young_cells, old_cells)]
    counts_matrix <- as.data.frame(counts_matrix)
    gc()
    
    #mapping ids
    # do check that the order is the same
    # all(unique(rownames(counts_matrix) == genes$SYMBOL))
    counts_matrix$ENTREZID <- genes$ENTREZID
    counts_matrix <- counts_matrix[complete.cases(counts_matrix),]
    rownames(counts_matrix) <- counts_matrix$ENTREZID
    counts_matrix$ENTREZID <- NULL
    
    #running gage
    gage_results <- gage(exprs = counts_matrix, 
                         gsets = kegg_genesets$kg.sets, 
                         ref = (1:length(young_cells)), 
                         samp = ((length(young_cells)+1) : ncol(counts_matrix)), 
                         same.dir = F, 
                         compare = "as.group", 
                         set.size = c(15, 1000)
                         )
    gc()
    
    saveRDS(object = gage_results, 
            file = paste0(loc_pathview_results, "astro/gage_results_cerebellar_04m.rds"), 
            compress = T)
    
  }
  #running gage analysis for IL2 vs GFP in old
  {
    #loading metadata
    metadata <- astro_data@meta.data
    
    #getting cell names and finding column indices
    young_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "24m.PHP.GFAP-GFP"))
    old_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "24m.PHP.GFAP-IL2"))
    young_cells <- which(colnames(astro_data[["SCT"]]@data) %in% young_cells)
    old_cells <- which(colnames(astro_data[["SCT"]]@data) %in% old_cells)
    
    #extracting and formatting log counts matrix to minimise ram footprint
    counts_matrix <- astro_data[["SCT"]]@data[,c(young_cells, old_cells)]
    counts_matrix <- as.data.frame(counts_matrix)
    gc()
    
    #mapping ids
    # do check that the order is the same
    # all(unique(rownames(counts_matrix) == genes$SYMBOL))
    counts_matrix$ENTREZID <- genes$ENTREZID
    counts_matrix <- counts_matrix[complete.cases(counts_matrix),]
    rownames(counts_matrix) <- counts_matrix$ENTREZID
    counts_matrix$ENTREZID <- NULL
    
    #running gage
    gage_results <- gage(exprs = counts_matrix, 
                         gsets = kegg_genesets$kg.sets, 
                         ref = (1:length(young_cells)), 
                         samp = ((length(young_cells)+1) : ncol(counts_matrix)), 
                         same.dir = F, 
                         compare = "as.group", 
                         set.size = c(15, 1000)
                         )
    gc()
    
    saveRDS(object = gage_results, 
            file = paste0(loc_pathview_results, "astro/gage_results_cerebellar_24m.rds"), 
            compress = T)
    
  }
  #running pathview analysis on treatment results
  {
    #loading GFP results
    gage_results_gfp <- readRDS(file = paste0(loc_pathview_results, 
                                              "astro/gage_results_cerebellar_GFP.rds"))
    diff_exprs_results_gfp <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "astro/subpops/Cerebellar.Astrocytes_markers_list_gfp_nofilter.rds"))[[2]]
    diff_exprs_results_gfp <- 
      diff_exprs_results_gfp[genes_to_select, "avg_log2FC", 
                         drop = F]
    
    #loading IL2 results
    gage_results_il2 <- readRDS(file = paste0(loc_pathview_results, 
                                              "astro/gage_results_cerebellar_IL2.rds"))
    diff_exprs_results_il2 <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "astro/subpops/Cerebellar.Astrocytes_markers_list_il2_nofilter.rds"))[[2]]
    diff_exprs_results_il2 <- 
      diff_exprs_results_il2[genes_to_select, "avg_log2FC", 
                         drop = F]
    
    #loading young results
    gage_results_04m <- readRDS(file = paste0(loc_pathview_results, 
                                              "astro/gage_results_cerebellar_04m.rds"))
    diff_exprs_results_04m <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "astro/subpops/Cerebellar.Astrocytes_markers_list_04m_nofilter.rds"))[[2]]
    diff_exprs_results_04m <- 
      diff_exprs_results_04m[genes_to_select, "avg_log2FC", 
                         drop = F]
    
    #loading old results
    gage_results_24m <- readRDS(file = paste0(loc_pathview_results, 
                                              "astro/gage_results_cerebellar_24m.rds"))
    diff_exprs_results_24m <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "astro/subpops/Cerebellar.Astrocytes_markers_list_24m_nofilter.rds"))[[2]]
    diff_exprs_results_24m <- 
      diff_exprs_results_24m[genes_to_select, "avg_log2FC", 
                         drop = F]
    
    #merging diff exprs data
    diff_exprs_results <- merge(x = diff_exprs_results_gfp, 
                                y = diff_exprs_results_il2, 
                                by = "row.names", 
                                all = T)
    diff_exprs_results_2 <- merge(x = diff_exprs_results_04m, 
                                y = diff_exprs_results_24m, 
                                by = "row.names", 
                                all = T)
    diff_exprs_results <- merge(x = diff_exprs_results, 
                                y = diff_exprs_results_2, 
                                by = "Row.names", 
                                all = T)
    rownames(diff_exprs_results) <- diff_exprs_results$Row.names
    diff_exprs_results$Row.names <- NULL
    diff_exprs_results_2 <- NULL
    colnames(diff_exprs_results) <- paste0("Contrast ", 
                                           1:ncol(diff_exprs_results))
    
    #to exclude those extreme cases where a gene was not detected in a contrast
    diff_exprs_results <- diff_exprs_results[complete.cases(diff_exprs_results),]
    
    #setting up pathview
    data(gene.idtype.bods)
    work.dir <- getwd()
    path <- paste0(loc_pathview_results, "astro/cerebellar_pathview_treatment/")
    dir.create(path = path, recursive = T)
    setwd(path)
    
    #writing out the gage results
    write.table(x = gage_results_gfp$greater, 
                    file = "gage_results_GFP.tab", 
                    quote = F, 
                    sep = "\t", 
                    row.names = T)
    write.table(x = gage_results_il2$greater, 
                    file = "gage_results_IL2.tab", 
                    quote = F, 
                    sep = "\t", 
                    row.names = T)
    write.table(x = gage_results_04m$greater, 
                    file = "gage_results_04m.tab", 
                    quote = F, 
                    sep = "\t", 
                    row.names = T)
    write.table(x = gage_results_24m$greater, 
                    file = "gage_results_24m.tab", 
                    quote = F, 
                    sep = "\t", 
                    row.names = T)
    
    #getting pathway ids from the gage results
    pathway_ids <- 
      c(rownames(gage_results_gfp$greater), 
        rownames(gage_results_il2$greater), 
        rownames(gage_results_04m$greater), 
        rownames(gage_results_24m$greater))
    pathway_ids <- gsub(pattern = " .*$", 
                        replacement = "", 
                        x = pathway_ids, 
                        ignore.case = T)
    pathway_ids <- unique(pathway_ids)
    #adding missing pathways
    pathway_ids <- c(pathway_ids, "mmu05022", "mmu04370", "mmu04012", "mmu04014", 
                     "mmu04211", "mmu04657", "mmu01200", "mmu04310", "mmu04024", 
                     "mmu04213", "mmu04371", "mmu04145", "mmu04150", "mmu04015", 
                     "mmu04722", "mmu04068", "mmu04210", "mmu04218", "mmu04136")
    
    #running pathview in batches due to an undocumented limitation of a maximum
    # number of KEGG ids that can be queried in a single call. Not sure what value
    # and whether the limitation is in pathview or on KEGG's end
    pathway_ids <- split(pathway_ids, ceiling(seq_along(pathway_ids)/20))
    for(batch in pathway_ids){
      pathview(gene.data = diff_exprs_results, 
               cpd.data = NULL, 
               pathway.id = batch, 
               species = "mouse", 
               kegg.dir = path, 
               gene.idtype = "SYMBOL", 
               low = list(gene = "#2a7fff", cpd = "blue"), 
               mid = list(gene = "white", cpd = "gray"), 
               high = list(gene = "#ff5555", cpd = "yellow"), 
               limit = list(gene = 0.5, cpd = 1), 
               bins = list(gene = 10, cpd = 10), 
               kegg.native = T, 
               na.col = "gray", 
               same.layer = F)
      }
    
    #resetting work dir
    setwd(work.dir)
    
  }
  
  
  ## olfactory
  astro_data <- readRDS(file = paste0(loc_superclusters,
                                      "astro/astro_data.rds"))
  astro_data <- 
    subset(astro_data, 
           seurat_clusters_renamed == "Olfactory\nAstrocytes")
  #running gage analysis for old vs young in GFP
  {
    #loading metadata
    metadata <- astro_data@meta.data
    
    #getting cell names and finding column indices
    young_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "04m.PHP.GFAP-GFP"))
    old_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "24m.PHP.GFAP-GFP"))
    young_cells <- which(colnames(astro_data[["SCT"]]@data) %in% young_cells)
    old_cells <- which(colnames(astro_data[["SCT"]]@data) %in% old_cells)
    
    #extracting and formatting log counts matrix to minimise ram footprint
    counts_matrix <- astro_data[["SCT"]]@data[,c(young_cells, old_cells)]
    counts_matrix <- as.data.frame(counts_matrix)
    gc()
    
    #mapping ids
    # do check that the order is the same
    # all(unique(rownames(counts_matrix) == genes$SYMBOL))
    counts_matrix$ENTREZID <- genes$ENTREZID
    counts_matrix <- counts_matrix[complete.cases(counts_matrix),]
    rownames(counts_matrix) <- counts_matrix$ENTREZID
    counts_matrix$ENTREZID <- NULL
    
    #running gage
    gage_results <- gage(exprs = counts_matrix, 
                         gsets = kegg_genesets$kg.sets, 
                         ref = (1:length(young_cells)), 
                         samp = ((length(young_cells)+1) : ncol(counts_matrix)), 
                         same.dir = F, 
                         compare = "as.group", 
                         set.size = c(15, 1000)
                         )
    gc()
    
    saveRDS(object = gage_results, 
            file = paste0(loc_pathview_results, "astro/gage_results_olfactory_GFP.rds"), 
            compress = T)
    
  }
  #running gage analysis for old vs young in IL2
  {
    #loading metadata
    metadata <- astro_data@meta.data
    
    #getting cell names and finding column indices
    young_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "04m.PHP.GFAP-IL2"))
    old_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "24m.PHP.GFAP-IL2"))
    young_cells <- which(colnames(astro_data[["SCT"]]@data) %in% young_cells)
    old_cells <- which(colnames(astro_data[["SCT"]]@data) %in% old_cells)
    
    #extracting and formatting log counts matrix to minimise ram footprint
    counts_matrix <- astro_data[["SCT"]]@data[,c(young_cells, old_cells)]
    counts_matrix <- as.data.frame(counts_matrix)
    gc()
    
    #mapping ids
    # do check that the order is the same
    # all(unique(rownames(counts_matrix) == genes$SYMBOL))
    counts_matrix$ENTREZID <- genes$ENTREZID
    counts_matrix <- counts_matrix[complete.cases(counts_matrix),]
    rownames(counts_matrix) <- counts_matrix$ENTREZID
    counts_matrix$ENTREZID <- NULL
    
    #running gage
    gage_results <- gage(exprs = counts_matrix, 
                         gsets = kegg_genesets$kg.sets, 
                         ref = (1:length(young_cells)), 
                         samp = ((length(young_cells)+1) : ncol(counts_matrix)), 
                         same.dir = F, 
                         compare = "as.group", 
                         set.size = c(15, 1000)
                         )
    gc()
    
    saveRDS(object = gage_results, 
            file = paste0(loc_pathview_results, "astro/gage_results_olfactory_IL2.rds"), 
            compress = T)
    
  }
  #running gage analysis for IL2 vs GFP in young
  {
    #loading metadata
    metadata <- astro_data@meta.data
    
    #getting cell names and finding column indices
    young_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "04m.PHP.GFAP-GFP"))
    old_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "04m.PHP.GFAP-IL2"))
    young_cells <- which(colnames(astro_data[["SCT"]]@data) %in% young_cells)
    old_cells <- which(colnames(astro_data[["SCT"]]@data) %in% old_cells)
    
    #extracting and formatting log counts matrix to minimise ram footprint
    counts_matrix <- astro_data[["SCT"]]@data[,c(young_cells, old_cells)]
    counts_matrix <- as.data.frame(counts_matrix)
    gc()
    
    #mapping ids
    # do check that the order is the same
    # all(unique(rownames(counts_matrix) == genes$SYMBOL))
    counts_matrix$ENTREZID <- genes$ENTREZID
    counts_matrix <- counts_matrix[complete.cases(counts_matrix),]
    rownames(counts_matrix) <- counts_matrix$ENTREZID
    counts_matrix$ENTREZID <- NULL
    
    #running gage
    gage_results <- gage(exprs = counts_matrix, 
                         gsets = kegg_genesets$kg.sets, 
                         ref = (1:length(young_cells)), 
                         samp = ((length(young_cells)+1) : ncol(counts_matrix)), 
                         same.dir = F, 
                         compare = "as.group", 
                         set.size = c(15, 1000)
                         )
    gc()
    
    saveRDS(object = gage_results, 
            file = paste0(loc_pathview_results, "astro/gage_results_olfactory_04m.rds"), 
            compress = T)
    
  }
  #running gage analysis for IL2 vs GFP in old
  {
    #loading metadata
    metadata <- astro_data@meta.data
    
    #getting cell names and finding column indices
    young_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "24m.PHP.GFAP-GFP"))
    old_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "24m.PHP.GFAP-IL2"))
    young_cells <- which(colnames(astro_data[["SCT"]]@data) %in% young_cells)
    old_cells <- which(colnames(astro_data[["SCT"]]@data) %in% old_cells)
    
    #extracting and formatting log counts matrix to minimise ram footprint
    counts_matrix <- astro_data[["SCT"]]@data[,c(young_cells, old_cells)]
    counts_matrix <- as.data.frame(counts_matrix)
    gc()
    
    #mapping ids
    # do check that the order is the same
    # all(unique(rownames(counts_matrix) == genes$SYMBOL))
    counts_matrix$ENTREZID <- genes$ENTREZID
    counts_matrix <- counts_matrix[complete.cases(counts_matrix),]
    rownames(counts_matrix) <- counts_matrix$ENTREZID
    counts_matrix$ENTREZID <- NULL
    
    #running gage
    gage_results <- gage(exprs = counts_matrix, 
                         gsets = kegg_genesets$kg.sets, 
                         ref = (1:length(young_cells)), 
                         samp = ((length(young_cells)+1) : ncol(counts_matrix)), 
                         same.dir = F, 
                         compare = "as.group", 
                         set.size = c(15, 1000)
                         )
    gc()
    
    saveRDS(object = gage_results, 
            file = paste0(loc_pathview_results, "astro/gage_results_olfactory_24m.rds"), 
            compress = T)
    
  }
  #running pathview analysis on treatment results
  {
    #loading GFP results
    gage_results_gfp <- readRDS(file = paste0(loc_pathview_results, 
                                              "astro/gage_results_olfactory_GFP.rds"))
    diff_exprs_results_gfp <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "astro/subpops/Olfactory.Astrocytes_markers_list_gfp_nofilter.rds"))[[2]]
    diff_exprs_results_gfp <- 
      diff_exprs_results_gfp[genes_to_select, "avg_log2FC", 
                         drop = F]
    
    #loading IL2 results
    gage_results_il2 <- readRDS(file = paste0(loc_pathview_results, 
                                              "astro/gage_results_olfactory_IL2.rds"))
    diff_exprs_results_il2 <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "astro/subpops/Olfactory.Astrocytes_markers_list_il2_nofilter.rds"))[[2]]
    diff_exprs_results_il2 <- 
      diff_exprs_results_il2[genes_to_select, "avg_log2FC", 
                         drop = F]
    
    #loading young results
    gage_results_04m <- readRDS(file = paste0(loc_pathview_results, 
                                              "astro/gage_results_olfactory_04m.rds"))
    diff_exprs_results_04m <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "astro/subpops/Olfactory.Astrocytes_markers_list_04m_nofilter.rds"))[[2]]
    diff_exprs_results_04m <- 
      diff_exprs_results_04m[genes_to_select, "avg_log2FC", 
                         drop = F]
    
    #loading old results
    gage_results_24m <- readRDS(file = paste0(loc_pathview_results, 
                                              "astro/gage_results_olfactory_24m.rds"))
    diff_exprs_results_24m <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "astro/subpops/Olfactory.Astrocytes_markers_list_24m_nofilter.rds"))[[2]]
    diff_exprs_results_24m <- 
      diff_exprs_results_24m[genes_to_select, "avg_log2FC", 
                         drop = F]
    
    #merging diff exprs data
    diff_exprs_results <- merge(x = diff_exprs_results_gfp, 
                                y = diff_exprs_results_il2, 
                                by = "row.names", 
                                all = T)
    diff_exprs_results_2 <- merge(x = diff_exprs_results_04m, 
                                y = diff_exprs_results_24m, 
                                by = "row.names", 
                                all = T)
    diff_exprs_results <- merge(x = diff_exprs_results, 
                                y = diff_exprs_results_2, 
                                by = "Row.names", 
                                all = T)
    rownames(diff_exprs_results) <- diff_exprs_results$Row.names
    diff_exprs_results$Row.names <- NULL
    diff_exprs_results_2 <- NULL
    colnames(diff_exprs_results) <- paste0("Contrast ", 
                                           1:ncol(diff_exprs_results))
    
    #to exclude those extreme cases where a gene was not detected in a contrast
    diff_exprs_results <- diff_exprs_results[complete.cases(diff_exprs_results),]
    
    #setting up pathview
    data(gene.idtype.bods)
    work.dir <- getwd()
    path <- paste0(loc_pathview_results, "astro/olfactory_pathview_treatment/")
    dir.create(path = path, recursive = T)
    setwd(path)
    
    #writing out the gage results
    write.table(x = gage_results_gfp$greater, 
                    file = "gage_results_GFP.tab", 
                    quote = F, 
                    sep = "\t", 
                    row.names = T)
    write.table(x = gage_results_il2$greater, 
                    file = "gage_results_IL2.tab", 
                    quote = F, 
                    sep = "\t", 
                    row.names = T)
    write.table(x = gage_results_04m$greater, 
                    file = "gage_results_04m.tab", 
                    quote = F, 
                    sep = "\t", 
                    row.names = T)
    write.table(x = gage_results_24m$greater, 
                    file = "gage_results_24m.tab", 
                    quote = F, 
                    sep = "\t", 
                    row.names = T)
    
    #getting pathway ids from the gage results
    pathway_ids <- 
      c(rownames(gage_results_gfp$greater), 
        rownames(gage_results_il2$greater), 
        rownames(gage_results_04m$greater), 
        rownames(gage_results_24m$greater))
    pathway_ids <- gsub(pattern = " .*$", 
                        replacement = "", 
                        x = pathway_ids, 
                        ignore.case = T)
    pathway_ids <- unique(pathway_ids)
    #adding missing pathways
    pathway_ids <- c(pathway_ids, "mmu05022", "mmu04370", "mmu04012", "mmu04014", 
                     "mmu04211", "mmu04657", "mmu01200", "mmu04310", "mmu04024", 
                     "mmu04213", "mmu04371", "mmu04145", "mmu04150", "mmu04015", 
                     "mmu04722", "mmu04068", "mmu04210", "mmu04218", "mmu04136")
    
    #running pathview in batches due to an undocumented limitation of a maximum
    # number of KEGG ids that can be queried in a single call. Not sure what value
    # and whether the limitation is in pathview or on KEGG's end
    pathway_ids <- split(pathway_ids, ceiling(seq_along(pathway_ids)/20))
    for(batch in pathway_ids){
      pathview(gene.data = diff_exprs_results, 
               cpd.data = NULL, 
               pathway.id = batch, 
               species = "mouse", 
               kegg.dir = path, 
               gene.idtype = "SYMBOL", 
               low = list(gene = "#2a7fff", cpd = "blue"), 
               mid = list(gene = "white", cpd = "gray"), 
               high = list(gene = "#ff5555", cpd = "yellow"), 
               limit = list(gene = 0.5, cpd = 1), 
               bins = list(gene = 10, cpd = 10), 
               kegg.native = T, 
               na.col = "gray", 
               same.layer = F)
      }
    
    #resetting work dir
    setwd(work.dir)
    
  }
  
  
  ## striatal
  astro_data <- readRDS(file = paste0(loc_superclusters,
                                      "astro/astro_data.rds"))
  astro_data <- 
    subset(astro_data, 
           seurat_clusters_renamed == "Striatal\nAstrocytes")
  #running gage analysis for old vs young in GFP
  {
    #loading metadata
    metadata <- astro_data@meta.data
    
    #getting cell names and finding column indices
    young_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "04m.PHP.GFAP-GFP"))
    old_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "24m.PHP.GFAP-GFP"))
    young_cells <- which(colnames(astro_data[["SCT"]]@data) %in% young_cells)
    old_cells <- which(colnames(astro_data[["SCT"]]@data) %in% old_cells)
    
    #extracting and formatting log counts matrix to minimise ram footprint
    counts_matrix <- astro_data[["SCT"]]@data[,c(young_cells, old_cells)]
    counts_matrix <- as.data.frame(counts_matrix)
    gc()
    
    #mapping ids
    # do check that the order is the same
    # all(unique(rownames(counts_matrix) == genes$SYMBOL))
    counts_matrix$ENTREZID <- genes$ENTREZID
    counts_matrix <- counts_matrix[complete.cases(counts_matrix),]
    rownames(counts_matrix) <- counts_matrix$ENTREZID
    counts_matrix$ENTREZID <- NULL
    
    #running gage
    gage_results <- gage(exprs = counts_matrix, 
                         gsets = kegg_genesets$kg.sets, 
                         ref = (1:length(young_cells)), 
                         samp = ((length(young_cells)+1) : ncol(counts_matrix)), 
                         same.dir = F, 
                         compare = "as.group", 
                         set.size = c(15, 1000)
                         )
    gc()
    
    saveRDS(object = gage_results, 
            file = paste0(loc_pathview_results, "astro/gage_results_striatal_GFP.rds"), 
            compress = T)
    
  }
  #running gage analysis for old vs young in IL2
  {
    #loading metadata
    metadata <- astro_data@meta.data
    
    #getting cell names and finding column indices
    young_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "04m.PHP.GFAP-IL2"))
    old_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "24m.PHP.GFAP-IL2"))
    young_cells <- which(colnames(astro_data[["SCT"]]@data) %in% young_cells)
    old_cells <- which(colnames(astro_data[["SCT"]]@data) %in% old_cells)
    
    #extracting and formatting log counts matrix to minimise ram footprint
    counts_matrix <- astro_data[["SCT"]]@data[,c(young_cells, old_cells)]
    counts_matrix <- as.data.frame(counts_matrix)
    gc()
    
    #mapping ids
    # do check that the order is the same
    # all(unique(rownames(counts_matrix) == genes$SYMBOL))
    counts_matrix$ENTREZID <- genes$ENTREZID
    counts_matrix <- counts_matrix[complete.cases(counts_matrix),]
    rownames(counts_matrix) <- counts_matrix$ENTREZID
    counts_matrix$ENTREZID <- NULL
    
    #running gage
    gage_results <- gage(exprs = counts_matrix, 
                         gsets = kegg_genesets$kg.sets, 
                         ref = (1:length(young_cells)), 
                         samp = ((length(young_cells)+1) : ncol(counts_matrix)), 
                         same.dir = F, 
                         compare = "as.group", 
                         set.size = c(15, 1000)
                         )
    gc()
    
    saveRDS(object = gage_results, 
            file = paste0(loc_pathview_results, "astro/gage_results_striatal_IL2.rds"), 
            compress = T)
    
  }
  #running gage analysis for IL2 vs GFP in young
  {
    #loading metadata
    metadata <- astro_data@meta.data
    
    #getting cell names and finding column indices
    young_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "04m.PHP.GFAP-GFP"))
    old_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "04m.PHP.GFAP-IL2"))
    young_cells <- which(colnames(astro_data[["SCT"]]@data) %in% young_cells)
    old_cells <- which(colnames(astro_data[["SCT"]]@data) %in% old_cells)
    
    #extracting and formatting log counts matrix to minimise ram footprint
    counts_matrix <- astro_data[["SCT"]]@data[,c(young_cells, old_cells)]
    counts_matrix <- as.data.frame(counts_matrix)
    gc()
    
    #mapping ids
    # do check that the order is the same
    # all(unique(rownames(counts_matrix) == genes$SYMBOL))
    counts_matrix$ENTREZID <- genes$ENTREZID
    counts_matrix <- counts_matrix[complete.cases(counts_matrix),]
    rownames(counts_matrix) <- counts_matrix$ENTREZID
    counts_matrix$ENTREZID <- NULL
    
    #running gage
    gage_results <- gage(exprs = counts_matrix, 
                         gsets = kegg_genesets$kg.sets, 
                         ref = (1:length(young_cells)), 
                         samp = ((length(young_cells)+1) : ncol(counts_matrix)), 
                         same.dir = F, 
                         compare = "as.group", 
                         set.size = c(15, 1000)
                         )
    gc()
    
    saveRDS(object = gage_results, 
            file = paste0(loc_pathview_results, "astro/gage_results_striatal_04m.rds"), 
            compress = T)
    
  }
  #running gage analysis for IL2 vs GFP in old
  {
    #loading metadata
    metadata <- astro_data@meta.data
    
    #getting cell names and finding column indices
    young_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "24m.PHP.GFAP-GFP"))
    old_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "24m.PHP.GFAP-IL2"))
    young_cells <- which(colnames(astro_data[["SCT"]]@data) %in% young_cells)
    old_cells <- which(colnames(astro_data[["SCT"]]@data) %in% old_cells)
    
    #extracting and formatting log counts matrix to minimise ram footprint
    counts_matrix <- astro_data[["SCT"]]@data[,c(young_cells, old_cells)]
    counts_matrix <- as.data.frame(counts_matrix)
    gc()
    
    #mapping ids
    # do check that the order is the same
    # all(unique(rownames(counts_matrix) == genes$SYMBOL))
    counts_matrix$ENTREZID <- genes$ENTREZID
    counts_matrix <- counts_matrix[complete.cases(counts_matrix),]
    rownames(counts_matrix) <- counts_matrix$ENTREZID
    counts_matrix$ENTREZID <- NULL
    
    #running gage
    gage_results <- gage(exprs = counts_matrix, 
                         gsets = kegg_genesets$kg.sets, 
                         ref = (1:length(young_cells)), 
                         samp = ((length(young_cells)+1) : ncol(counts_matrix)), 
                         same.dir = F, 
                         compare = "as.group", 
                         set.size = c(15, 1000)
                         )
    gc()
    
    saveRDS(object = gage_results, 
            file = paste0(loc_pathview_results, "astro/gage_results_striatal_24m.rds"), 
            compress = T)
    
  }
  #running pathview analysis on treatment results
  {
    #loading GFP results
    gage_results_gfp <- readRDS(file = paste0(loc_pathview_results, 
                                              "astro/gage_results_striatal_GFP.rds"))
    diff_exprs_results_gfp <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "astro/subpops/Striatal.Astrocytes_markers_list_gfp_nofilter.rds"))[[2]]
    diff_exprs_results_gfp <- 
      diff_exprs_results_gfp[genes_to_select, "avg_log2FC", 
                         drop = F]
    
    #loading IL2 results
    gage_results_il2 <- readRDS(file = paste0(loc_pathview_results, 
                                              "astro/gage_results_striatal_IL2.rds"))
    diff_exprs_results_il2 <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "astro/subpops/Striatal.Astrocytes_markers_list_il2_nofilter.rds"))[[2]]
    diff_exprs_results_il2 <- 
      diff_exprs_results_il2[genes_to_select, "avg_log2FC", 
                         drop = F]
    
    #loading young results
    gage_results_04m <- readRDS(file = paste0(loc_pathview_results, 
                                              "astro/gage_results_striatal_04m.rds"))
    diff_exprs_results_04m <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "astro/subpops/Striatal.Astrocytes_markers_list_04m_nofilter.rds"))[[2]]
    diff_exprs_results_04m <- 
      diff_exprs_results_04m[genes_to_select, "avg_log2FC", 
                         drop = F]
    
    #loading old results
    gage_results_24m <- readRDS(file = paste0(loc_pathview_results, 
                                              "astro/gage_results_striatal_24m.rds"))
    diff_exprs_results_24m <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            "astro/subpops/Striatal.Astrocytes_markers_list_24m_nofilter.rds"))[[2]]
    diff_exprs_results_24m <- 
      diff_exprs_results_24m[genes_to_select, "avg_log2FC", 
                         drop = F]
    
    #merging diff exprs data
    diff_exprs_results <- merge(x = diff_exprs_results_gfp, 
                                y = diff_exprs_results_il2, 
                                by = "row.names", 
                                all = T)
    diff_exprs_results_2 <- merge(x = diff_exprs_results_04m, 
                                y = diff_exprs_results_24m, 
                                by = "row.names", 
                                all = T)
    diff_exprs_results <- merge(x = diff_exprs_results, 
                                y = diff_exprs_results_2, 
                                by = "Row.names", 
                                all = T)
    rownames(diff_exprs_results) <- diff_exprs_results$Row.names
    diff_exprs_results$Row.names <- NULL
    diff_exprs_results_2 <- NULL
    colnames(diff_exprs_results) <- paste0("Contrast ", 
                                           1:ncol(diff_exprs_results))
    
    #to exclude those extreme cases where a gene was not detected in a contrast
    diff_exprs_results <- diff_exprs_results[complete.cases(diff_exprs_results),]
    
    #setting up pathview
    data(gene.idtype.bods)
    work.dir <- getwd()
    path <- paste0(loc_pathview_results, "astro/striatal_pathview_treatment/")
    dir.create(path = path, recursive = T)
    setwd(path)
    
    #writing out the gage results
    write.table(x = gage_results_gfp$greater, 
                    file = "gage_results_GFP.tab", 
                    quote = F, 
                    sep = "\t", 
                    row.names = T)
    write.table(x = gage_results_il2$greater, 
                    file = "gage_results_IL2.tab", 
                    quote = F, 
                    sep = "\t", 
                    row.names = T)
    write.table(x = gage_results_04m$greater, 
                    file = "gage_results_04m.tab", 
                    quote = F, 
                    sep = "\t", 
                    row.names = T)
    write.table(x = gage_results_24m$greater, 
                    file = "gage_results_24m.tab", 
                    quote = F, 
                    sep = "\t", 
                    row.names = T)
    
    #getting pathway ids from the gage results
    pathway_ids <- 
      c(rownames(gage_results_gfp$greater), 
        rownames(gage_results_il2$greater), 
        rownames(gage_results_04m$greater), 
        rownames(gage_results_24m$greater))
    pathway_ids <- gsub(pattern = " .*$", 
                        replacement = "", 
                        x = pathway_ids, 
                        ignore.case = T)
    pathway_ids <- unique(pathway_ids)
    #adding missing pathways
    pathway_ids <- c(pathway_ids, "mmu05022", "mmu04370", "mmu04012", "mmu04014", 
                     "mmu04211", "mmu04657", "mmu01200", "mmu04310", "mmu04024", 
                     "mmu04213", "mmu04371", "mmu04145", "mmu04150", "mmu04015", 
                     "mmu04722", "mmu04068", "mmu04210", "mmu04218", "mmu04136")
    
    #running pathview in batches due to an undocumented limitation of a maximum
    # number of KEGG ids that can be queried in a single call. Not sure what value
    # and whether the limitation is in pathview or on KEGG's end
    pathway_ids <- split(pathway_ids, ceiling(seq_along(pathway_ids)/20))
    for(batch in pathway_ids){
      pathview(gene.data = diff_exprs_results, 
               cpd.data = NULL, 
               pathway.id = batch, 
               species = "mouse", 
               kegg.dir = path, 
               gene.idtype = "SYMBOL", 
               low = list(gene = "#2a7fff", cpd = "blue"), 
               mid = list(gene = "white", cpd = "gray"), 
               high = list(gene = "#ff5555", cpd = "yellow"), 
               limit = list(gene = 0.5, cpd = 1), 
               bins = list(gene = 10, cpd = 10), 
               kegg.native = T, 
               na.col = "gray", 
               same.layer = F)
      }
    
    #resetting work dir
    setwd(work.dir)
    
  }
  
  
}
#running gagePipe and gageComp on all micro
{
  #loading metadata
  metadata <- astro_data@meta.data
  
  #creating the customised kegg.gs object
  kegg.gs <- kegg_genesets$kg.sets
  
  #getting cell names and finding column indices
  {
    old_gfp_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "24m.PHP.GFAP-GFP"))
    old_il2_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "24m.PHP.GFAP-IL2"))
    young_gfp_cells <- rownames(subset(metadata,
                                 metadata$age_treatment == "04m.PHP.GFAP-GFP"))
    young_il2_cells <- rownames(subset(metadata, 
                                 metadata$age_treatment == "04m.PHP.GFAP-IL2"))
    
    old_gfp_cells <- which(colnames(astro_data[["SCT"]]@data) %in% 
                             old_gfp_cells)
    old_il2_cells <- which(colnames(astro_data[["SCT"]]@data) %in% 
                             old_il2_cells)
    young_gfp_cells <- which(colnames(astro_data[["SCT"]]@data) %in% 
                               young_gfp_cells)
    young_il2_cells <- which(colnames(astro_data[["SCT"]]@data) %in% 
                               young_il2_cells)
  }
  
  #creating sample and reference lists
  sampList <- list(old_gfp_cells, old_il2_cells, young_il2_cells, old_il2_cells)
  refList <- list(young_gfp_cells, young_il2_cells, young_gfp_cells, 
                   old_gfp_cells)
  sampnames <- c("c_GFP.24m.vs.GFP.04m", 
                 "c_IL2.24m.vs.IL2.04m", 
                 "c_04m.IL2.vs.04m.GFP", 
                 "c_24m.IL2.vs.24m.GFP")
  
  #loading counts and mapping ids
  # do check that the order is the same
  # all(unique(rownames(counts_data) == genes$SYMBOL))
  counts_data <- as.data.frame(astro_data@assays$SCT@data)
  counts_data$ENTREZID <- genes$ENTREZID
  counts_data <- counts_data[complete.cases(counts_data),]
  rownames(counts_data) <- counts_data$ENTREZID
  counts_data$ENTREZID <- NULL
  
  #saving work.dir and restoring it later
  work.dir <- getwd()
  setwd(paste0(loc_pathview_results, "astro/gage_combined/"))
  
  #running gage pipe
  gagePipe(arraydata = counts_data, 
           dataname = "All.Astrocytes", trim.at = F, 
           sampnames = sampnames, 
           gsname = "kegg.gs", 
           ref.list = refList, 
           samp.list = sampList, 
           comp.list = "as.group", 
           heatmap = T)
  
  #running gage comparison
  gageComp(sampnames = sampnames, 
           dataname = "All.Astrocytes", 
           gsname = "kegg.gs")
  
  #restoring work.dir
  setwd(work.dir)
  
  #garbage collection
  gc()
  
  #saving work.dir for the following loop and restoring it later
  work.dir <- getwd()
  setwd(paste0(loc_pathview_results, "astro/gage_combined/"))
  
  for(cell_type in unique(metadata$seurat_clusters_renamed)){
    #getting cell names and finding column indices
    {
      old_gfp_cells <- 
        rownames(subset(metadata, 
                        metadata$age_treatment == "24m.PHP.GFAP-GFP" & 
                          metadata$seurat_clusters_renamed == cell_type))
      old_il2_cells <- 
        rownames(subset(metadata, 
                        metadata$age_treatment == "24m.PHP.GFAP-IL2" & 
                          metadata$seurat_clusters_renamed == cell_type))
      young_gfp_cells <- 
        rownames(subset(metadata, 
                        metadata$age_treatment == "04m.PHP.GFAP-GFP" & 
                          metadata$seurat_clusters_renamed == cell_type))
      young_il2_cells <- 
        rownames(subset(metadata, 
                        metadata$age_treatment == "04m.PHP.GFAP-IL2" & 
                          metadata$seurat_clusters_renamed == cell_type))
      
      old_gfp_cells <- which(colnames(astro_data[["SCT"]]@data) %in% 
                               old_gfp_cells)
      old_il2_cells <- which(colnames(astro_data[["SCT"]]@data) %in% 
                               old_il2_cells)
      young_gfp_cells <- which(colnames(astro_data[["SCT"]]@data) %in% 
                                 young_gfp_cells)
      young_il2_cells <- which(colnames(astro_data[["SCT"]]@data) %in% 
                                 young_il2_cells)
    }
    
    #creating sample and reference lists
    sampList <- list(old_gfp_cells, old_il2_cells, 
                     young_il2_cells, old_il2_cells)
    refList <- list(young_gfp_cells, young_il2_cells, 
                    young_gfp_cells, old_gfp_cells)
    cell_type <- gsub(pattern = "(\n| |-)", 
                      replacement = ".", 
                      x = cell_type, 
                      ignore.case = T)
    sampnames <- c("c_GFP.24m.vs.GFP.04m", 
                   "c_IL2.24m.vs.IL2.04m", 
                   "c_04m.IL2.vs.04m.GFP", 
                   "c_24m.IL2.vs.24m.GFP")
    
    #running gage pipe
    gagePipe(arraydata = counts_data, 
             dataname = cell_type, trim.at = F, 
             sampnames = sampnames, 
             gsname = "kegg.gs", 
             ref.list = refList, 
             samp.list = sampList, 
             comp.list = "as.group", 
             heatmap = T)
    
    #running gage comparison
    gageComp(sampnames = sampnames, 
             dataname = cell_type, 
             gsname = "kegg.gs")
  }
  
  #restoring work.dir
  setwd(work.dir)
  
  #garbage collection
  gc()
  
}
#creating the manual summarised comparison -- horizontal/cbind -- ref contrast 
# -- only doing this for wide format table
{
  gene_sets <- kegg_genesets$kg.sets
  
  for(i in seq(1, length(gene_sets))){
    gene_sets[[i]] <- genes$SYMBOL[which(genes$ENTREZID %in% gene_sets[[i]])]
  }
  
  #loading differential expression results for each subpop
  subpops <- paste0("astro/subpops/", 
                    c("Bergmann.Glia", "Cerebellar.Astrocytes", 
                      "Non-telencephalon.Astrocytes", "Olfactory.Astrocytes", 
                      "Striatal.Astrocytes", "Telencephalon.Astrocytes", 
                      "All.Astrocytes"))
  
  #initialising genes_to_select for the loop to get a unified list for all astro
  genes_to_select <- character()
  for(subpop in subpops){
    #loading differential expression results
    #loading GFP results
    diff_exprs_results_gfp <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            subpop, 
                            "_markers_list_gfp_nofilter.rds"))[[2]]
    diff_exprs_results_gfp <-
      diff_exprs_results_gfp[diff_exprs_results_gfp$p_val_adj < 0.1, "avg_log2FC",
                         drop = F]
    
    #loading IL2 results
    diff_exprs_results_il2 <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            subpop, 
                            "_markers_list_il2_nofilter.rds"))[[2]]
    diff_exprs_results_il2 <-
      diff_exprs_results_il2[diff_exprs_results_il2$p_val_adj < 0.1, "avg_log2FC",
                         drop = F]
    
    #loading 04m results
    diff_exprs_results_04m <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            subpop, 
                            "_markers_list_04m_nofilter.rds"))[[2]]
    diff_exprs_results_04m <-
      diff_exprs_results_04m[diff_exprs_results_04m$p_val_adj < 0.1, "avg_log2FC",
                         drop = F]
    
    #loading 24m results
    diff_exprs_results_24m <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            subpop, 
                            "_markers_list_24m_nofilter.rds"))[[2]]
    diff_exprs_results_24m <-
      diff_exprs_results_24m[diff_exprs_results_24m$p_val_adj < 0.1, "avg_log2FC",
                         drop = F] 
    
    genes_to_select <- unique(c(genes_to_select, 
                                rownames(diff_exprs_results_gfp), 
                                rownames(diff_exprs_results_il2), 
                                rownames(diff_exprs_results_04m), 
                                rownames(diff_exprs_results_24m)))
  }
  
  #constructing data
  avg_sign_df_list <- list()
  for(subpop in subpops){
    #loading differential expression results
    {
      #loading GFP results
      diff_exprs_results_gfp <- 
        readRDS(file = paste0(loc_diff_exprs_results, 
                              subpop, 
                              "_markers_list_gfp_nofilter.rds"))[[2]]
      diff_exprs_results_gfp <-
        diff_exprs_results_gfp[genes_to_select, "avg_log2FC",
                           drop = F]
      
      #loading IL2 results
      diff_exprs_results_il2 <- 
        readRDS(file = paste0(loc_diff_exprs_results, 
                              subpop, 
                              "_markers_list_il2_nofilter.rds"))[[2]]
      diff_exprs_results_il2 <-
        diff_exprs_results_il2[genes_to_select, "avg_log2FC",
                           drop = F]
      
      #loading 04m results
      diff_exprs_results_04m <- 
        readRDS(file = paste0(loc_diff_exprs_results, 
                              subpop, 
                              "_markers_list_04m_nofilter.rds"))[[2]]
      diff_exprs_results_04m <-
        diff_exprs_results_04m[genes_to_select, "avg_log2FC",
                           drop = F]
      
      #loading 24m results
      diff_exprs_results_24m <- 
        readRDS(file = paste0(loc_diff_exprs_results, 
                              subpop, 
                              "_markers_list_24m_nofilter.rds"))[[2]]
      diff_exprs_results_24m <-
        diff_exprs_results_24m[genes_to_select, "avg_log2FC",
                           drop = F]
    }
    
    #formatting data frame of expression directions
    avg_sign_df <- data.frame(A = double(), B = double(), 
                              C = double(), D = double())
    for(i in seq(1, length(gene_sets))){
      #importing contrast data
      cont_ref <- diff_exprs_results_gfp[gene_sets[[i]],]
      cont_a <- diff_exprs_results_il2[gene_sets[[i]],]
      cont_b <- diff_exprs_results_04m[gene_sets[[i]],]
      cont_c <- diff_exprs_results_24m[gene_sets[[i]],]
      
      #checking which indices are negative in ref contrast
      flip_index <- which(cont_ref < 0)
      
      #if even a single negative value is found
      if(length(flip_index) > 0){
        cont_ref[flip_index] <- -(cont_ref[flip_index])
        cont_a[flip_index] <- -(cont_a[flip_index])
        cont_b[flip_index] <- -(cont_b[flip_index])
        cont_c[flip_index] <- -(cont_c[flip_index])
      }
      
      #formatting entry
      entry <- 
        c(mean(sign(cont_ref), na.rm = T), 
          mean(sign(cont_a), na.rm = T), 
          mean(sign(cont_b), na.rm = T), 
          mean(sign(cont_c), na.rm = T)
          )
      
      #if nothing was found in any contrast (no qualifying genes in any contrast
      # for a given pathway)
      if(all(is.na(entry))){
        next
      }
      
      #appending entry
      avg_sign_df <- rbind(avg_sign_df, entry)
      rownames(avg_sign_df)[nrow(avg_sign_df)] <- names(gene_sets)[i]
    }
    
    #assigning colnames
    colnames(avg_sign_df) <- c("GFP 24m vs GFP 04m (ref)", 
                               "IL2 24m vs IL2 04m", 
                               "04m IL2 vs 04m GFP", 
                               "24m IL2 vs 24m GFP")
    
    # #adding Pathway and Cell type information -- only used for long format
    # avg_sign_df$Pathway <- rownames(avg_sign_df)
    # avg_sign_df$Cell <- gsub(pattern = "^.*/", 
    #                          replacement = "", 
    #                          x = subpop, 
    #                          ignore.case = T)
    
    #appending to list and named
    avg_sign_df_list <- append(avg_sign_df_list, list(avg_sign_df))
    names(avg_sign_df_list)[length(avg_sign_df_list)] <- 
      gsub(pattern = "^.*/", 
           replacement = "", 
           x = subpop, 
           ignore.case = T)
  }
  
  #wide format requires same gene_sets/pathways in all contrasts
  common_rows <- rownames(avg_sign_df_list[[1]])
  for(i in seq(2, length(avg_sign_df_list))){
    common_rows <- intersect(common_rows, rownames(avg_sign_df_list[[i]]))
  }
  #retaining only common rows
  for(i in seq(1, length(avg_sign_df_list))){
    avg_sign_df_list[[i]] <- avg_sign_df_list[[i]][common_rows,]
  }
  
  #cbinding to a single dataframe
  avg_sign_df <- do.call(cbind.data.frame, avg_sign_df_list)
  
  #removing all na/nan rows
  avg_sign_df <- 
    avg_sign_df[!is.na(rowSums(avg_sign_df, na.rm = T)),]
  
  #writing out full results
  write.table(x = avg_sign_df %>% rownames_to_column(var = "Pathway"), 
              file = paste0(loc_04mv24m_results, 
                            "/00_figures/Figure6/", 
                            "astrocytes_direction_data_wide.tab"), 
              quote = F, sep = "\t", row.names = F, col.names = T)
  
}
#creating the manual summarised comparison -- vertical/rbind/faceted -- ref cont
# -- figure 6
{
  gene_sets <- kegg_genesets$kg.sets
  
  for(i in seq(1, length(gene_sets))){
    gene_sets[[i]] <- genes$SYMBOL[which(genes$ENTREZID %in% gene_sets[[i]])]
  }
  
  #loading differential expression results for each subpop
  subpops <- paste0("astro/subpops/", 
                    c("Bergmann.Glia", "Cerebellar.Astrocytes", 
                      "Non-telencephalon.Astrocytes", "Olfactory.Astrocytes", 
                      "Striatal.Astrocytes", "Telencephalon.Astrocytes", 
                      "All.Astrocytes"))
  #initialising genes_to_select for the loop to get a unified list for all astro
  genes_to_select <- character()
  for(subpop in subpops){
    #loading differential expression results
    #loading GFP results
    diff_exprs_results_gfp <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            subpop, 
                            "_markers_list_gfp_nofilter.rds"))[[2]]
    diff_exprs_results_gfp <-
      diff_exprs_results_gfp[diff_exprs_results_gfp$p_val_adj < 0.1, "avg_log2FC",
                         drop = F]
    
    #loading IL2 results
    diff_exprs_results_il2 <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            subpop, 
                            "_markers_list_il2_nofilter.rds"))[[2]]
    diff_exprs_results_il2 <-
      diff_exprs_results_il2[diff_exprs_results_il2$p_val_adj < 0.1, "avg_log2FC",
                         drop = F]
    
    #loading 04m results
    diff_exprs_results_04m <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            subpop, 
                            "_markers_list_04m_nofilter.rds"))[[2]]
    diff_exprs_results_04m <-
      diff_exprs_results_04m[diff_exprs_results_04m$p_val_adj < 0.1, "avg_log2FC",
                         drop = F]
    
    #loading 24m results
    diff_exprs_results_24m <- 
      readRDS(file = paste0(loc_diff_exprs_results, 
                            subpop, 
                            "_markers_list_24m_nofilter.rds"))[[2]]
    diff_exprs_results_24m <-
      diff_exprs_results_24m[diff_exprs_results_24m$p_val_adj < 0.1, "avg_log2FC",
                         drop = F] 
    
    genes_to_select <- unique(c(genes_to_select, 
                                rownames(diff_exprs_results_gfp), 
                                rownames(diff_exprs_results_il2), 
                                rownames(diff_exprs_results_04m), 
                                rownames(diff_exprs_results_24m)))
  }
  
  #constructing data
  avg_sign_df_list <- list()
  for(subpop in subpops){
    #loading differential expression results
    {
      #loading GFP results
      diff_exprs_results_gfp <- 
        readRDS(file = paste0(loc_diff_exprs_results, 
                              subpop, 
                              "_markers_list_gfp_nofilter.rds"))[[2]]
      diff_exprs_results_gfp <-
        diff_exprs_results_gfp[genes_to_select, "avg_log2FC",
                           drop = F]
      
      #loading IL2 results
      diff_exprs_results_il2 <- 
        readRDS(file = paste0(loc_diff_exprs_results, 
                              subpop, 
                              "_markers_list_il2_nofilter.rds"))[[2]]
      diff_exprs_results_il2 <-
        diff_exprs_results_il2[genes_to_select, "avg_log2FC",
                           drop = F]
      
      #loading 04m results
      diff_exprs_results_04m <- 
        readRDS(file = paste0(loc_diff_exprs_results, 
                              subpop, 
                              "_markers_list_04m_nofilter.rds"))[[2]]
      diff_exprs_results_04m <-
        diff_exprs_results_04m[genes_to_select, "avg_log2FC",
                           drop = F]
      
      #loading 24m results
      diff_exprs_results_24m <- 
        readRDS(file = paste0(loc_diff_exprs_results, 
                              subpop, 
                              "_markers_list_24m_nofilter.rds"))[[2]]
      diff_exprs_results_24m <-
        diff_exprs_results_24m[genes_to_select, "avg_log2FC",
                           drop = F]
    }
    
    #formatting data frame of expression directions
    avg_sign_df <- data.frame(A = double(), B = double(), 
                              C = double(), D = double())
    for(i in seq(1, length(gene_sets))){
      #importing contrast data
      cont_ref <- diff_exprs_results_gfp[gene_sets[[i]],]
      cont_a <- diff_exprs_results_il2[gene_sets[[i]],]
      cont_b <- diff_exprs_results_04m[gene_sets[[i]],]
      cont_c <- diff_exprs_results_24m[gene_sets[[i]],]
      
      #checking which indices are negative in ref contrast
      flip_index <- which(cont_ref < 0)
      
      #if even a single negative value is found
      if(length(flip_index) > 0){
        cont_ref[flip_index] <- -(cont_ref[flip_index])
        cont_a[flip_index] <- -(cont_a[flip_index])
        cont_b[flip_index] <- -(cont_b[flip_index])
        cont_c[flip_index] <- -(cont_c[flip_index])
      }
      
      #formatting entry
      entry <- 
        c(mean(sign(cont_ref), na.rm = T), 
          mean(sign(cont_a), na.rm = T), 
          mean(sign(cont_b), na.rm = T), 
          mean(sign(cont_c), na.rm = T)
          )
      
      #if nothing was found in any contrast (no qualifying genes in any contrast
      # for a given pathway)
      if(all(is.na(entry))){
        next
      }
      
      #appending entry
      avg_sign_df <- rbind(avg_sign_df, entry)
      rownames(avg_sign_df)[nrow(avg_sign_df)] <- names(gene_sets)[i]
    }
    
    #assigning colnames
    colnames(avg_sign_df) <- c("GFP 24m vs GFP 04m (ref)", "IL2 24m vs IL2 04m", 
                             "04m IL2 vs 04m GFP", "24m IL2 vs 24m GFP")
    
    #adding Pathway and Cell type information
    avg_sign_df$Pathway <- rownames(avg_sign_df)
    avg_sign_df$Cell <- gsub(pattern = "^.*/", 
                             replacement = "", 
                             x = subpop, 
                             ignore.case = T)
    
    #appending to list and named
    avg_sign_df_list <- append(avg_sign_df_list, list(avg_sign_df))
    names(avg_sign_df_list)[length(avg_sign_df_list)] <- 
      gsub(pattern = "^.*/", 
           replacement = "", 
           x = subpop, 
           ignore.case = T)
    
  }
  
  #rbinding list to a single dataframe
  avg_sign_df <- do.call(rbind.data.frame, avg_sign_df_list)
  
  #removing all na/nan rows
  avg_sign_df <- 
    avg_sign_df[!is.na(rowSums(avg_sign_df[,1:4], na.rm = T)),]
  
  #writing out full results
  write.table(x = avg_sign_df, 
              file = paste0(loc_04mv24m_results, 
                            "/00_figures/Figure6/", 
                            "astrocytes_direction_data_long.tab"), 
              quote = F, sep = "\t", row.names = F, col.names = T)
  
  #reordering data frame for heatmap
  ordering <- rowMeans(avg_sign_df[,1:4], na.rm = T)
  ordering <- ordering[order(ordering, decreasing = T)]
  avg_sign_df <- 
    rbind(avg_sign_df[rownames(avg_sign_df) %in% names(ordering),], 
          avg_sign_df[!rownames(avg_sign_df) %in% names(ordering),])
  ordering <- rev(rownames(avg_sign_df))
  #removing leading cell type from the ordering
  ordering <- gsub(pattern = "^.*\\.", 
                   replacement = "", 
                   x = ordering, 
                   ignore.case = T)
  #retaining only unique pathways in ordering for making factor later on
  ordering <- unique(ordering)
  
  #removing '.' from cell names for facet labels
  avg_sign_df$Cell <- gsub(pattern = "\\.", 
                           replacement = " ", 
                           x = avg_sign_df$Cell, 
                           ignore.case = T)
  
  #importing pre-selected pathways
  selected_mmu <- c("mmu05010", "mmu05012", "mmu04910", "mmu03013", "mmu04066", 
                    "mmu00190", "mmu01200", "mmu03010", "mmu03040", "mmu03050", 
                    "mmu04140", "mmu04141", "mmu04142", "mmu04144", "mmu04150", 
                    "mmu04810")
  avg_sign_df <- avg_sign_df[str_extract(rownames(avg_sign_df),
                                         pattern = "mmu[0-9]+") %in%
                               selected_mmu,]
                               # selected_mmu$ID,]
  
  #word wrapping ordering to map to Pathway column later on
  ordering <- str_wrap(string = ordering, width = 35)
  #formatting word wrapping
  avg_sign_df$Pathway <- str_wrap(string = avg_sign_df$Pathway, width = 35)
  
  #converting to heatmap
  avg_sign_df <- avg_sign_df %>% 
    gather(colname, value, -Pathway, -Cell)
  
  #str wrapping for facet labels
  avg_sign_df$Cell <- str_wrap(string = avg_sign_df$Cell, width = 22)
  
  #converting pathway names and colnames to factor
  avg_sign_df$Pathway <- factor(x = avg_sign_df$Pathway, 
                                levels = ordering, ordered = T)
  avg_sign_df$colname <- factor(x = avg_sign_df$colname, 
                                levels = c("GFP 24m vs GFP 04m (ref)", 
                                           "IL2 24m vs IL2 04m", 
                                           "04m IL2 vs 04m GFP", 
                                           "24m IL2 vs 24m GFP"), 
                                ordered = T)
  
  #inverting values for colour scale in ggplot2
  avg_sign_df$value <- -(avg_sign_df$value)
  
  ggplot(avg_sign_df, 
         aes(x = colname, 
             y = Pathway, 
             fill = value)) + 
    geom_tile() + 
    scale_fill_gradient2(low = ("#3DBEDB"), #munsell::mnsl("5B 7/8")
                         mid = "white", 
                         high = ("#CAAE42"), #munsell::mnsl("5Y 7/8") 
                         breaks = c(-1, -0.5, 0, 0.5, 1), 
                         labels = c("Closer to\naged control", "", "", "", 
                                    "Closer to\nyoung control"), 
                         limits = c(-1, 1)) + 
    xlab(NULL) + ylab(NULL) + labs(fill = NULL) + 
    theme_light() + 
    theme(axis.text.x = element_text(size = 13, angle = 25, 
                                     hjust = 0.9, vjust = 0.9), 
          text = element_text(size = 16), 
          axis.text.y =  element_text(size = 14)) + 
    facet_grid(. ~ Cell)
  
}


```


